{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "18db57d9-6d77-4cce-b8f0-bacdd9b58ade",
   "metadata": {},
   "source": [
    "# Parabricks Hands-On Workshop"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc22de43-5a66-4913-b723-836b78a80db7",
   "metadata": {},
   "source": [
    "## Tutorial 1: Germline Workflow\n",
    "This workflow will demonstrate how to perform secondary analysis for DNA sequencing data from next-generation sequencing. This is relevant for population sequencing, rare disease detection, common disease sequencing, etc.\n",
    "- Sample: 30x whole-genome sequencing from a human individual, HC002.\n",
    "- Analysis: 1) Read alignment, 2) Variant calling\n",
    "\n",
    "To save time from downloading data, all necessary data has been put in TWCC Cloud Object Services (COS), including sample data and reference genome library."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "727be743-7a89-4242-83d3-3920884a7327",
   "metadata": {},
   "source": [
    "#### Download Data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af6a84dc-62c8-404a-b25d-d32c29c8e1b2",
   "metadata": {},
   "source": [
    "We will download the reference human genome, its index files, a known variants file, and \n",
    "a 30x whole-genome sequencing sample for germline analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1b939766-26a3-4d9f-93f3-3f6062c42627",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "URL transformed to HTTPS due to an HSTS policy\n",
      "--2025-09-03 12:17:31--  https://cos.twcc.ai/pbworkshop/hg38/parabricks_sample.tar.gz\n",
      "Resolving cos.twcc.ai (cos.twcc.ai)... 203.145.219.21\n",
      "Connecting to cos.twcc.ai (cos.twcc.ai)|203.145.219.21|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 11767623602 (11G) [binary/octet-stream]\n",
      "Saving to: ‘parabricks_sample.tar.gz.1’\n",
      "\n",
      "parabricks_sample.t 100%[===================>]  10.96G   173MB/s    in 1m 41s  \n",
      "\n",
      "2025-09-03 12:19:12 (112 MB/s) - ‘parabricks_sample.tar.gz.1’ saved [11767623602/11767623602]\n",
      "\n",
      "parabricks_sample/\n",
      "parabricks_sample/Data/\n",
      "parabricks_sample/Data/markdup_input.bam\n",
      "parabricks_sample/Data/sample_2.fq.gz\n",
      "parabricks_sample/Data/sample_1.fq.gz\n",
      "parabricks_sample/Data/single_ended.bam\n",
      "parabricks_sample/Ref/\n",
      "parabricks_sample/Ref/Homo_sapiens_assembly38.fasta.sa\n",
      "parabricks_sample/Ref/Homo_sapiens_assembly38.known_indels.vcf.gz.tbi\n",
      "parabricks_sample/Ref/Homo_sapiens_assembly38.dict\n",
      "parabricks_sample/Ref/Homo_sapiens_assembly38.fasta\n",
      "parabricks_sample/Ref/Homo_sapiens_assembly38.fasta.fai\n",
      "parabricks_sample/Ref/Homo_sapiens_assembly38.fasta.pac\n",
      "parabricks_sample/Ref/Homo_sapiens_assembly38.known_indels.vcf.gz\n",
      "parabricks_sample/Ref/Homo_sapiens_assembly38.fasta.bwt\n",
      "parabricks_sample/Ref/Homo_sapiens_assembly38.fasta.ann\n",
      "parabricks_sample/Ref/Homo_sapiens_assembly38.fasta.amb\n",
      "mv: cannot move 'parabricks_sample/Ref' to './Ref': File exists\n"
     ]
    }
   ],
   "source": [
    "# The tar file is 9.3GB and, when extracted, an additional 14GB\n",
    "!wget http://cos.twcc.ai/pbworkshop/hg38/parabricks_sample.tar.gz\n",
    "!tar -xzvf parabricks_sample.tar.gz\n",
    "!mv parabricks_sample/* .\n",
    "!rm -r parabricks_sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "af9bccea-d790-4073-9937-3c32718f1faf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "markdup_input.bam  sample_1.fq.gz  sample_2.fq.gz  single_ended.bam\n"
     ]
    }
   ],
   "source": [
    "!mv Data germline_data\n",
    "!ls germline_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4449d1d7-09a9-459d-b71d-7ccc75037583",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Homo_sapiens_assembly38.dict\n",
      "Homo_sapiens_assembly38.fasta\n",
      "Homo_sapiens_assembly38.fasta.amb\n",
      "Homo_sapiens_assembly38.fasta.ann\n",
      "Homo_sapiens_assembly38.fasta.bwt\n",
      "Homo_sapiens_assembly38.fasta.fai\n",
      "Homo_sapiens_assembly38.fasta.pac\n",
      "Homo_sapiens_assembly38.fasta.sa\n",
      "Homo_sapiens_assembly38.known_indels.vcf.gz\n",
      "Homo_sapiens_assembly38.known_indels.vcf.gz.tbi\n",
      "ctat_genome_lib_build_dir\n",
      "gencode.v48.primary_assembly.annotation.gtf\n",
      "genomelib\n",
      "genomelib.tar.gz\n"
     ]
    }
   ],
   "source": [
    "!ls Ref"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "daf00c34-1843-4502-a212-218c4346d803",
   "metadata": {},
   "source": [
    "Make a new directory to store the output from this tutorial."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3e865f3a-813c-4435-825b-0f022207f713",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘germline_output’: File exists\n"
     ]
    }
   ],
   "source": [
    "!mkdir germline_output"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81d854c7-c48b-4e62-a843-eadcca5ab261",
   "metadata": {},
   "source": [
    "#### GPU Monitoring"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d5523e52-751a-4059-9751-f36883738c2f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wed Sep  3 12:20:44 2025       \n",
      "+---------------------------------------------------------------------------------------+\n",
      "| NVIDIA-SMI 535.161.08             Driver Version: 535.161.08   CUDA Version: 12.3     |\n",
      "|-----------------------------------------+----------------------+----------------------+\n",
      "| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |\n",
      "| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |\n",
      "|                                         |                      |               MIG M. |\n",
      "|=========================================+======================+======================|\n",
      "|   0  Tesla V100-SXM2-32GB           On  | 00000000:1B:00.0 Off |                    0 |\n",
      "| N/A   27C    P0              42W / 300W |      0MiB / 32768MiB |      0%      Default |\n",
      "|                                         |                      |                  N/A |\n",
      "+-----------------------------------------+----------------------+----------------------+\n",
      "                                                                                         \n",
      "+---------------------------------------------------------------------------------------+\n",
      "| Processes:                                                                            |\n",
      "|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |\n",
      "|        ID   ID                                                             Usage      |\n",
      "|=======================================================================================|\n",
      "|  No running processes found                                                           |\n",
      "+---------------------------------------------------------------------------------------+\n"
     ]
    }
   ],
   "source": [
    "!nvidia-smi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0374156-867b-4477-a3bf-1bfcd17fef31",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run the command below in the terminal\n",
    "### watch -n 0.5 nvidia-smi\n",
    "#"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cefbf268-8ad8-40ca-9433-b63312d3068e",
   "metadata": {},
   "source": [
    "### Alignment: FASTQ to BAM"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f2513fc-1ee5-4059-8edf-375b76e7fb75",
   "metadata": {},
   "source": [
    "The alignment and variant calling steps in the germline workflow can run separately with these commands. Again, `--bwa-nstreams` is set to `1` because we are limited to a single V100 GPU. With more GPU memory, you can run more streams (i.g., 16) and shorten the time for running."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8a661081-f1c6-4e63-bfc5-e15eee8f2a31",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please visit https://docs.nvidia.com/clara/#parabricks for detailed documentation\n",
      "\n",
      "\n",
      "\n",
      "[Parabricks Options Mesg]: Checking argument compatibility\n",
      "[Parabricks Options Mesg]: Set --bwa-options=\"-K #\" to produce compatible pair-ended results with previous versions of\n",
      "fq2bam or BWA MEM.\n",
      "[Parabricks Options Mesg]: Automatically generating ID prefix\n",
      "[Parabricks Options Mesg]: Read group created for /home/yingja1227/germline_data/sample_1.fq.gz and\n",
      "/home/yingja1227/germline_data/sample_2.fq.gz\n",
      "[Parabricks Options Mesg]: @RG\\tID:HK3TJBCX2.1\\tLB:lib1\\tPL:bar\\tSM:sample\\tPU:HK3TJBCX2.1\n",
      "[PB Info 2025-Sep-03 12:20:44] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Sep-03 12:20:44] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2025-Sep-03 12:20:44] ||                              Version 4.4.0-1                             ||\n",
      "[PB Info 2025-Sep-03 12:20:44] ||                      GPU-PBBWA mem, Sorting Phase-I                      ||\n",
      "[PB Info 2025-Sep-03 12:20:44] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Sep-03 12:20:44] Mode = pair-ended-gpu\n",
      "[PB Info 2025-Sep-03 12:20:44] Running with 1 GPU(s), using 1 stream(s) per device with 16 worker threads per GPU\n",
      "[PB Info 2025-Sep-03 12:20:54] # 79  4  0  1  0   0 pool:  0 0 bases/GPU/minute: 0.0 \n",
      "[PB Info 2025-Sep-03 12:21:04] # 100  0  6  0  0   0 pool:  1 248787200 bases/GPU/minute: 1492723200.0 \n",
      "[PB Info 2025-Sep-03 12:21:14] # 100  0  6  0  0   0 pool:  1 407102515 bases/GPU/minute: 949891890.0 \n",
      "[PB Info 2025-Sep-03 12:21:24] # 100  0  6  0  0   0 pool:  1 542801298 bases/GPU/minute: 814192698.0 \n",
      "[PB Info 2025-Sep-03 12:21:34] # 100  2  4  1  0   0 pool:  0 716193003 bases/GPU/minute: 1040350230.0 \n",
      "[PB Info 2025-Sep-03 12:21:44] # 100 21  0  6  0   0 pool:  0 851893171 bases/GPU/minute: 814201008.0 \n",
      "[PB Info 2025-Sep-03 12:21:54] # 100 34  0  6  0   0 pool:  0 1017752304 bases/GPU/minute: 995154798.0 \n",
      "[PB Info 2025-Sep-03 12:22:04] # 100 39  0  5  0   0 pool:  1 1198683695 bases/GPU/minute: 1085588346.0 \n",
      "[PB Info 2025-Sep-03 12:22:14] # 100 47  1  5  0   0 pool:  1 1356997291 bases/GPU/minute: 949881576.0 \n",
      "[PB Info 2025-Sep-03 12:22:24] # 100 57  0  5  0   0 pool:  1 1515307812 bases/GPU/minute: 949863126.0 \n",
      "[PB Info 2025-Sep-03 12:22:34] # 100 68  0  6  0   0 pool:  0 1673630338 bases/GPU/minute: 949935156.0 \n",
      "[PB Info 2025-Sep-03 12:22:44] # 100 84  0  6  0   0 pool:  0 1801791453 bases/GPU/minute: 768966690.0 \n",
      "[PB Info 2025-Sep-03 12:22:54] # 100 98  1  5  0   0 pool:  1 1945033152 bases/GPU/minute: 859450194.0 \n",
      "[PB Info 2025-Sep-03 12:23:04] # 100 100  1  5  0   0 pool:  0 2065653272 bases/GPU/minute: 723720720.0 \n",
      "[PB Info 2025-Sep-03 12:23:14] # 100 100  0  5  0   0 pool:  1 2246583288 bases/GPU/minute: 1085580096.0 \n",
      "[PB Info 2025-Sep-03 12:23:24] # 100 100  1  5  0   0 pool:  0 2419981116 bases/GPU/minute: 1040386968.0 \n",
      "[PB Info 2025-Sep-03 12:23:34] # 100 100  1  5  0   0 pool:  0 2563220743 bases/GPU/minute: 859437762.0 \n",
      "[PB Info 2025-Sep-03 12:23:44] # 100 100  1  5  0   0 pool:  0 2698922135 bases/GPU/minute: 814208352.0 \n",
      "[PB Info 2025-Sep-03 12:23:54] # 100 100  1  5  0   0 pool:  0 2834618337 bases/GPU/minute: 814177212.0 \n",
      "[PB Info 2025-Sep-03 12:24:04] # 100 100  1  5  0   0 pool:  0 2962776994 bases/GPU/minute: 768951942.0 \n",
      "[PB Info 2025-Sep-03 12:24:14] # 100 98  3  3  0   0 pool:  2 3143707962 bases/GPU/minute: 1085585808.0 \n",
      "[PB Info 2025-Sep-03 12:24:24] # 100 100  1  5  0   0 pool:  0 3324647367 bases/GPU/minute: 1085636430.0 \n",
      "[PB Info 2025-Sep-03 12:24:34] # 100 100  1  4  0   0 pool:  1 3482963598 bases/GPU/minute: 949897386.0 \n",
      "[PB Info 2025-Sep-03 12:24:44] # 100 98  3  3  0   0 pool:  2 3648816625 bases/GPU/minute: 995118162.0 \n",
      "[PB Info 2025-Sep-03 12:24:54] # 100 100  1  5  0   0 pool:  0 3814662958 bases/GPU/minute: 995077998.0 \n",
      "[PB Info 2025-Sep-03 12:25:04] # 100 96  5  1  0   0 pool:  4 3972982945 bases/GPU/minute: 949919922.0 \n",
      "[PB Info 2025-Sep-03 12:25:14] # 100 100  1  5  0   0 pool:  0 4116224068 bases/GPU/minute: 859446738.0 \n",
      "[PB Info 2025-Sep-03 12:25:24] # 100 99  2  4  0   0 pool:  1 4266998045 bases/GPU/minute: 904643862.0 \n",
      "[PB Info 2025-Sep-03 12:25:34] # 100 82 19  0  0   0 pool:  5 4493156720 bases/GPU/minute: 1356952050.0 \n",
      "[PB Info 2025-Sep-03 12:25:39] Time spent reading: 32.280512 seconds\n",
      "[PB Info 2025-Sep-03 12:25:44] # 83 85 16  9  0   0 pool: 13 4666551953 bases/GPU/minute: 1040371398.0 \n",
      "[PB Info 2025-Sep-03 12:25:54] # 50 95  6 19  0   0 pool: 36 4839950851 bases/GPU/minute: 1040393388.0 \n",
      "[PB Info 2025-Sep-03 12:26:04] # 19 100  1 25  0   0 pool: 61 5028423151 bases/GPU/minute: 1130833800.0 \n",
      "[PB Info 2025-Sep-03 12:26:14] #  0 96  0 26  0   0 pool: 85 5209354282 bases/GPU/minute: 1085586786.0 \n",
      "[PB Info 2025-Sep-03 12:26:24] #  0 73  0 26  0   0 pool: 100 5375204821 bases/GPU/minute: 995103234.0 \n",
      "[PB Info 2025-Sep-03 12:26:34] #  0 56  0 25  0   0 pool: 100 5503367187 bases/GPU/minute: 768974196.0 \n",
      "[PB Info 2025-Sep-03 12:26:44] #  0 34  0 25  0   0 pool: 100 5661687351 bases/GPU/minute: 949920984.0 \n",
      "[PB Info 2025-Sep-03 12:26:54] #  0  8  0 25  0   0 pool: 100 5865234268 bases/GPU/minute: 1221281502.0 \n",
      "[PB Info 2025-Sep-03 12:26:57] GPU 0 exited\n",
      "[PB Info 2025-Sep-03 12:27:04] #  0  0  0  5  0   0 pool: 100 6083859454 bases/GPU/minute: 1311751116.0 \n",
      "[PB Info 2025-Sep-03 12:27:06] A CPU has exited\n",
      "[PB Info 2025-Sep-03 12:27:14] #  0  0  0  0  0   0 pool: 100 6133353829 bases/GPU/minute: 296966250.0 \n",
      "[PB Info 2025-Sep-03 12:27:24] #  0  0  0  0  0   0 pool: 100 6133353829 bases/GPU/minute: 0.0 \n",
      "[PB Info 2025-Sep-03 12:27:34] Rate stats (based on sampling every 10 seconds):\n",
      "\tmin rate: 0.0 bases/GPU/minute\n",
      "\tmax rate: 1492723200.0 bases/GPU/minute\n",
      "\tavg rate: 920003074.4 bases/GPU/minute\n",
      "[PB Info 2025-Sep-03 12:27:34] Time spent monitoring (multiple of 10): 410.021\n",
      "[PB Info 2025-Sep-03 12:27:34] bwalib run finished in 404.077 seconds\n",
      "[PB Info 2025-Sep-03 12:27:34] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Sep-03 12:27:34] ||        Program:                    GPU-PBBWA mem, Sorting Phase-I        ||\n",
      "[PB Info 2025-Sep-03 12:27:34] ||        Version:                                           4.4.0-1        ||\n",
      "[PB Info 2025-Sep-03 12:27:34] ||        Start Time:                       Wed Sep  3 12:20:44 2025        ||\n",
      "[PB Info 2025-Sep-03 12:27:34] ||        End Time:                         Wed Sep  3 12:27:34 2025        ||\n",
      "[PB Info 2025-Sep-03 12:27:34] ||        Total Time:                           6 minutes 50 seconds        ||\n",
      "[PB Info 2025-Sep-03 12:27:34] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Sep-03 12:27:35] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Sep-03 12:27:35] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2025-Sep-03 12:27:35] ||                              Version 4.4.0-1                             ||\n",
      "[PB Info 2025-Sep-03 12:27:35] ||                             Sorting Phase-II                             ||\n",
      "[PB Info 2025-Sep-03 12:27:35] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Sep-03 12:27:35] progressMeter - Percentage\n",
      "[PB Info 2025-Sep-03 12:27:35] 0.0\n",
      "[PB Info 2025-Sep-03 12:27:40] 35.5\n",
      "[PB Info 2025-Sep-03 12:27:45] 73.0\n",
      "[PB Info 2025-Sep-03 12:27:50] Sorting and Marking: 15.003 seconds\n",
      "[PB Info 2025-Sep-03 12:27:50] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Sep-03 12:27:50] ||        Program:                                  Sorting Phase-II        ||\n",
      "[PB Info 2025-Sep-03 12:27:50] ||        Version:                                           4.4.0-1        ||\n",
      "[PB Info 2025-Sep-03 12:27:50] ||        Start Time:                       Wed Sep  3 12:27:35 2025        ||\n",
      "[PB Info 2025-Sep-03 12:27:50] ||        End Time:                         Wed Sep  3 12:27:50 2025        ||\n",
      "[PB Info 2025-Sep-03 12:27:50] ||        Total Time:                                     15 seconds        ||\n",
      "[PB Info 2025-Sep-03 12:27:50] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Sep-03 12:27:50] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Sep-03 12:27:50] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2025-Sep-03 12:27:50] ||                              Version 4.4.0-1                             ||\n",
      "[PB Info 2025-Sep-03 12:27:50] ||                         Marking Duplicates, BQSR                         ||\n",
      "[PB Info 2025-Sep-03 12:27:50] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Sep-03 12:27:50] Using PBBinBamFile for BAM writing\n",
      "[PB Info 2025-Sep-03 12:27:50] progressMeter -\tPercentage\n",
      "[PB Info 2025-Sep-03 12:28:00] 8.4\n",
      "[PB Info 2025-Sep-03 12:28:10] 18.2\n",
      "[PB Info 2025-Sep-03 12:28:20] 29.0\n",
      "[PB Info 2025-Sep-03 12:28:30] 43.8\n",
      "[PB Info 2025-Sep-03 12:28:40] 58.7\n",
      "[PB Info 2025-Sep-03 12:28:50] 74.0\n",
      "[PB Info 2025-Sep-03 12:29:00] 85.3\n",
      "[PB Info 2025-Sep-03 12:29:10] 100.0\n",
      "[PB Info 2025-Sep-03 12:29:10] BQSR and writing final BAM:  80.045 seconds\n",
      "[PB Info 2025-Sep-03 12:29:10] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Sep-03 12:29:10] ||        Program:                          Marking Duplicates, BQSR        ||\n",
      "[PB Info 2025-Sep-03 12:29:10] ||        Version:                                           4.4.0-1        ||\n",
      "[PB Info 2025-Sep-03 12:29:10] ||        Start Time:                       Wed Sep  3 12:27:50 2025        ||\n",
      "[PB Info 2025-Sep-03 12:29:10] ||        End Time:                         Wed Sep  3 12:29:10 2025        ||\n",
      "[PB Info 2025-Sep-03 12:29:10] ||        Total Time:                            1 minute 20 seconds        ||\n",
      "[PB Info 2025-Sep-03 12:29:10] ------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "!pbrun fq2bam \\\n",
    "      --ref Ref/Homo_sapiens_assembly38.fasta \\\n",
    "      --in-fq germline_data/sample_1.fq.gz germline_data/sample_2.fq.gz \\\n",
    "      --out-bam germline_output/fq2bam_output.bam \\\n",
    "      --num-gpus 1 \\\n",
    "      --bwa-nstreams 1 "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b640e24c-121d-417b-b7f6-c41f4b649d6a",
   "metadata": {},
   "source": [
    "### Variant Calling: BAM to VCF"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2e1e51d-b363-4cff-8cfa-2dec27843353",
   "metadata": {},
   "source": [
    "#### GATK Haplotypecaller"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4c4fb290-443b-403a-a72c-964912ca0f4c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please visit https://docs.nvidia.com/clara/#parabricks for detailed documentation\n",
      "\n",
      "usage: pbrun haplotypecaller <options>\n",
      "Help: pbrun haplotypecaller -h\n",
      "\n",
      "Run HaplotypeCaller to convert BAM/CRAM to VCF.\n",
      "\n",
      "options:\n",
      "  -h, --help            show this help message and exit\n",
      "\n",
      "Input Output file options:\n",
      "  Options for Input and Output files for this tool.\n",
      "\n",
      "  --ref REF             Path to the reference file. (default: None)\n",
      "  --in-bam IN_BAM       Path to the input BAM/CRAM file for variant calling.\n",
      "                        The argument may also be a local folder containing\n",
      "                        several BAM files. (default: None)\n",
      "  --in-recal-file IN_RECAL_FILE\n",
      "                        Path to the input BQSR report. (default: None)\n",
      "  --interval-file INTERVAL_FILE\n",
      "                        Path to an interval file in one of these formats:\n",
      "                        Picard-style (.interval_list or .picard), GATK-style\n",
      "                        (.list or .intervals), or BED file (.bed). This option\n",
      "                        can be used multiple times. (default: None)\n",
      "  --htvc-bam-output HTVC_BAM_OUTPUT\n",
      "                        File to which assembled haplotypes should be written.\n",
      "                        If passing with --run-partition, multiple bam files\n",
      "                        will be outputed. (default: None)\n",
      "  --out-variants OUT_VARIANTS\n",
      "                        Path of the vcf/g.vcf/gvcf.gz file after variant\n",
      "                        calling. (default: None)\n",
      "  --htvc-alleles HTVC_ALLELES\n",
      "                        Path of the vcf.gz force-call file. The set of alleles\n",
      "                        to force-call regardless of evidence. (default: None)\n",
      "\n",
      "Tool Options:\n",
      "  Options specific to the tool.\n",
      "\n",
      "  --haplotypecaller-options HAPLOTYPECALLER_OPTIONS\n",
      "                        Pass supported haplotype caller options as one string.\n",
      "                        The following are currently supported original\n",
      "                        haplotypecaller options: -A\n",
      "                        <AS_BaseQualityRankSumTest, AS_FisherStrand,\n",
      "                        AS_InbreedingCoeff, AS_MappingQualityRankSumTest,\n",
      "                        AS_QualByDepth, AS_RMSMappingQuality,\n",
      "                        AS_ReadPosRankSumTest, AS_StrandOddsRatio,\n",
      "                        BaseQualityRankSumTest, ChromosomeCounts,\n",
      "                        ClippingRankSumTest, Coverage, DepthPerAlleleBySample,\n",
      "                        DepthPerSampleHC, ExcessHet, FisherStrand,\n",
      "                        InbreedingCoeff, MappingQualityRankSumTest,\n",
      "                        QualByDepth, RMSMappingQuality, ReadPosRankSumTest,\n",
      "                        ReferenceBases, StrandBiasBySample, StrandOddsRatio,\n",
      "                        TandemRepeat, AssemblyComplexity>,-AX <same options as\n",
      "                        -A>,--output-mode <EMIT_VARIANTS_ONLY,\n",
      "                        EMIT_ALL_CONFIDENT_SITES, EMIT_ALL_ACTIVE_SITES>\n",
      "                        ,-max-reads-per-alignment-start <int>, -min-dangling-\n",
      "                        branch-length <int>, -min-pruning <int>, -pcr-indel-\n",
      "                        model <NONE, HOSTILE, AGGRESSIVE, CONSERVATIVE>,\n",
      "                        -standard-min-confidence-threshold-for-calling\n",
      "                        <int>(e.g. --haplotypecaller-options=\"-min-pruning 4\n",
      "                        -standard-min-confidence-threshold-for-calling 30\").\n",
      "                        (default: None)\n",
      "  --static-quantized-quals STATIC_QUANTIZED_QUALS\n",
      "                        Use static quantized quality scores to a given number\n",
      "                        of levels. Repeat this option multiple times for\n",
      "                        multiple bins. (default: None)\n",
      "  --gvcf                Generate variant calls in .gvcf Format. (default:\n",
      "                        None)\n",
      "  --disable-read-filter DISABLE_READ_FILTER\n",
      "                        Disable the read filters for BAM entries. Currently,\n",
      "                        the supported read filters that can be disabled are\n",
      "                        MappingQualityAvailableReadFilter,\n",
      "                        MappingQualityReadFilter,\n",
      "                        NotSecondaryAlignmentReadFilter, and\n",
      "                        WellformedReadFilter. (default: None)\n",
      "  --max-alternate-alleles MAX_ALTERNATE_ALLELES\n",
      "                        Maximum number of alternate alleles to genotype.\n",
      "                        (default: None)\n",
      "  -G ANNOTATION_GROUP, --annotation-group ANNOTATION_GROUP\n",
      "                        The groups of annotations to add to the output variant\n",
      "                        calls. Currently supported annotation groups are\n",
      "                        StandardAnnotation, StandardHCAnnotation, and\n",
      "                        AS_StandardAnnotation. (default: None)\n",
      "  -GQB GVCF_GQ_BANDS, --gvcf-gq-bands GVCF_GQ_BANDS\n",
      "                        Exclusive upper bounds for reference confidence GQ\n",
      "                        bands. Must be in the range [1, 100] and specified in\n",
      "                        increasing order. (default: None)\n",
      "  --rna                 Run haplotypecaller optimized for RNA data. (default:\n",
      "                        None)\n",
      "  --dont-use-soft-clipped-bases\n",
      "                        Don't use soft clipped bases for variant calling.\n",
      "                        (default: None)\n",
      "  --minimum-mapping-quality MINIMUM_MAPPING_QUALITY\n",
      "                        Minimum mapping quality to keep (inclusive). (default:\n",
      "                        None)\n",
      "  --mapping-quality-threshold-for-genotyping MAPPING_QUALITY_THRESHOLD_FOR_GENOTYPING\n",
      "                        Control the threshold for discounting reads from the\n",
      "                        genotyper due to mapping quality after the active\n",
      "                        region detection and assembly steps but before\n",
      "                        genotyping. (default: None)\n",
      "  --enable-dynamic-read-disqualification-for-genotyping\n",
      "                        Will enable less strict read disqualification low base\n",
      "                        quality reads. (default: None)\n",
      "  --min-base-quality-score MIN_BASE_QUALITY_SCORE\n",
      "                        Minimum base quality required to consider a base for\n",
      "                        calling. (default: None)\n",
      "  --adaptive-pruning    Use adaptive graph pruning algorithm when pruning De\n",
      "                        Bruijn graph. (default: None)\n",
      "  --force-call-filtered-alleles\n",
      "                        Force-call filtered alleles included in the resource\n",
      "                        specified by --alleles. (default: None)\n",
      "  --filter-reads-too-long\n",
      "                        Ignore all input bam reads with size > 500bp (default:\n",
      "                        None)\n",
      "  --no-alt-contigs      Get rid of output records for alternate contigs.\n",
      "                        (default: None)\n",
      "  --ploidy PLOIDY       Ploidy assumed for the BAM file. Currently only\n",
      "                        haploid (ploidy 1) and diploid (ploidy 2) are\n",
      "                        supported. (default: 2)\n",
      "  -L INTERVAL, --interval INTERVAL\n",
      "                        Interval within which to call the variants from the\n",
      "                        BAM/CRAM file. All intervals will have a padding of\n",
      "                        100 to get read records, and overlapping intervals\n",
      "                        will be combined. Interval files should be passed\n",
      "                        using the --interval-file option. This option can be\n",
      "                        used multiple times (e.g. \"-L chr1 -L chr2:10000 -L\n",
      "                        chr3:20000+ -L chr4:10000-20000\"). (default: None)\n",
      "  -ip INTERVAL_PADDING, --interval-padding INTERVAL_PADDING\n",
      "                        Amount of padding (in base pairs) to add to each\n",
      "                        interval you are including. (default: None)\n",
      "  --sample-sex SAMPLE_SEX\n",
      "                        Sex of the sample input. This option will override the\n",
      "                        sex determined from any X/Y read ratio range. Must be\n",
      "                        either male or female. (default: None)\n",
      "  --range-male RANGE_MALE\n",
      "                        Inclusive male range for the X/Y read ratio. The sex\n",
      "                        is declared male if the actual ratio falls in the\n",
      "                        specified range. Syntax is \"<min>-<max>\" (e.g. \"--\n",
      "                        range-male 1-10\"). (default: None)\n",
      "  --range-female RANGE_FEMALE\n",
      "                        Inclusive female range for the X/Y read ratio. The sex\n",
      "                        is declared female if the actual ratio falls in the\n",
      "                        specified range. Syntax is \"<min>-<max>\" (e.g. \"--\n",
      "                        range-female 150-250\"). (default: None)\n",
      "  --use-GRCh37-regions  Use the pseudoautosomal regions for GRCh37 reference\n",
      "                        types. This flag should be used for GRCh37 and UCSC\n",
      "                        hg19 references. By default, GRCh38 regions are used.\n",
      "                        (default: None)\n",
      "\n",
      "Performance Options:\n",
      "  Performance Tuning Options.\n",
      "\n",
      "  --htvc-low-memory     Use low memory mode in htvc. (default: None)\n",
      "  --num-htvc-threads NUM_HTVC_THREADS\n",
      "                        Number of CPU threads. (default: 5)\n",
      "  --run-partition       Divide the whole genome into multiple partitions and\n",
      "                        run multiple processes at the same time, each on one\n",
      "                        partition. (default: None)\n",
      "\n",
      "Run Options:\n",
      "  Options related to resource/verbosity/etc.\n",
      "\n",
      "  --verbose             Enable verbose output. (default: None)\n",
      "  --x3                  Show full command line arguments. (default: None)\n",
      "  --logfile LOGFILE     Path to the log file. If not specified, messages will\n",
      "                        only be written to the standard error output.\n",
      "                        (default: None)\n",
      "  --tmp-dir TMP_DIR     Full path to the directory where temporary files will\n",
      "                        be stored. (default: .)\n",
      "  --with-petagene-dir WITH_PETAGENE_DIR\n",
      "                        Full path to the PetaGene installation directory. By\n",
      "                        default, this should have been installed at\n",
      "                        /opt/petagene. Use of this option also requires that\n",
      "                        the PetaLink library has been preloaded by setting the\n",
      "                        LD_PRELOAD environment variable. Optionally set the\n",
      "                        PETASUITE_REFPATH and PGCLOUD_CREDPATH environment\n",
      "                        variables that are used for data and credentials.\n",
      "                        Optionally set the PetaLinkMode environment variable\n",
      "                        that is used to further configure PetaLink, notably\n",
      "                        setting it to \"+write\" to enable outputting compressed\n",
      "                        BAM and .fastq files. (default: None)\n",
      "  --keep-tmp            Do not delete the directory storing temporary files\n",
      "                        after completion. (default: None)\n",
      "  --no-seccomp-override\n",
      "                        Do not override seccomp options for docker. (default:\n",
      "                        None)\n",
      "  --version             View compatible software versions. (default: None)\n",
      "  --num-gpus NUM_GPUS   Number of GPUs to use for a run. (default: 1)\n"
     ]
    }
   ],
   "source": [
    "!pbrun haplotypecaller -h"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "372f1a89-db78-48fa-991f-6654c331d343",
   "metadata": {},
   "source": [
    "This command will generate `.vcf` file from `.bam` file that we produced in the previous step.\n",
    "For population studies, you may want to generate `.gvcf` files to me merged for joint genotyping in later steps. Otherwise, you can delete the` --gvcf` option and output as `.vcf` file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "bc65cc13-6ad2-4dcd-b3fd-0b284c2d9414",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please visit https://docs.nvidia.com/clara/#parabricks for detailed documentation\n",
      "\n",
      "/usr/local/parabricks/binaries/bin/htvc /home/yingja1227/Ref/Homo_sapiens_assembly38.fasta /home/yingja1227/germline_output/fq2bam_output.bam 1 -o /home/yingja1227/germline_output/variants_gatk.g.vcf -nt 5 --low-memory\n",
      "[PB Info 2025-Sep-03 12:29:13] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Sep-03 12:29:13] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2025-Sep-03 12:29:13] ||                              Version 4.4.0-1                             ||\n",
      "[PB Info 2025-Sep-03 12:29:13] ||                         GPU-GATK4 HaplotypeCaller                        ||\n",
      "[PB Info 2025-Sep-03 12:29:13] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Sep-03 12:29:25] 0 /home/yingja1227/germline_output/fq2bam_output.bam/home/yingja1227/germline_output/variants_gatk.g.vcf\n",
      "[PB Info 2025-Sep-03 12:29:25] ProgressMeter -\tCurrent-Locus\tElapsed-Minutes\tRegions-Processed\tRegions/Minute\n",
      "[PB Info 2025-Sep-03 12:29:35] chr1:10296001\t0.2\t68284\t409704\n",
      "[PB Info 2025-Sep-03 12:29:45] chr1:39878401\t0.3\t191777\t575331\n",
      "[PB Info 2025-Sep-03 12:29:55] chr1:64142401\t0.5\t323143\t646286\n",
      "[PB Info 2025-Sep-03 12:30:05] chr1:125179201\t0.7\t573294\t859941\n",
      "[PB Info 2025-Sep-03 12:30:15] chr1:161356801\t0.8\t658894\t790672\n",
      "[PB Info 2025-Sep-03 12:30:25] chr1:198264001\t1.0\t814408\t814408\n",
      "[PB Info 2025-Sep-03 12:30:35] chr2:8342401\t1.2\t1061821\t910132\n",
      "[PB Info 2025-Sep-03 12:30:45] chr2:45100801\t1.3\t1243899\t932924\n",
      "[PB Info 2025-Sep-03 12:30:55] chr2:95937601\t1.5\t1449927\t966618\n",
      "[PB Info 2025-Sep-03 12:31:05] chr2:144398401\t1.7\t1676230\t1005738\n",
      "[PB Info 2025-Sep-03 12:31:15] chr2:201806401\t1.8\t1912994\t1043451\n",
      "[PB Info 2025-Sep-03 12:31:25] chr3:9244801\t2.0\t2130080\t1065040\n",
      "[PB Info 2025-Sep-03 12:31:35] chr3:52219201\t2.2\t2323699\t1072476\n",
      "[PB Info 2025-Sep-03 12:31:45] chr3:107102401\t2.3\t2539185\t1088222\n",
      "[PB Info 2025-Sep-03 12:31:55] chr3:150288001\t2.5\t2752745\t1101098\n",
      "[PB Info 2025-Sep-03 12:32:05] chr4:124801\t2.7\t2950890\t1106583\n",
      "[PB Info 2025-Sep-03 12:32:15] chr4:49708801\t2.8\t3184395\t1123904\n",
      "[PB Info 2025-Sep-03 12:32:25] chr4:109344001\t3.0\t3429011\t1143003\n",
      "[PB Info 2025-Sep-03 12:32:35] chr4:152563201\t3.2\t3615892\t1141860\n",
      "[PB Info 2025-Sep-03 12:32:45] chr5:20457601\t3.3\t3872156\t1161646\n",
      "[PB Info 2025-Sep-03 12:32:55] chr5:72067201\t3.5\t4075795\t1164512\n",
      "[PB Info 2025-Sep-03 12:33:05] chr5:139420801\t3.7\t4362464\t1189762\n",
      "[PB Info 2025-Sep-03 12:33:15] chr5:177091201\t3.8\t4521485\t1179517\n",
      "[PB Info 2025-Sep-03 12:33:25] chr6:44246401\t4.0\t4733390\t1183347\n",
      "[PB Info 2025-Sep-03 12:33:35] chr6:101400001\t4.2\t4948970\t1187752\n",
      "[PB Info 2025-Sep-03 12:33:45] chr6:157084801\t4.3\t5208852\t1202042\n",
      "[PB Info 2025-Sep-03 12:33:55] chr7:19684801\t4.5\t5357241\t1190498\n",
      "[PB Info 2025-Sep-03 12:34:05] chr7:87460801\t4.7\t5630335\t1206500\n",
      "[PB Info 2025-Sep-03 12:34:15] chr7:116793601\t4.8\t5783706\t1196628\n",
      "[PB Info 2025-Sep-03 12:34:25] chr8:9552001\t5.0\t5966199\t1193239\n",
      "[PB Info 2025-Sep-03 12:34:35] chr8:40056001\t5.2\t6128566\t1186174\n",
      "[PB Info 2025-Sep-03 12:34:45] chr8:62707201\t5.3\t6224700\t1167131\n",
      "[PB Info 2025-Sep-03 12:34:55] chr8:96057601\t5.5\t6368190\t1157852\n",
      "[PB Info 2025-Sep-03 12:35:05] chr9:2088001\t5.7\t6585262\t1162105\n",
      "[PB Info 2025-Sep-03 12:35:15] chr9:30686401\t5.8\t6697091\t1148072\n",
      "[PB Info 2025-Sep-03 12:35:25] chr9:95472001\t6.0\t6904641\t1150773\n",
      "[PB Info 2025-Sep-03 12:35:35] chr9:131227201\t6.2\t7057401\t1144443\n",
      "[PB Info 2025-Sep-03 12:35:45] chr10:41884801\t6.3\t7252324\t1145103\n",
      "[PB Info 2025-Sep-03 12:35:55] chr10:95510401\t6.5\t7507273\t1154965\n",
      "[PB Info 2025-Sep-03 12:36:05] chr10:131952001\t6.7\t7662379\t1149356\n",
      "[PB Info 2025-Sep-03 12:36:15] chr11:33859201\t6.8\t7820245\t1144426\n",
      "[PB Info 2025-Sep-03 12:36:25] chr11:62980801\t7.0\t7945260\t1135037\n",
      "[PB Info 2025-Sep-03 12:36:35] chr11:77337601\t7.2\t8015243\t1118406\n",
      "[PB Info 2025-Sep-03 12:36:45] chr11:103996801\t7.3\t8137917\t1109715\n",
      "[PB Info 2025-Sep-03 12:36:55] chr11:125635201\t7.5\t8212574\t1095009\n",
      "[PB Info 2025-Sep-03 12:37:05] chr12:2644801\t7.7\t8274911\t1079336\n",
      "[PB Info 2025-Sep-03 12:37:15] chr12:18494401\t7.8\t8334801\t1064017\n",
      "[PB Info 2025-Sep-03 12:37:25] chr12:45811201\t8.0\t8456914\t1057114\n",
      "[PB Info 2025-Sep-03 12:37:35] chr12:53976001\t8.2\t8498537\t1040637\n",
      "[PB Info 2025-Sep-03 12:37:45] chr12:65827201\t8.3\t8558315\t1026997\n",
      "[PB Info 2025-Sep-03 12:37:55] chr12:80467201\t8.5\t8623650\t1014547\n",
      "[PB Info 2025-Sep-03 12:38:05] chr12:114672001\t8.7\t8751604\t1009800\n",
      "[PB Info 2025-Sep-03 12:38:15] chr13:18297601\t8.8\t8849062\t1001780\n",
      "[PB Info 2025-Sep-03 12:38:25] chr13:26774401\t9.0\t8888431\t987603\n",
      "[PB Info 2025-Sep-03 12:38:35] chr13:38304001\t9.2\t8947424\t976082\n",
      "[PB Info 2025-Sep-03 12:38:45] chr13:52132801\t9.3\t9019081\t966330\n",
      "[PB Info 2025-Sep-03 12:38:55] chr13:94704001\t9.5\t9185874\t966934\n",
      "[PB Info 2025-Sep-03 12:39:05] chr14:23179201\t9.7\t9303064\t962385\n",
      "[PB Info 2025-Sep-03 12:39:15] chr14:37795201\t9.8\t9373197\t953206\n",
      "[PB Info 2025-Sep-03 12:39:25] chr14:52296001\t10.0\t9435610\t943561\n",
      "[PB Info 2025-Sep-03 12:39:35] chr14:75278401\t10.2\t9530838\t937459\n",
      "[PB Info 2025-Sep-03 12:39:45] chr15:25627201\t10.3\t9666666\t935483\n",
      "[PB Info 2025-Sep-03 12:39:55] chr15:43430401\t10.5\t9746720\t928259\n",
      "[PB Info 2025-Sep-03 12:40:05] chr15:58689601\t10.7\t9800423\t918789\n",
      "[PB Info 2025-Sep-03 12:40:15] chr15:87878401\t10.8\t9921087\t915792\n",
      "[PB Info 2025-Sep-03 12:40:25] chr15:94934401\t11.0\t9972953\t906632\n",
      "[PB Info 2025-Sep-03 12:40:35] chr16:10891201\t11.2\t10050612\t900054\n",
      "[PB Info 2025-Sep-03 12:40:45] chr16:29342401\t11.3\t10130307\t893850\n",
      "[PB Info 2025-Sep-03 12:40:55] chr16:56827201\t11.5\t10209252\t887761\n",
      "[PB Info 2025-Sep-03 12:41:05] chr16:81916801\t11.7\t10295633\t882482\n",
      "[PB Info 2025-Sep-03 12:41:15] chr17:6115201\t11.8\t10376809\t876913\n",
      "[PB Info 2025-Sep-03 12:41:25] chr17:30768001\t12.0\t10464730\t872060\n",
      "[PB Info 2025-Sep-03 12:41:35] chr17:43089601\t12.2\t10515131\t864257\n",
      "[PB Info 2025-Sep-03 12:41:45] chr17:61382401\t12.3\t10592299\t858835\n",
      "[PB Info 2025-Sep-03 12:41:55] chr17:78129601\t12.5\t10666692\t853335\n",
      "[PB Info 2025-Sep-03 12:42:05] chr18:196801\t12.7\t10697367\t844528\n",
      "[PB Info 2025-Sep-03 12:42:15] chr18:14155201\t12.8\t10765509\t838870\n",
      "[PB Info 2025-Sep-03 12:42:25] chr18:37310401\t13.0\t10849004\t834538\n",
      "[PB Info 2025-Sep-03 12:42:35] chr18:61497601\t13.2\t10934577\t830474\n",
      "[PB Info 2025-Sep-03 12:42:45] chr19:4929601\t13.3\t11038468\t827885\n",
      "[PB Info 2025-Sep-03 12:42:55] chr19:10363201\t13.5\t11071828\t820135\n",
      "[PB Info 2025-Sep-03 12:43:05] chr19:33374401\t13.7\t11161158\t816670\n",
      "[PB Info 2025-Sep-03 12:43:15] chr19:38568001\t13.8\t11192085\t809066\n",
      "[PB Info 2025-Sep-03 12:43:25] chr19:49680001\t14.0\t11237520\t802680\n",
      "[PB Info 2025-Sep-03 12:43:35] chr20:6427201\t14.2\t11295564\t797333\n",
      "[PB Info 2025-Sep-03 12:43:45] chr20:31387201\t14.3\t11388376\t794537\n",
      "[PB Info 2025-Sep-03 12:43:55] chr20:41452801\t14.5\t11449401\t789613\n",
      "[PB Info 2025-Sep-03 12:44:05] chr20:55003201\t14.7\t11519400\t785413\n",
      "[PB Info 2025-Sep-03 12:44:15] chr21:10353601\t14.8\t11565790\t779716\n",
      "[PB Info 2025-Sep-03 12:44:25] chr21:29342401\t15.0\t11632393\t775492\n",
      "[PB Info 2025-Sep-03 12:44:35] chr21:41740801\t15.2\t11690956\t770832\n",
      "[PB Info 2025-Sep-03 12:44:45] chr22:19948801\t15.3\t11744611\t765952\n",
      "[PB Info 2025-Sep-03 12:44:55] chr22:28684801\t15.5\t11783399\t760219\n",
      "[PB Info 2025-Sep-03 12:45:05] chr22:34003201\t15.7\t11809599\t753804\n",
      "[PB Info 2025-Sep-03 12:45:15] chr22:41140801\t15.8\t11838654\t747704\n",
      "[PB Info 2025-Sep-03 12:45:25] chr22:50265601\t16.0\t11873156\t742072\n",
      "[PB Info 2025-Sep-03 12:45:35] chrX:35860801\t16.2\t12068543\t746507\n",
      "[PB Info 2025-Sep-03 12:45:45] chrX:72571201\t16.3\t12207338\t747388\n",
      "[PB Info 2025-Sep-03 12:45:55] chrX:111163201\t16.5\t12338668\t747798\n",
      "[PB Info 2025-Sep-03 12:46:05] chrX:139488001\t16.7\t12444143\t746648\n",
      "[PB Info 2025-Sep-03 12:46:15] chrY:12811201\t16.8\t12524150\t744008\n",
      "[PB Info 2025-Sep-03 12:46:25] chrUn_KI270746v1:48001\t17.0\t12582569\t740151\n",
      "[PB Info 2025-Sep-03 12:46:35] HLA-DRB1*10:01:01:9601\t17.2\t12635027\t736020\n",
      "[PB Info 2025-Sep-03 12:46:46] Total time taken: 1040.678141\n",
      "[PB Info 2025-Sep-03 12:46:46] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Sep-03 12:46:46] ||        Program:                         GPU-GATK4 HaplotypeCaller        ||\n",
      "[PB Info 2025-Sep-03 12:46:46] ||        Version:                                           4.4.0-1        ||\n",
      "[PB Info 2025-Sep-03 12:46:46] ||        Start Time:                       Wed Sep  3 12:29:13 2025        ||\n",
      "[PB Info 2025-Sep-03 12:46:46] ||        End Time:                         Wed Sep  3 12:46:46 2025        ||\n",
      "[PB Info 2025-Sep-03 12:46:46] ||        Total Time:                          17 minutes 33 seconds        ||\n",
      "[PB Info 2025-Sep-03 12:46:46] ------------------------------------------------------------------------------\n",
      "Variant caller done, total time: 17.6 min\n"
     ]
    }
   ],
   "source": [
    "!pbrun haplotypecaller \\\n",
    "      --ref Ref/Homo_sapiens_assembly38.fasta \\\n",
    "      --in-bam germline_output/fq2bam_output.bam \\\n",
    "      --out-variants germline_output/variants_gatk.g.vcf \\\n",
    "      --num-gpus 1\\\n",
    "      --htvc-low-memory\n",
    "#    --gvcf   #Use this option if you want to generate .gvcf file"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9992893-7139-4202-9c8e-cae64b39bae4",
   "metadata": {},
   "source": [
    "The key difference between a regular VCF and a GVCF is that the GVCF has records for all sites, whether there is a variant call there or not. The goal is to have every site represented in the file in order to do joint analysis of a cohort in subsequent steps. (https://gatk.broadinstitute.org/hc/en-us/articles/360035531812-GVCF-Genomic-Variant-Call-Format)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "533cf659-bdae-4e2c-ac82-4700792aa5fc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 9.1G\n",
      "-rw-r--r-- 1 yingja1227 ACD113150 4.5G Sep  3 12:29 fq2bam_output.bam\n",
      "-rw-r--r-- 1 yingja1227 ACD113150 3.1M Sep  3 12:29 fq2bam_output.bam.bai\n",
      "-rw-r--r-- 1 yingja1227 ACD113150  86K Sep  3 12:29 fq2bam_output_chrs.txt\n",
      "-rw-r--r-- 1 yingja1227 ACD113150 4.5G Aug 29 13:57 germline_out.bam\n",
      "-rw-r--r-- 1 yingja1227 ACD113150 3.1M Aug 29 13:57 germline_out.bam.bai\n",
      "-rw-r--r-- 1 yingja1227 ACD113150  86K Aug 29 13:57 germline_out_chrs.txt\n",
      "-rw-r--r-- 1 yingja1227 ACD113150 387K Aug 29 13:56 recal.table\n",
      "-rw-r--r-- 1 yingja1227 ACD113150  22M Aug 29 14:10 variants.vcf\n",
      "-rw-r--r-- 1 yingja1227 ACD113150  23M Sep  3 12:46 variants_gatk.g.vcf\n",
      "-rw-r--r-- 1 yingja1227 ACD113150 168K Sep  3 12:46 variants_gatk.g.vcf.idx\n"
     ]
    }
   ],
   "source": [
    "!ls -lh germline_output/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ca6968e-4835-4138-ae7b-b3294867d3a6",
   "metadata": {},
   "source": [
    "#### DeepVariant\n",
    "Beside the Haplotype Caller from GATK, another popular variant caller is DeepVariant, which uses AI methods for variant calling. Parabricks speeeds up DeepVariant and can be used for both short-read and long-read sequencing by choosing the right option."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "774b562b-378c-42da-8579-f645f3a7a84e",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please visit https://docs.nvidia.com/clara/#parabricks for detailed documentation\n",
      "\n",
      "usage: pbrun deepvariant <options>\n",
      "Help: pbrun deepvariant -h\n",
      "\n",
      "Run DeepVariant to convert BAM/CRAM to VCF.\n",
      "\n",
      "optional arguments:\n",
      "  -h, --help            show this help message and exit\n",
      "\n",
      "Input Output file options:\n",
      "  Options for Input and Output files for this tool:\n",
      "\n",
      "  --ref REF             Path to the reference file. (default: None)\n",
      "  --in-bam IN_BAM       Path to the input BAM/CRAM file for variant calling.\n",
      "                        (default: None)\n",
      "  --interval-file INTERVAL_FILE\n",
      "                        Path to a BED file (.bed) for selective access. This\n",
      "                        option can be used multiple times. (default: None)\n",
      "  --out-variants OUT_VARIANTS\n",
      "                        Path of the vcf/g.vcf/g.vcf.gz file after variant\n",
      "                        calling. (default: None)\n",
      "  --pb-model-file PB_MODEL_FILE\n",
      "                        Path to a non-default parabricks model file for\n",
      "                        deepvariant. (default: None)\n",
      "  --pb-model-dir PB_MODEL_DIR\n",
      "                        Path to a non-default parabricks model dir that\n",
      "                        contains multiple engine files for one model (default:\n",
      "                        None)\n",
      "  --proposed-variants PROPOSED_VARIANTS\n",
      "                        Path of the vcf.gz file, which has proposed variants\n",
      "                        for the make examples stage. (default: None)\n",
      "\n",
      "Tool Options:\n",
      "  Options specific to the tool\n",
      "\n",
      "  --disable-use-window-selector-model\n",
      "                        Change the window selector model from Allele Count\n",
      "                        Linear to Variant Reads. This option will increase the\n",
      "                        accuracy and runtime. (default: None)\n",
      "  --gvcf                Generate variant calls in .gvcf Format. (default:\n",
      "                        None)\n",
      "  --norealign-reads     Do not locally realign reads before calling variants.\n",
      "                        Reads longer than 500 bp are never realigned.\n",
      "                        (default: None)\n",
      "  --sort-by-haplotypes  Reads are sorted by haplotypes (using HP tag).\n",
      "                        (default: None)\n",
      "  --keep-duplicates     Keep reads that are duplicate. (default: None)\n",
      "  --vsc-min-count-snps VSC_MIN_COUNT_SNPS\n",
      "                        SNP alleles occurring at least this many times in the\n",
      "                        AlleleCount will be advanced as candidates. (default:\n",
      "                        2)\n",
      "  --vsc-min-count-indels VSC_MIN_COUNT_INDELS\n",
      "                        Indel alleles occurring at least this many times in\n",
      "                        the AlleleCount will be advanced as candidates.\n",
      "                        (default: 2)\n",
      "  --vsc-min-fraction-snps VSC_MIN_FRACTION_SNPS\n",
      "                        SNP alleles occurring at least this fraction of all\n",
      "                        counts in the AlleleCount will be advanced as\n",
      "                        candidates. (default: 0.12)\n",
      "  --vsc-min-fraction-indels VSC_MIN_FRACTION_INDELS\n",
      "                        Indel alleles occurring at least this fraction of all\n",
      "                        counts in the AlleleCount will be advanced as\n",
      "                        candidates. (default: None)\n",
      "  --min-mapping-quality MIN_MAPPING_QUALITY\n",
      "                        By default, reads with any mapping quality are kept.\n",
      "                        Setting this field to a positive integer i will only\n",
      "                        keep reads that have a MAPQ >= i. Note this only\n",
      "                        applies to aligned reads. (default: 5)\n",
      "  --min-base-quality MIN_BASE_QUALITY\n",
      "                        Minimum base quality. This option enforces a minimum\n",
      "                        base quality score for alternate alleles. Alternate\n",
      "                        alleles will only be considered if all bases in the\n",
      "                        allele have a quality greater than min_base_quality.\n",
      "                        (default: 10)\n",
      "  --mode MODE           Value can be one of [shortread, pacbio, ont]. By\n",
      "                        default, it is shortread. If mode is set to pacbio,\n",
      "                        the following defaults are used: --norealign-reads,\n",
      "                        --alt-aligned-pileup diff_channels, --vsc-min-\n",
      "                        fraction-indels 0.12. If mode is set to ont, the\n",
      "                        following defaults are used: -norealign-reads,\n",
      "                        --variant-caller VCF_CANDIDATE_IMPORTER. (default:\n",
      "                        shortread)\n",
      "  --alt-aligned-pileup ALT_ALIGNED_PILEUP\n",
      "                        Value can be one of [none, diff_channels]. Include\n",
      "                        alignments of reads against each candidate alternate\n",
      "                        allele in the pileup image. (default: None)\n",
      "  --variant-caller VARIANT_CALLER\n",
      "                        Value can be one of [VERY_SENSITIVE_CALLER,\n",
      "                        VCF_CANDIDATE_IMPORTER]. The caller to use to make\n",
      "                        examples. If you use VCF_CANDIDATE_IMPORTER, it\n",
      "                        implies force calling. Default is\n",
      "                        VERY_SENSITIVE_CALLER. (default: None)\n",
      "  --add-hp-channel      Add another channel to represent HP tags per read.\n",
      "                        (default: None)\n",
      "  --parse-sam-aux-fields\n",
      "                        Auxiliary fields of the BAM/CRAM records are parsed.\n",
      "                        If either --sort-by-haplotypes or --add-hp-channel is\n",
      "                        set, then this option must also be set. (default:\n",
      "                        None)\n",
      "  --use-wes-model       If passed, the WES model file will be used. Only used\n",
      "                        in shortread mode. (default: None)\n",
      "  --run-partition       Divide the whole genome into multiple partitions and\n",
      "                        run multiple processes at the same time, each on one\n",
      "                        partition. (default: None)\n",
      "  --include-med-dp      If True, include MED_DP in the output gVCF records.\n",
      "                        (default: None)\n",
      "  --normalize-reads     If True, allele counter left align INDELs for each\n",
      "                        read. (default: None)\n",
      "  --channel-insert-size\n",
      "                        If True, add insert_size channel into pileup image. By\n",
      "                        default, this parameter is true in WGS and WES mode.\n",
      "                        (default: None)\n",
      "  --no-channel-insert-size\n",
      "                        If True, don't add insert_size channel into the pileup\n",
      "                        image. (default: None)\n",
      "  --max-read-size-512   Allow deepvariant to run on reads of size 512bp. The\n",
      "                        default size is 320 bp. (default: None)\n",
      "  --prealign-helper-thread\n",
      "                        Use an extra thread for the pre-align step. This\n",
      "                        parameter is more useful when --max-reads-size-512 is\n",
      "                        set. (default: None)\n",
      "  --max-reads-per-partition MAX_READS_PER_PARTITION\n",
      "                        The maximum number of reads per partition that are\n",
      "                        considered before following processing such as\n",
      "                        sampling and realignment. (default: 1500)\n",
      "  --partition-size PARTITION_SIZE\n",
      "                        The maximum number of basepairs allowed in a region\n",
      "                        before splitting it into multiple smaller subregions.\n",
      "                        (default: 1000)\n",
      "  --track-ref-reads     If True, allele counter keeps track of reads\n",
      "                        supporting ref. By default, allele counter keeps a\n",
      "                        simple count of the number of reads supporting ref.\n",
      "                        (default: None)\n",
      "  --phase-reads         Calculate phases and add HP tag to all reads\n",
      "                        automatically. (default: None)\n",
      "  --dbg-min-base-quality DBG_MIN_BASE_QUALITY\n",
      "                        Minimum base quality in a k-mer sequence to consider.\n",
      "                        (default: 15)\n",
      "  --ws-min-windows-distance WS_MIN_WINDOWS_DISTANCE\n",
      "                        Minimum distance between candidate windows for local\n",
      "                        assembly (default: 80)\n",
      "  --channel-gc-content  If True, add gc_content channel into pileup image\n",
      "                        (default: None)\n",
      "  --channel-hmer-deletion-quality\n",
      "                        If True, add hmer deletion quality channel into pileup\n",
      "                        image (default: None)\n",
      "  --channel-hmer-insertion-quality\n",
      "                        If True, add hmer insertion quality channel into\n",
      "                        pileup image (default: None)\n",
      "  --channel-non-hmer-insertion-quality\n",
      "                        If True, add non-hmer insertion quality channel into\n",
      "                        pileup image (default: None)\n",
      "  --skip-bq-channel     If True, ignore base quality channel. (default: None)\n",
      "  --aux-fields-to-keep AUX_FIELDS_TO_KEEP\n",
      "                        Comma-delimited list of auxiliary BAM fields to keep.\n",
      "                        Values can be [HP, tp, t0] (default: HP)\n",
      "  --vsc-min-fraction-hmer-indels VSC_MIN_FRACTION_HMER_INDELS\n",
      "                        Hmer Indel alleles occurring at least this be advanced\n",
      "                        as candidates. Use this threshold if hmer and non-hmer\n",
      "                        indels should be treated differently (Ultima\n",
      "                        reads)Default will use the same threshold for hmer and\n",
      "                        non-hmer indels, as defined in\n",
      "                        vsc_min_fraction_indels. (default: None)\n",
      "  --vsc-turn-on-non-hmer-ins-proxy-support\n",
      "                        Add read-support from soft-clipped reads and other\n",
      "                        non-hmer insertion alleles,to the most frequent non-\n",
      "                        hmer insertion allele. (default: None)\n",
      "  --consider-strand-bias\n",
      "                        If True, expect SB field in calls and write it to vcf\n",
      "                        (default: None)\n",
      "  --p-error P_ERROR     Basecalling error for reference confidence model.\n",
      "                        (default: 0.001)\n",
      "  --channel-ins-size    If true, add another channel to represent size of\n",
      "                        insertions. (good for flow-based sequencing) (default:\n",
      "                        None)\n",
      "  --max-ins-size MAX_INS_SIZE\n",
      "                        Max insertion size for ins_size_channel, larger\n",
      "                        insertions will look like max (have max intensity)\n",
      "                        (default: 10)\n",
      "  --disable-group-variants\n",
      "                        If using vcf_candidate_importer and multi-allelic\n",
      "                        sites are split across multiple lines in VCF, set to\n",
      "                        True so that variants are not grouped when\n",
      "                        transforming CallVariantsOutput to Variants. (default:\n",
      "                        None)\n",
      "  -L INTERVAL, --interval INTERVAL\n",
      "                        Interval within which to call the variants from the\n",
      "                        BAM/CRAM file. Overlapping intervals will be combined.\n",
      "                        Interval files should be passed using the --interval-\n",
      "                        file option. This option can be used multiple times\n",
      "                        (e.g. \"-L chr1 -L chr2:10000 -L chr3:20000+ -L\n",
      "                        chr4:10000-20000\"). (default: None)\n",
      "\n",
      "Run Options:\n",
      "  Options related to resource/verbosity/etc\n",
      "\n",
      "  --logfile LOGFILE     Path to the log file. If not specified, messages will\n",
      "                        only be written to the standard error output.\n",
      "                        (default: None)\n",
      "  --tmp-dir TMP_DIR     Full path to the directory where temporary files will\n",
      "                        be stored. (default: .)\n",
      "  --with-petagene-dir WITH_PETAGENE_DIR\n",
      "                        Full path to the PetaGene installation directory. By\n",
      "                        default, this should have been installed at\n",
      "                        /opt/petagene. Use of this option also requires that\n",
      "                        the PetaLink library has been preloaded by setting the\n",
      "                        LD_PRELOAD environment variable. Optionally set the\n",
      "                        PETASUITE_REFPATH and PGCLOUD_CREDPATH environment\n",
      "                        variables that are used for data and credentials.\n",
      "                        Optionally set the PetaLinkMode environment variable\n",
      "                        that is used to further configure PetaLink, notably\n",
      "                        setting it to \"+write\" to enable outputting compressed\n",
      "                        BAM and .fastq files. (default: None)\n",
      "  --keep-tmp            Do not delete the directory storing temporary files\n",
      "                        after completion. (default: None)\n",
      "  --no-seccomp-override\n",
      "                        Do not override seccomp options for docker. (default:\n",
      "                        None)\n",
      "  --version             View compatible software versions. (default: None)\n",
      "  --num-gpus NUM_GPUS   Number of GPUs to use for a run. (default: 1)\n"
     ]
    }
   ],
   "source": [
    "!pbrun deepvariant -h"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d947725a-7a36-41cf-a1b2-feb33fcf8d3b",
   "metadata": {},
   "source": [
    "Default is to run variant calling on short reads like Illumina."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "fedb93a4-8ca8-4afa-9069-938a2052d457",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please visit https://docs.nvidia.com/clara/#parabricks for detailed documentation\n",
      "\n",
      "Detected 1 CUDA Capable device(s), considering 1 device(s)\n",
      "  CUDA Driver Version / Runtime Version          12.2 / 12.0\n",
      "Using model for CUDA Capability Major/Minor version number:    89\n",
      "CUDA_VISIBLE_DEVICES=0 /usr/local/parabricks/binaries//bin/deepvariant /workdir/sample_data/Ref/Homo_sapiens_assembly38.fasta /workdir/outputdir/fq2bam_output.bam 1 2 -o 0.vcf  -L \"chr1\" -L \"chr2\" -L \"chr3\" -L \"chr4\" -L \"chr5\" -L \"chr6\" -L \"chr7\" -L \"chrX\" -L \"chr8\" -L \"chr9\" -L \"chr11\" -L \"chr10\" -L \"chr12\" -L \"chr13\" -L \"chr14\" -L \"chr15\" -L \"chr16\" -L \"chr17\" -L \"chr18\" -L \"chr20\" -L \"chr19\" -L \"chrY\" -L \"chr22\" -L \"chr21\" -L \"chrM\" -n 6 --model /usr/local/parabricks/binaries//model/80+/shortread/deepvariant.eng --channel_insert_size --pileup_image_width 221 --max_reads_per_partition 1500 --partition_size 1000 --vsc_min_count_snps 2 --vsc_min_count_indels 2 --vsc_min_fraction_snps 0.12 --min_mapping_quality 5 --min_base_quality 10 --alt_aligned_pileup none --variant_caller VERY_SENSITIVE_CALLER --dbg_min_base_quality 15 --ws_min_windows_distance 80 --aux_fields_to_keep HP --p_error 0.001 --max_ins_size 10   \n",
      "[PB Info 2023-Nov-30 03:22:57] ------------------------------------------------------------------------------\n",
      "[PB Info 2023-Nov-30 03:22:57] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2023-Nov-30 03:22:57] ||                              Version 4.1.1-1                             ||\n",
      "[PB Info 2023-Nov-30 03:22:57] ||                                deepvariant                               ||\n",
      "[PB Info 2023-Nov-30 03:22:57] ------------------------------------------------------------------------------\n",
      "[PB Info 2023-Nov-30 03:22:57] Starting DeepVariant\n",
      "[PB Info 2023-Nov-30 03:22:57] Running with 1 GPU devices, each with 2 group instances and 6 workers\n",
      "[PB Info 2023-Nov-30 03:23:15] ProgressMeter -\tCurrent-Locus\tElapsed-Minutes\n",
      "[PB Info 2023-Nov-30 03:23:21] ProgressMeter -\tchr1:84514000\t0.1\n",
      "[PB Info 2023-Nov-30 03:23:27] ProgressMeter -\tchr1:221012000\t0.2\n",
      "[PB Info 2023-Nov-30 03:23:33] ProgressMeter -\tchr2:98381000\t0.3\n",
      "[PB Info 2023-Nov-30 03:23:39] ProgressMeter -\tchr2:233769000\t0.4\n",
      "[PB Info 2023-Nov-30 03:23:45] ProgressMeter -\tchr3:101383000\t0.5\n",
      "[PB Info 2023-Nov-30 03:23:51] ProgressMeter -\tchr4:35893000\t0.6\n",
      "[PB Info 2023-Nov-30 03:23:57] ProgressMeter -\tchr4:161946000\t0.7\n",
      "[PB Info 2023-Nov-30 03:24:03] ProgressMeter -\tchr5:124932000\t0.8\n",
      "[PB Info 2023-Nov-30 03:24:09] ProgressMeter -\tchr6:78953000\t0.9\n",
      "[PB Info 2023-Nov-30 03:24:15] ProgressMeter -\tchr7:427000\t1.0\n",
      "[PB Info 2023-Nov-30 03:24:21] ProgressMeter -\tchr7:114886000\t1.1\n",
      "[PB Info 2023-Nov-30 03:24:27] ProgressMeter -\tchr8:56103000\t1.2\n",
      "[PB Info 2023-Nov-30 03:24:33] ProgressMeter -\tchr9:601000\t1.3\n",
      "[PB Info 2023-Nov-30 03:24:39] ProgressMeter -\tchr9:133883000\t1.4\n",
      "[PB Info 2023-Nov-30 03:24:45] ProgressMeter -\tchr10:100392000\t1.5\n",
      "[PB Info 2023-Nov-30 03:24:51] ProgressMeter -\tchr11:70842000\t1.6\n",
      "[PB Info 2023-Nov-30 03:24:57] ProgressMeter -\tchr12:34362000\t1.7\n",
      "[PB Info 2023-Nov-30 03:25:03] ProgressMeter -\tchr12:118507000\t1.8\n",
      "[PB Info 2023-Nov-30 03:25:09] ProgressMeter -\tchr14:20508000\t1.9\n",
      "[PB Info 2023-Nov-30 03:25:15] ProgressMeter -\tchr15:18952000\t2.0\n",
      "[PB Info 2023-Nov-30 03:25:21] ProgressMeter -\tchr16:54020000\t2.1\n",
      "[PB Info 2023-Nov-30 03:25:27] ProgressMeter -\tchr17:21727000\t2.2\n",
      "[PB Info 2023-Nov-30 03:25:33] ProgressMeter -\tchr18:8232000\t2.3\n",
      "[PB Info 2023-Nov-30 03:25:39] ProgressMeter -\tchr20:1511000\t2.4\n",
      "[PB Info 2023-Nov-30 03:25:45] ProgressMeter -\tchr22:21339000\t2.5\n",
      "[PB Info 2023-Nov-30 03:25:51] ProgressMeter -\tchrX:55143000\t2.6\n",
      "[PB Info 2023-Nov-30 03:25:57] ProgressMeter -\tchrM:4000\t2.7\n",
      "[PB Info 2023-Nov-30 03:26:04] Deepvariant is finished, total time is 168.718 seconds\n",
      "[PB Info 2023-Nov-30 03:26:04] ------------------------------------------------------------------------------\n",
      "[PB Info 2023-Nov-30 03:26:04] ||        Program:                                       deepvariant        ||\n",
      "[PB Info 2023-Nov-30 03:26:04] ||        Version:                                           4.1.1-1        ||\n",
      "[PB Info 2023-Nov-30 03:26:04] ||        Start Time:                       Thu Nov 30 03:22:57 2023        ||\n",
      "[PB Info 2023-Nov-30 03:26:04] ||        End Time:                         Thu Nov 30 03:26:04 2023        ||\n",
      "[PB Info 2023-Nov-30 03:26:04] ||        Total Time:                            3 minutes 7 seconds        ||\n",
      "[PB Info 2023-Nov-30 03:26:04] ------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "!pbrun deepvariant \\\n",
    "    --ref Ref/Homo_sapiens_assembly38.fasta \\\n",
    "    --in-bam germline_output/fq2bam_output.bam \\\n",
    "    --out-variants germline_output/variants_dv.vcf \\\n",
    "    --num-streams-per-gpu 2 \\\n",
    "    --run-partition \\\n",
    "    --gpu-num-per-partition 1 \\\n",
    "    --num-gpus 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "232ecee4-d01c-4e22-a39f-0b26202ef336",
   "metadata": {},
   "source": [
    "DeepVariant from Parabricks has the ability to use multiple streams on a GPU. The number of streams that can be used depends on the available resources. The default number of streams is set to two but can be increased up to a maximum of six to get better performance. This is something that has to be experimented with, before getting the optimal number on your system. (https://docs.nvidia.com/clara/parabricks/4.1.0/bestperformance.html#best-performance-for-deepvariant)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "d459010a-4946-4e70-b144-e637299028ad",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please visit https://docs.nvidia.com/clara/#parabricks for detailed documentation\n",
      "\n",
      "Detected 1 CUDA Capable device(s), considering 1 device(s)\n",
      "  CUDA Driver Version / Runtime Version          12.2 / 12.0\n",
      "Using model for CUDA Capability Major/Minor version number:    89\n",
      "CUDA_VISIBLE_DEVICES=0 /usr/local/parabricks/binaries//bin/deepvariant /workdir/sample_data/Ref/Homo_sapiens_assembly38.fasta /workdir/outputdir/fq2bam_output.bam 1 4 -o 0.vcf  -L \"chr1\" -L \"chr2\" -L \"chr3\" -L \"chr4\" -L \"chr5\" -L \"chr6\" -L \"chr7\" -L \"chrX\" -L \"chr8\" -L \"chr9\" -L \"chr11\" -L \"chr10\" -L \"chr12\" -L \"chr13\" -L \"chr14\" -L \"chr15\" -L \"chr16\" -L \"chr17\" -L \"chr18\" -L \"chr20\" -L \"chr19\" -L \"chrY\" -L \"chr22\" -L \"chr21\" -L \"chrM\" -n 6 --model /usr/local/parabricks/binaries//model/80+/shortread/deepvariant.eng --channel_insert_size --pileup_image_width 221 --max_reads_per_partition 1500 --partition_size 1000 --vsc_min_count_snps 2 --vsc_min_count_indels 2 --vsc_min_fraction_snps 0.12 --min_mapping_quality 5 --min_base_quality 10 --alt_aligned_pileup none --variant_caller VERY_SENSITIVE_CALLER --dbg_min_base_quality 15 --ws_min_windows_distance 80 --aux_fields_to_keep HP --p_error 0.001 --max_ins_size 10   \n",
      "[PB Info 2023-Nov-30 03:34:50] ------------------------------------------------------------------------------\n",
      "[PB Info 2023-Nov-30 03:34:50] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2023-Nov-30 03:34:50] ||                              Version 4.1.1-1                             ||\n",
      "[PB Info 2023-Nov-30 03:34:50] ||                                deepvariant                               ||\n",
      "[PB Info 2023-Nov-30 03:34:50] ------------------------------------------------------------------------------\n",
      "[PB Info 2023-Nov-30 03:34:50] Starting DeepVariant\n",
      "[PB Info 2023-Nov-30 03:34:50] Running with 1 GPU devices, each with 4 group instances and 6 workers\n",
      "[PB Info 2023-Nov-30 03:35:10] ProgressMeter -\tCurrent-Locus\tElapsed-Minutes\n",
      "[PB Info 2023-Nov-30 03:35:16] ProgressMeter -\t:0\t0.1\n",
      "[PB Info 2023-Nov-30 03:35:22] ProgressMeter -\tchr2:138283000\t0.2\n",
      "[PB Info 2023-Nov-30 03:35:28] ProgressMeter -\tchr3:103721000\t0.3\n",
      "[PB Info 2023-Nov-30 03:35:34] ProgressMeter -\tchr4:144007000\t0.4\n",
      "[PB Info 2023-Nov-30 03:35:40] ProgressMeter -\tchr5:158697000\t0.5\n",
      "[PB Info 2023-Nov-30 03:35:46] ProgressMeter -\tchr7:74992000\t0.6\n",
      "[PB Info 2023-Nov-30 03:35:52] ProgressMeter -\tchr8:113130000\t0.7\n",
      "[PB Info 2023-Nov-30 03:35:58] ProgressMeter -\tchr10:7808000\t0.8\n",
      "[PB Info 2023-Nov-30 03:36:04] ProgressMeter -\tchr11:62775000\t0.9\n",
      "[PB Info 2023-Nov-30 03:36:10] ProgressMeter -\tchr12:78455000\t1.0\n",
      "[PB Info 2023-Nov-30 03:36:16] ProgressMeter -\tchr14:103141000\t1.1\n",
      "[PB Info 2023-Nov-30 03:36:22] ProgressMeter -\tchr17:5978000\t1.2\n",
      "[PB Info 2023-Nov-30 03:36:28] ProgressMeter -\tchr18:53913000\t1.3\n",
      "[PB Info 2023-Nov-30 03:36:34] ProgressMeter -\tchr22:19606000\t1.4\n",
      "[PB Info 2023-Nov-30 03:36:41] Deepvariant is finished, total time is 90.839 seconds\n",
      "[PB Info 2023-Nov-30 03:36:41] ------------------------------------------------------------------------------\n",
      "[PB Info 2023-Nov-30 03:36:41] ||        Program:                                       deepvariant        ||\n",
      "[PB Info 2023-Nov-30 03:36:41] ||        Version:                                           4.1.1-1        ||\n",
      "[PB Info 2023-Nov-30 03:36:41] ||        Start Time:                       Thu Nov 30 03:34:50 2023        ||\n",
      "[PB Info 2023-Nov-30 03:36:41] ||        End Time:                         Thu Nov 30 03:36:41 2023        ||\n",
      "[PB Info 2023-Nov-30 03:36:41] ||        Total Time:                            1 minute 51 seconds        ||\n",
      "[PB Info 2023-Nov-30 03:36:41] ------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "!pbrun deepvariant \\\n",
    "    --ref /workdir/sample_data/Ref/Homo_sapiens_assembly38.fasta \\\n",
    "    --in-bam /workdir/outputdir/fq2bam_output.bam \\\n",
    "    --out-variants /workdir/outputdir/variants_dv.vcf \\\n",
    "    --num-streams-per-gpu 4 \\\n",
    "    --run-partition \\\n",
    "    --gpu-num-per-partition 1 \\\n",
    "    --num-gpus 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a85f500f-adbc-43ef-b395-fa4a497e66a1",
   "metadata": {},
   "source": [
    "Using the --run-partition, --proposed-variants, and --gvcf options at the same time will lead to a substantial slowdown."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9f60d6c-985c-4d78-ae85-d735d5c1cc46",
   "metadata": {},
   "source": [
    "- gvcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "fc8ae15a-562b-48b3-908c-5352ab87c9b7",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please visit https://docs.nvidia.com/clara/#parabricks for detailed documentation\n",
      "\n",
      "Detected 1 CUDA Capable device(s), considering 1 device(s)\n",
      "  CUDA Driver Version / Runtime Version          12.3 / 12.3\n",
      "Using model for CUDA Capability Major/Minor version number:    70\n",
      "/usr/local/parabricks/binaries/bin/deepvariant /tutorial/sample_data/Ref/Homo_sapiens_assembly38.fasta /tutorial/outputdir/germline_out.bam 1 1 -o /tutorial/outputdir/variants_dv.g.vcf -n 6 --model /usr/local/parabricks/binaries/model/70/shortread/deepvariant.eng -g --channel_insert_size --pileup_image_width 221 --max_reads_per_partition 1500 --partition_size 1000 --vsc_min_count_snps 2 --vsc_min_count_indels 2 --vsc_min_fraction_snps 0.12 --min_mapping_quality 5 --min_base_quality 10 --alt_aligned_pileup none --variant_caller VERY_SENSITIVE_CALLER --dbg_min_base_quality 15 --ws_min_windows_distance 80 --aux_fields_to_keep HP --p_error 0.001 --max_ins_size 10\n",
      "[PB Info 2024-Oct-11 03:53:55] ------------------------------------------------------------------------------\n",
      "[PB Info 2024-Oct-11 03:53:55] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2024-Oct-11 03:53:55] ||                              Version 4.3.2-1                             ||\n",
      "[PB Info 2024-Oct-11 03:53:55] ||                                deepvariant                               ||\n",
      "[PB Info 2024-Oct-11 03:53:55] ------------------------------------------------------------------------------\n",
      "[PB Info 2024-Oct-11 03:53:55] Starting DeepVariant\n",
      "[PB Info 2024-Oct-11 03:53:55] Running with 1 GPU devices, each with 1 group instances and 6 workers\n",
      "[PB Info 2024-Oct-11 03:54:10] ProgressMeter -\tCurrent-Locus\tElapsed-Minutes\n",
      "[PB Info 2024-Oct-11 03:54:16] ProgressMeter -\tchr1:5594000\t0.1\n",
      "[PB Info 2024-Oct-11 03:54:22] ProgressMeter -\tchr1:17597000\t0.2\n",
      "[PB Info 2024-Oct-11 03:54:28] ProgressMeter -\tchr1:34373000\t0.3\n",
      "[PB Info 2024-Oct-11 03:54:34] ProgressMeter -\tchr1:43585000\t0.4\n",
      "[PB Info 2024-Oct-11 03:54:40] ProgressMeter -\tchr1:57858000\t0.5\n",
      "[PB Info 2024-Oct-11 03:54:46] ProgressMeter -\tchr1:79296000\t0.6\n",
      "[PB Info 2024-Oct-11 03:54:52] ProgressMeter -\tchr1:95542000\t0.7\n",
      "[PB Info 2024-Oct-11 03:54:58] ProgressMeter -\tchr1:122929000\t0.8\n",
      "[PB Info 2024-Oct-11 03:55:04] ProgressMeter -\tchr1:152394000\t0.9\n",
      "[PB Info 2024-Oct-11 03:55:10] ProgressMeter -\tchr1:166634000\t1.0\n",
      "[PB Info 2024-Oct-11 03:55:16] ProgressMeter -\tchr1:187231000\t1.1\n",
      "[PB Info 2024-Oct-11 03:55:22] ProgressMeter -\tchr1:202299000\t1.2\n",
      "[PB Info 2024-Oct-11 03:55:28] ProgressMeter -\tchr1:211480000\t1.3\n",
      "[PB Info 2024-Oct-11 03:55:34] ProgressMeter -\tchr1:235405000\t1.4\n",
      "[PB Info 2024-Oct-11 03:55:40] ProgressMeter -\tchr2:12536000\t1.5\n",
      "[PB Info 2024-Oct-11 03:55:46] ProgressMeter -\tchr2:23725000\t1.6\n",
      "[PB Info 2024-Oct-11 03:55:52] ProgressMeter -\tchr2:34267000\t1.7\n",
      "[PB Info 2024-Oct-11 03:55:58] ProgressMeter -\tchr2:48581000\t1.8\n",
      "[PB Info 2024-Oct-11 03:56:04] ProgressMeter -\tchr2:70298000\t1.9\n",
      "[PB Info 2024-Oct-11 03:56:10] ProgressMeter -\tchr2:88565000\t2.0\n",
      "[PB Info 2024-Oct-11 03:56:16] ProgressMeter -\tchr2:110938000\t2.1\n",
      "[PB Info 2024-Oct-11 03:56:22] ProgressMeter -\tchr2:126439000\t2.2\n",
      "[PB Info 2024-Oct-11 03:56:28] ProgressMeter -\tchr2:143486000\t2.3\n",
      "[PB Info 2024-Oct-11 03:56:34] ProgressMeter -\tchr2:156425000\t2.4\n",
      "[PB Info 2024-Oct-11 03:56:40] ProgressMeter -\tchr2:171996000\t2.5\n",
      "[PB Info 2024-Oct-11 03:56:46] ProgressMeter -\tchr2:179208000\t2.6\n",
      "[PB Info 2024-Oct-11 03:56:52] ProgressMeter -\tchr2:195891000\t2.7\n",
      "[PB Info 2024-Oct-11 03:56:58] ProgressMeter -\tchr2:208240000\t2.8\n",
      "[PB Info 2024-Oct-11 03:57:04] ProgressMeter -\tchr2:217700000\t2.9\n",
      "[PB Info 2024-Oct-11 03:57:10] ProgressMeter -\tchr2:235929000\t3.0\n",
      "[PB Info 2024-Oct-11 03:57:16] ProgressMeter -\tchr3:11738000\t3.1\n",
      "[PB Info 2024-Oct-11 03:57:22] ProgressMeter -\tchr3:21319000\t3.2\n",
      "[PB Info 2024-Oct-11 03:57:28] ProgressMeter -\tchr3:33062000\t3.3\n",
      "[PB Info 2024-Oct-11 03:57:34] ProgressMeter -\tchr3:50154000\t3.4\n",
      "[PB Info 2024-Oct-11 03:57:40] ProgressMeter -\tchr3:70981000\t3.5\n",
      "[PB Info 2024-Oct-11 03:57:46] ProgressMeter -\tchr3:89518000\t3.6\n",
      "[PB Info 2024-Oct-11 03:57:52] ProgressMeter -\tchr3:106843000\t3.7\n",
      "[PB Info 2024-Oct-11 03:57:58] ProgressMeter -\tchr3:120361000\t3.8\n",
      "[PB Info 2024-Oct-11 03:58:04] ProgressMeter -\tchr3:123683000\t3.9\n",
      "[PB Info 2024-Oct-11 03:58:10] ProgressMeter -\tchr3:138135000\t4.0\n",
      "[PB Info 2024-Oct-11 03:58:16] ProgressMeter -\tchr3:147850000\t4.1\n",
      "[PB Info 2024-Oct-11 03:58:22] ProgressMeter -\tchr3:170285000\t4.2\n",
      "[PB Info 2024-Oct-11 03:58:28] ProgressMeter -\tchr3:185888000\t4.3\n",
      "[PB Info 2024-Oct-11 03:58:34] ProgressMeter -\tchr4:12322000\t4.4\n",
      "[PB Info 2024-Oct-11 03:58:40] ProgressMeter -\tchr4:33849000\t4.5\n",
      "[PB Info 2024-Oct-11 03:58:46] ProgressMeter -\tchr4:50275000\t4.6\n",
      "[PB Info 2024-Oct-11 03:58:52] ProgressMeter -\tchr4:65341000\t4.7\n",
      "[PB Info 2024-Oct-11 03:58:58] ProgressMeter -\tchr4:84023000\t4.8\n",
      "[PB Info 2024-Oct-11 03:59:04] ProgressMeter -\tchr4:102985000\t4.9\n",
      "[PB Info 2024-Oct-11 03:59:10] ProgressMeter -\tchr4:113362000\t5.0\n",
      "[PB Info 2024-Oct-11 03:59:16] ProgressMeter -\tchr4:134551000\t5.1\n",
      "[PB Info 2024-Oct-11 03:59:22] ProgressMeter -\tchr4:145921000\t5.2\n",
      "[PB Info 2024-Oct-11 03:59:28] ProgressMeter -\tchr4:163202000\t5.3\n",
      "[PB Info 2024-Oct-11 03:59:34] ProgressMeter -\tchr4:181996000\t5.4\n",
      "[PB Info 2024-Oct-11 03:59:40] ProgressMeter -\tchr5:9467000\t5.5\n",
      "[PB Info 2024-Oct-11 03:59:46] ProgressMeter -\tchr5:20779000\t5.6\n",
      "[PB Info 2024-Oct-11 03:59:52] ProgressMeter -\tchr5:36146000\t5.7\n",
      "[PB Info 2024-Oct-11 03:59:58] ProgressMeter -\tchr5:52980000\t5.8\n",
      "[PB Info 2024-Oct-11 04:00:04] ProgressMeter -\tchr5:64489000\t5.9\n",
      "[PB Info 2024-Oct-11 04:00:10] ProgressMeter -\tchr5:85302000\t6.0\n",
      "[PB Info 2024-Oct-11 04:00:16] ProgressMeter -\tchr5:103664000\t6.1\n",
      "[PB Info 2024-Oct-11 04:00:22] ProgressMeter -\tchr5:119226000\t6.2\n",
      "[PB Info 2024-Oct-11 04:00:28] ProgressMeter -\tchr5:134392000\t6.3\n",
      "[PB Info 2024-Oct-11 04:00:34] ProgressMeter -\tchr5:146544000\t6.4\n",
      "[PB Info 2024-Oct-11 04:00:40] ProgressMeter -\tchr5:156469000\t6.5\n",
      "[PB Info 2024-Oct-11 04:00:46] ProgressMeter -\tchr5:166156000\t6.6\n",
      "[PB Info 2024-Oct-11 04:00:52] ProgressMeter -\tchr6:386000\t6.7\n",
      "[PB Info 2024-Oct-11 04:00:58] ProgressMeter -\tchr6:17504000\t6.8\n",
      "[PB Info 2024-Oct-11 04:01:04] ProgressMeter -\tchr6:27773000\t6.9\n",
      "[PB Info 2024-Oct-11 04:01:10] ProgressMeter -\tchr6:44170000\t7.0\n",
      "[PB Info 2024-Oct-11 04:01:16] ProgressMeter -\tchr6:70360000\t7.1\n",
      "[PB Info 2024-Oct-11 04:01:22] ProgressMeter -\tchr6:93221000\t7.2\n",
      "[PB Info 2024-Oct-11 04:01:28] ProgressMeter -\tchr6:107333000\t7.3\n",
      "[PB Info 2024-Oct-11 04:01:34] ProgressMeter -\tchr6:119497000\t7.4\n",
      "[PB Info 2024-Oct-11 04:01:40] ProgressMeter -\tchr6:135220000\t7.5\n",
      "[PB Info 2024-Oct-11 04:01:46] ProgressMeter -\tchr6:142450000\t7.6\n",
      "[PB Info 2024-Oct-11 04:01:52] ProgressMeter -\tchr6:152429000\t7.7\n",
      "[PB Info 2024-Oct-11 04:01:58] ProgressMeter -\tchr6:166362000\t7.8\n",
      "[PB Info 2024-Oct-11 04:02:04] ProgressMeter -\tchr7:9619000\t7.9\n",
      "[PB Info 2024-Oct-11 04:02:10] ProgressMeter -\tchr7:27113000\t8.0\n",
      "[PB Info 2024-Oct-11 04:02:16] ProgressMeter -\tchr7:37936000\t8.1\n",
      "[PB Info 2024-Oct-11 04:02:22] ProgressMeter -\tchr7:51084000\t8.2\n",
      "[PB Info 2024-Oct-11 04:02:28] ProgressMeter -\tchr7:68069000\t8.3\n",
      "[PB Info 2024-Oct-11 04:02:34] ProgressMeter -\tchr7:79574000\t8.4\n",
      "[PB Info 2024-Oct-11 04:02:40] ProgressMeter -\tchr7:93677000\t8.5\n",
      "[PB Info 2024-Oct-11 04:02:46] ProgressMeter -\tchr7:110258000\t8.6\n",
      "[PB Info 2024-Oct-11 04:02:52] ProgressMeter -\tchr7:120093000\t8.7\n",
      "[PB Info 2024-Oct-11 04:02:58] ProgressMeter -\tchr7:138221000\t8.8\n",
      "[PB Info 2024-Oct-11 04:03:04] ProgressMeter -\tchr7:153986000\t8.9\n",
      "[PB Info 2024-Oct-11 04:03:10] ProgressMeter -\tchr8:9070000\t9.0\n",
      "[PB Info 2024-Oct-11 04:03:16] ProgressMeter -\tchr8:19365000\t9.1\n",
      "[PB Info 2024-Oct-11 04:03:22] ProgressMeter -\tchr8:30710000\t9.2\n",
      "[PB Info 2024-Oct-11 04:03:28] ProgressMeter -\tchr8:43393000\t9.3\n",
      "[PB Info 2024-Oct-11 04:03:34] ProgressMeter -\tchr8:59253000\t9.4\n",
      "[PB Info 2024-Oct-11 04:03:40] ProgressMeter -\tchr8:74134000\t9.5\n",
      "[PB Info 2024-Oct-11 04:03:46] ProgressMeter -\tchr8:89942000\t9.6\n",
      "[PB Info 2024-Oct-11 04:03:52] ProgressMeter -\tchr8:105450000\t9.7\n",
      "[PB Info 2024-Oct-11 04:03:58] ProgressMeter -\tchr8:130659000\t9.8\n",
      "[PB Info 2024-Oct-11 04:04:04] ProgressMeter -\tchr8:141793000\t9.9\n",
      "[PB Info 2024-Oct-11 04:04:10] ProgressMeter -\tchr9:15688000\t10.0\n",
      "[PB Info 2024-Oct-11 04:04:16] ProgressMeter -\tchr9:22809000\t10.1\n",
      "[PB Info 2024-Oct-11 04:04:22] ProgressMeter -\tchr9:68305000\t10.2\n",
      "[PB Info 2024-Oct-11 04:04:28] ProgressMeter -\tchr9:82390000\t10.3\n",
      "[PB Info 2024-Oct-11 04:04:34] ProgressMeter -\tchr9:93778000\t10.4\n",
      "[PB Info 2024-Oct-11 04:04:40] ProgressMeter -\tchr9:107608000\t10.5\n",
      "[PB Info 2024-Oct-11 04:04:46] ProgressMeter -\tchr9:123806000\t10.6\n",
      "[PB Info 2024-Oct-11 04:04:52] ProgressMeter -\tchr10:1301000\t10.7\n",
      "[PB Info 2024-Oct-11 04:04:58] ProgressMeter -\tchr10:12189000\t10.8\n",
      "[PB Info 2024-Oct-11 04:05:04] ProgressMeter -\tchr10:27472000\t10.9\n",
      "[PB Info 2024-Oct-11 04:05:10] ProgressMeter -\tchr10:46326000\t11.0\n",
      "[PB Info 2024-Oct-11 04:05:16] ProgressMeter -\tchr10:63174000\t11.1\n",
      "[PB Info 2024-Oct-11 04:05:22] ProgressMeter -\tchr10:72008000\t11.2\n",
      "[PB Info 2024-Oct-11 04:05:28] ProgressMeter -\tchr10:88193000\t11.3\n",
      "[PB Info 2024-Oct-11 04:05:34] ProgressMeter -\tchr10:100523000\t11.4\n",
      "[PB Info 2024-Oct-11 04:05:40] ProgressMeter -\tchr10:117512000\t11.5\n",
      "[PB Info 2024-Oct-11 04:05:46] ProgressMeter -\tchr10:132700000\t11.6\n",
      "[PB Info 2024-Oct-11 04:05:52] ProgressMeter -\tchr11:15153000\t11.7\n",
      "[PB Info 2024-Oct-11 04:05:58] ProgressMeter -\tchr11:29955000\t11.8\n",
      "[PB Info 2024-Oct-11 04:06:04] ProgressMeter -\tchr11:41430000\t11.9\n",
      "[PB Info 2024-Oct-11 04:06:10] ProgressMeter -\tchr11:60248000\t12.0\n",
      "[PB Info 2024-Oct-11 04:06:16] ProgressMeter -\tchr11:72234000\t12.1\n",
      "[PB Info 2024-Oct-11 04:06:22] ProgressMeter -\tchr11:83954000\t12.2\n",
      "[PB Info 2024-Oct-11 04:06:28] ProgressMeter -\tchr11:92020000\t12.3\n",
      "[PB Info 2024-Oct-11 04:06:34] ProgressMeter -\tchr11:110049000\t12.4\n",
      "[PB Info 2024-Oct-11 04:06:40] ProgressMeter -\tchr11:119585000\t12.5\n",
      "[PB Info 2024-Oct-11 04:06:46] ProgressMeter -\tchr11:131136000\t12.6\n",
      "[PB Info 2024-Oct-11 04:06:52] ProgressMeter -\tchr12:8192000\t12.7\n",
      "[PB Info 2024-Oct-11 04:06:58] ProgressMeter -\tchr12:25218000\t12.8\n",
      "[PB Info 2024-Oct-11 04:07:04] ProgressMeter -\tchr12:38090000\t12.9\n",
      "[PB Info 2024-Oct-11 04:07:10] ProgressMeter -\tchr12:48780000\t13.0\n",
      "[PB Info 2024-Oct-11 04:07:16] ProgressMeter -\tchr12:56669000\t13.1\n",
      "[PB Info 2024-Oct-11 04:07:22] ProgressMeter -\tchr12:80465000\t13.2\n",
      "[PB Info 2024-Oct-11 04:07:28] ProgressMeter -\tchr12:95023000\t13.3\n",
      "[PB Info 2024-Oct-11 04:07:34] ProgressMeter -\tchr12:104161000\t13.4\n",
      "[PB Info 2024-Oct-11 04:07:40] ProgressMeter -\tchr12:123804000\t13.5\n",
      "[PB Info 2024-Oct-11 04:07:46] ProgressMeter -\tchr12:132649000\t13.6\n",
      "[PB Info 2024-Oct-11 04:07:52] ProgressMeter -\tchr13:28314000\t13.7\n",
      "[PB Info 2024-Oct-11 04:07:58] ProgressMeter -\tchr13:42569000\t13.8\n",
      "[PB Info 2024-Oct-11 04:08:04] ProgressMeter -\tchr13:50070000\t13.9\n",
      "[PB Info 2024-Oct-11 04:08:10] ProgressMeter -\tchr13:63451000\t14.0\n",
      "[PB Info 2024-Oct-11 04:08:16] ProgressMeter -\tchr13:91348000\t14.1\n",
      "[PB Info 2024-Oct-11 04:08:22] ProgressMeter -\tchr13:104127000\t14.2\n",
      "[PB Info 2024-Oct-11 04:08:28] ProgressMeter -\tchr14:22922000\t14.3\n",
      "[PB Info 2024-Oct-11 04:08:34] ProgressMeter -\tchr14:34762000\t14.4\n",
      "[PB Info 2024-Oct-11 04:08:40] ProgressMeter -\tchr14:52730000\t14.5\n",
      "[PB Info 2024-Oct-11 04:08:46] ProgressMeter -\tchr14:61347000\t14.6\n",
      "[PB Info 2024-Oct-11 04:08:52] ProgressMeter -\tchr14:77417000\t14.7\n",
      "[PB Info 2024-Oct-11 04:08:58] ProgressMeter -\tchr14:95052000\t14.8\n",
      "[PB Info 2024-Oct-11 04:09:04] ProgressMeter -\tchr14:103601000\t14.9\n",
      "[PB Info 2024-Oct-11 04:09:10] ProgressMeter -\tchr15:39372000\t15.0\n",
      "[PB Info 2024-Oct-11 04:09:16] ProgressMeter -\tchr15:56314000\t15.1\n",
      "[PB Info 2024-Oct-11 04:09:22] ProgressMeter -\tchr15:66702000\t15.2\n",
      "[PB Info 2024-Oct-11 04:09:28] ProgressMeter -\tchr15:78826000\t15.3\n",
      "[PB Info 2024-Oct-11 04:09:34] ProgressMeter -\tchr15:88671000\t15.4\n",
      "[PB Info 2024-Oct-11 04:09:40] ProgressMeter -\tchr16:3084000\t15.5\n",
      "[PB Info 2024-Oct-11 04:09:46] ProgressMeter -\tchr16:19657000\t15.6\n",
      "[PB Info 2024-Oct-11 04:09:52] ProgressMeter -\tchr16:31677000\t15.7\n",
      "[PB Info 2024-Oct-11 04:09:58] ProgressMeter -\tchr16:53266000\t15.8\n",
      "[PB Info 2024-Oct-11 04:10:04] ProgressMeter -\tchr16:66454000\t15.9\n",
      "[PB Info 2024-Oct-11 04:10:10] ProgressMeter -\tchr16:77669000\t16.0\n",
      "[PB Info 2024-Oct-11 04:10:16] ProgressMeter -\tchr17:4450000\t16.1\n",
      "[PB Info 2024-Oct-11 04:10:22] ProgressMeter -\tchr17:14613000\t16.2\n",
      "[PB Info 2024-Oct-11 04:10:28] ProgressMeter -\tchr17:21982000\t16.3\n",
      "[PB Info 2024-Oct-11 04:10:34] ProgressMeter -\tchr17:36001000\t16.4\n",
      "[PB Info 2024-Oct-11 04:10:40] ProgressMeter -\tchr17:52745000\t16.5\n",
      "[PB Info 2024-Oct-11 04:10:46] ProgressMeter -\tchr17:62713000\t16.6\n",
      "[PB Info 2024-Oct-11 04:10:52] ProgressMeter -\tchr17:77356000\t16.7\n",
      "[PB Info 2024-Oct-11 04:10:58] ProgressMeter -\tchr18:2083000\t16.8\n",
      "[PB Info 2024-Oct-11 04:11:04] ProgressMeter -\tchr18:21020000\t16.9\n",
      "[PB Info 2024-Oct-11 04:11:10] ProgressMeter -\tchr18:34294000\t17.0\n",
      "[PB Info 2024-Oct-11 04:11:16] ProgressMeter -\tchr18:47671000\t17.1\n",
      "[PB Info 2024-Oct-11 04:11:22] ProgressMeter -\tchr18:62638000\t17.2\n",
      "[PB Info 2024-Oct-11 04:11:28] ProgressMeter -\tchr18:77230000\t17.3\n",
      "[PB Info 2024-Oct-11 04:11:34] ProgressMeter -\tchr19:6307000\t17.4\n",
      "[PB Info 2024-Oct-11 04:11:40] ProgressMeter -\tchr19:12867000\t17.5\n",
      "[PB Info 2024-Oct-11 04:11:46] ProgressMeter -\tchr19:32606000\t17.6\n",
      "[PB Info 2024-Oct-11 04:11:52] ProgressMeter -\tchr19:47899000\t17.7\n",
      "[PB Info 2024-Oct-11 04:11:58] ProgressMeter -\tchr20:5120000\t17.8\n",
      "[PB Info 2024-Oct-11 04:12:04] ProgressMeter -\tchr20:21996000\t17.9\n",
      "[PB Info 2024-Oct-11 04:12:10] ProgressMeter -\tchr20:41040000\t18.0\n",
      "[PB Info 2024-Oct-11 04:12:16] ProgressMeter -\tchr20:51113000\t18.1\n",
      "[PB Info 2024-Oct-11 04:12:22] ProgressMeter -\tchr20:61554000\t18.2\n",
      "[PB Info 2024-Oct-11 04:12:28] ProgressMeter -\tchr21:15181000\t18.3\n",
      "[PB Info 2024-Oct-11 04:12:34] ProgressMeter -\tchr21:28186000\t18.4\n",
      "[PB Info 2024-Oct-11 04:12:40] ProgressMeter -\tchr21:45035000\t18.5\n",
      "[PB Info 2024-Oct-11 04:12:46] ProgressMeter -\tchr22:25246000\t18.6\n",
      "[PB Info 2024-Oct-11 04:12:52] ProgressMeter -\tchr22:40297000\t18.7\n",
      "[PB Info 2024-Oct-11 04:12:58] ProgressMeter -\tchrX:13068000\t18.8\n",
      "[PB Info 2024-Oct-11 04:13:04] ProgressMeter -\tchrX:37492000\t18.9\n",
      "[PB Info 2024-Oct-11 04:13:10] ProgressMeter -\tchrX:54279000\t19.0\n",
      "[PB Info 2024-Oct-11 04:13:16] ProgressMeter -\tchrX:84551000\t19.1\n",
      "[PB Info 2024-Oct-11 04:13:22] ProgressMeter -\tchrX:120283000\t19.2\n",
      "[PB Info 2024-Oct-11 04:13:28] ProgressMeter -\tchrX:140040000\t19.3\n",
      "[PB Info 2024-Oct-11 04:13:34] Deepvariant is finished, total time is 1164.233 seconds\n",
      "[PB Info 2024-Oct-11 04:13:34] ------------------------------------------------------------------------------\n",
      "[PB Info 2024-Oct-11 04:13:34] ||        Program:                                       deepvariant        ||\n",
      "[PB Info 2024-Oct-11 04:13:34] ||        Version:                                           4.3.2-1        ||\n",
      "[PB Info 2024-Oct-11 04:13:34] ||        Start Time:                       Fri Oct 11 03:53:55 2024        ||\n",
      "[PB Info 2024-Oct-11 04:13:34] ||        End Time:                         Fri Oct 11 04:13:34 2024        ||\n",
      "[PB Info 2024-Oct-11 04:13:34] ||        Total Time:                          19 minutes 39 seconds        ||\n",
      "[PB Info 2024-Oct-11 04:13:34] ------------------------------------------------------------------------------\n",
      "Variant caller done, total time: 19.7 min\n"
     ]
    }
   ],
   "source": [
    "!pbrun deepvariant \\\n",
    "    --ref sample_data/Ref/Homo_sapiens_assembly38.fasta \\\n",
    "    --in-bam outputdir/germline_out.bam \\\n",
    "    --out-variants outputdir/variants_dv.g.vcf \\\n",
    "    --num-streams-per-gpu 1 \\\n",
    "    --gvcf \\\n",
    "    --num-gpus 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "243bd52f-c9be-421e-8dc2-7ecd34e0e411",
   "metadata": {},
   "source": [
    "### Run Germline Workflow: GATK (BWA-MEM + HaplotypeCaller)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa71dce3-2cf7-45ac-a7ad-17f74e8f5666",
   "metadata": {},
   "source": [
    "This workflow runs the GATK germline pipeline in one command. This command will run read alignment followed by variant calling using HaplotypeCaller. It will not only output the VCF file, but also output intermediate BAM and BQSR files if needed. In this workshop, we are limited to a single V100 GPU with 16GB RAM. Therefore, the `--low-memory` option is used."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "71d68e6d-182e-44ac-a1b0-84843727ace7",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please visit https://docs.nvidia.com/clara/#parabricks for detailed documentation\n",
      "\n",
      "\n",
      "usage: pbrun germline <options>\n",
      "Help: pbrun germline -h\n",
      "\n",
      "Run Germline pipeline to convert FASTQ to VCF.\n",
      "\n",
      "options:\n",
      "  -h, --help            show this help message and exit\n",
      "\n",
      "Input Output file options:\n",
      "  Options for Input and Output files for this tool.\n",
      "\n",
      "  --ref REF             Path to the reference file. (default: None)\n",
      "  --in-fq [IN_FQ ...]   Path to the pair-ended FASTQ files followed by\n",
      "                        optional read groups with quotes (Example:\n",
      "                        \"@RG\\tID:foo\\tLB:lib1\\tPL:bar\\tSM:sample\\tPU:foo\").\n",
      "                        The files must be in fastq or fastq.gz format. All\n",
      "                        sets of inputs should have a read group; otherwise,\n",
      "                        none should have a read group, and it will be\n",
      "                        automatically added by the pipeline. This option can\n",
      "                        be repeated multiple times. Example 1: --in-fq\n",
      "                        sampleX_1_1.fastq.gz sampleX_1_2.fastq.gz --in-fq\n",
      "                        sampleX_2_1.fastq.gz sampleX_2_2.fastq.gz. Example 2:\n",
      "                        --in-fq sampleX_1_1.fastq.gz sampleX_1_2.fastq.gz\n",
      "                        \"@RG\\tID:foo\\tLB:lib1\\tPL:bar\\tSM:sample\\tPU:unit1\"\n",
      "                        --in-fq sampleX_2_1.fastq.gz sampleX_2_2.fastq.gz\n",
      "                        \"@RG\\tID:foo2\\tLB:lib1\\tPL:bar\\tSM:sample\\tPU:unit2\".\n",
      "                        For the same sample, Read Groups should have the same\n",
      "                        sample name (SM) and a different ID and PU. (default:\n",
      "                        None)\n",
      "  --in-se-fq [IN_SE_FQ ...]\n",
      "                        Path to the single-ended FASTQ file followed by\n",
      "                        optional read group with quotes (Example:\n",
      "                        \"@RG\\tID:foo\\tLB:lib1\\tPL:bar\\tSM:sample\\tPU:foo\").\n",
      "                        The file must be in fastq or fastq.gz format. Either\n",
      "                        all sets of inputs have a read group, or none should\n",
      "                        have one, and it will be automatically added by the\n",
      "                        pipeline. This option can be repeated multiple times.\n",
      "                        Example 1: --in-se-fq sampleX_1.fastq.gz --in-se-fq\n",
      "                        sampleX_2.fastq.gz . Example 2: --in-se-fq\n",
      "                        sampleX_1.fastq.gz\n",
      "                        \"@RG\\tID:foo\\tLB:lib1\\tPL:bar\\tSM:sample\\tPU:unit1\"\n",
      "                        --in-se-fq sampleX_2.fastq.gz\n",
      "                        \"@RG\\tID:foo2\\tLB:lib1\\tPL:bar\\tSM:sample\\tPU:unit2\" .\n",
      "                        For the same sample, Read Groups should have the same\n",
      "                        sample name (SM) and a different ID and PU. (default:\n",
      "                        None)\n",
      "  --knownSites KNOWNSITES\n",
      "                        Path to a known indels file. The file must be in\n",
      "                        vcf.gz format. This option can be used multiple times.\n",
      "                        (default: None)\n",
      "  --interval-file INTERVAL_FILE\n",
      "                        Path to an interval file in one of these formats:\n",
      "                        Picard-style (.interval_list or .picard), GATK-style\n",
      "                        (.list or .intervals), or BED file (.bed). This option\n",
      "                        can be used multiple times. (default: None)\n",
      "  --out-recal-file OUT_RECAL_FILE\n",
      "                        Path of the report file after Base Quality Score\n",
      "                        Recalibration. (default: None)\n",
      "  --out-bam OUT_BAM     Path of BAM file after Marking Duplicates. (default:\n",
      "                        None)\n",
      "  --htvc-bam-output HTVC_BAM_OUTPUT\n",
      "                        File to which assembled haplotypes should be written\n",
      "                        in HaplotypeCaller. If passing with --run-partition,\n",
      "                        multiple bam files will be outputed. (default: None)\n",
      "  --out-variants OUT_VARIANTS\n",
      "                        Path of the vcf/gvcf/gvcf.gz file after variant\n",
      "                        calling. (default: None)\n",
      "  --out-duplicate-metrics OUT_DUPLICATE_METRICS\n",
      "                        Path of duplicate metrics file after Marking\n",
      "                        Duplicates. (default: None)\n",
      "  --htvc-alleles HTVC_ALLELES\n",
      "                        Path of the vcf.gz force-call file. The set of alleles\n",
      "                        to force-call regardless of evidence. (default: None)\n",
      "\n",
      "Tool Options:\n",
      "  Options specific to the tool.\n",
      "\n",
      "  --max-read-length MAX_READ_LENGTH\n",
      "                        Maximum read length/size (i.e., sequence length) used\n",
      "                        for bwa and filtering FASTQ input (default: 480)\n",
      "  --min-read-length MIN_READ_LENGTH\n",
      "                        Minimum read length/size (i.e., sequence length) used\n",
      "                        for bwa and filtering FASTQ input (default: 1)\n",
      "  -L INTERVAL, --interval INTERVAL\n",
      "                        Interval within which to call bqsr from the input\n",
      "                        reads. All intervals will have a padding of 100 to get\n",
      "                        read records, and overlapping intervals will be\n",
      "                        combined. Interval files should be passed using the\n",
      "                        --interval-file option. This option can be used\n",
      "                        multiple times (e.g. \"-L chr1 -L chr2:10000 -L\n",
      "                        chr3:20000+ -L chr4:10000-20000\"). (default: None)\n",
      "  --bwa-options BWA_OPTIONS\n",
      "                        Pass supported bwa mem options as one string. The\n",
      "                        current original bwa mem supported options are: -M,\n",
      "                        -Y, -C, -T, -B, -U, -L, and -K (e.g. --bwa-options=\"-M\n",
      "                        -Y\") (default: None)\n",
      "  --no-warnings         Suppress warning messages about system thread and\n",
      "                        memory usage. (default: None)\n",
      "  --filter-flag FILTER_FLAG\n",
      "                        Don't generate SAM entries in the output if the\n",
      "                        entry's flag's meet this criteria. Criteria: (flag &\n",
      "                        filter != 0) (default: 0)\n",
      "  --skip-multiple-hits  Filter SAM entries whose length of SA is not 0\n",
      "                        (default: None)\n",
      "  --align-only          Generate output BAM after bwa-mem. The output will not\n",
      "                        be co-ordinate sorted or duplicates will not be marked\n",
      "                        (default: None)\n",
      "  --no-markdups         Do not perform the Mark Duplicates step. Return BAM\n",
      "                        after sorting. (default: None)\n",
      "  --fix-mate            Add mate cigar (MC) and mate quality (MQ) tags to the\n",
      "                        output file. (default: None)\n",
      "  --markdups-assume-sortorder-queryname\n",
      "                        Assume the reads are sorted by queryname for Marking\n",
      "                        Duplicates. This will mark secondary, supplementary,\n",
      "                        and unmapped reads as duplicates as well. This flag\n",
      "                        will not impact variant calling while increasing\n",
      "                        processing times. (default: None)\n",
      "  --markdups-picard-version-2182\n",
      "                        Assume marking duplicates to be similar to Picard\n",
      "                        version 2.18.2. (default: None)\n",
      "  --monitor-usage       Monitor approximate CPU utilization and host memory\n",
      "                        usage during execution. (default: None)\n",
      "  --optical-duplicate-pixel-distance OPTICAL_DUPLICATE_PIXEL_DISTANCE\n",
      "                        The maximum offset between two duplicate clusters in\n",
      "                        order to consider them optical duplicates. Ignored if\n",
      "                        --out-duplicate-metrics is not passed. (default: None)\n",
      "  --read-group-sm READ_GROUP_SM\n",
      "                        SM tag for read groups in this run. (default: None)\n",
      "  --read-group-lb READ_GROUP_LB\n",
      "                        LB tag for read groups in this run. (default: None)\n",
      "  --read-group-pl READ_GROUP_PL\n",
      "                        PL tag for read groups in this run. (default: None)\n",
      "  --read-group-id-prefix READ_GROUP_ID_PREFIX\n",
      "                        Prefix for the ID and PU tags for read groups in this\n",
      "                        run. This prefix will be used for all pairs of fastq\n",
      "                        files in this run. The ID and PU tags will consist of\n",
      "                        this prefix and an identifier, that will be unique for\n",
      "                        a pair of fastq files. (default: None)\n",
      "  -ip INTERVAL_PADDING, --interval-padding INTERVAL_PADDING\n",
      "                        Amount of padding (in base pairs) to add to each\n",
      "                        interval you are including. (default: None)\n",
      "  --standalone-bqsr     Run standalone BQSR. (default: None)\n",
      "  --haplotypecaller-options HAPLOTYPECALLER_OPTIONS\n",
      "                        Pass supported haplotype caller options as one string.\n",
      "                        The following are currently supported original\n",
      "                        haplotypecaller options: -A\n",
      "                        <AS_BaseQualityRankSumTest, AS_FisherStrand,\n",
      "                        AS_InbreedingCoeff, AS_MappingQualityRankSumTest,\n",
      "                        AS_QualByDepth, AS_RMSMappingQuality,\n",
      "                        AS_ReadPosRankSumTest, AS_StrandOddsRatio,\n",
      "                        BaseQualityRankSumTest, ChromosomeCounts,\n",
      "                        ClippingRankSumTest, Coverage, DepthPerAlleleBySample,\n",
      "                        DepthPerSampleHC, ExcessHet, FisherStrand,\n",
      "                        InbreedingCoeff, MappingQualityRankSumTest,\n",
      "                        QualByDepth, RMSMappingQuality, ReadPosRankSumTest,\n",
      "                        ReferenceBases, StrandBiasBySample, StrandOddsRatio,\n",
      "                        TandemRepeat, AssemblyComplexity>,-AX <same options as\n",
      "                        -A>,--output-mode <EMIT_VARIANTS_ONLY,\n",
      "                        EMIT_ALL_CONFIDENT_SITES, EMIT_ALL_ACTIVE_SITES>\n",
      "                        ,-max-reads-per-alignment-start <int>, -min-dangling-\n",
      "                        branch-length <int>, -min-pruning <int>, -pcr-indel-\n",
      "                        model <NONE, HOSTILE, AGGRESSIVE, CONSERVATIVE>,\n",
      "                        -standard-min-confidence-threshold-for-calling\n",
      "                        <int>(e.g. --haplotypecaller-options=\"-min-pruning 4\n",
      "                        -standard-min-confidence-threshold-for-calling 30\").\n",
      "                        (default: None)\n",
      "  --static-quantized-quals STATIC_QUANTIZED_QUALS\n",
      "                        Use static quantized quality scores to a given number\n",
      "                        of levels. Repeat this option multiple times for\n",
      "                        multiple bins. (default: None)\n",
      "  --gvcf                Generate variant calls in .gvcf Format. (default:\n",
      "                        None)\n",
      "  --disable-read-filter DISABLE_READ_FILTER\n",
      "                        Disable the read filters for BAM entries. Currently,\n",
      "                        the supported read filters that can be disabled are\n",
      "                        MappingQualityAvailableReadFilter,\n",
      "                        MappingQualityReadFilter,\n",
      "                        NotSecondaryAlignmentReadFilter, and\n",
      "                        WellformedReadFilter. (default: None)\n",
      "  --max-alternate-alleles MAX_ALTERNATE_ALLELES\n",
      "                        Maximum number of alternate alleles to genotype.\n",
      "                        (default: None)\n",
      "  -G ANNOTATION_GROUP, --annotation-group ANNOTATION_GROUP\n",
      "                        The groups of annotations to add to the output variant\n",
      "                        calls. Currently supported annotation groups are\n",
      "                        StandardAnnotation, StandardHCAnnotation, and\n",
      "                        AS_StandardAnnotation. (default: None)\n",
      "  -GQB GVCF_GQ_BANDS, --gvcf-gq-bands GVCF_GQ_BANDS\n",
      "                        Exclusive upper bounds for reference confidence GQ\n",
      "                        bands. Must be in the range [1, 100] and specified in\n",
      "                        increasing order. (default: None)\n",
      "  --rna                 Run haplotypecaller optimized for RNA data. (default:\n",
      "                        None)\n",
      "  --dont-use-soft-clipped-bases\n",
      "                        Don't use soft clipped bases for variant calling.\n",
      "                        (default: None)\n",
      "  --minimum-mapping-quality MINIMUM_MAPPING_QUALITY\n",
      "                        Minimum mapping quality to keep (inclusive). (default:\n",
      "                        None)\n",
      "  --mapping-quality-threshold-for-genotyping MAPPING_QUALITY_THRESHOLD_FOR_GENOTYPING\n",
      "                        Control the threshold for discounting reads from the\n",
      "                        genotyper due to mapping quality after the active\n",
      "                        region detection and assembly steps but before\n",
      "                        genotyping. (default: None)\n",
      "  --enable-dynamic-read-disqualification-for-genotyping\n",
      "                        Will enable less strict read disqualification low base\n",
      "                        quality reads. (default: None)\n",
      "  --min-base-quality-score MIN_BASE_QUALITY_SCORE\n",
      "                        Minimum base quality required to consider a base for\n",
      "                        calling. (default: None)\n",
      "  --adaptive-pruning    Use adaptive graph pruning algorithm when pruning De\n",
      "                        Bruijn graph. (default: None)\n",
      "  --force-call-filtered-alleles\n",
      "                        Force-call filtered alleles included in the resource\n",
      "                        specified by --alleles. (default: None)\n",
      "  --filter-reads-too-long\n",
      "                        Ignore all input bam reads with size > 500bp (default:\n",
      "                        None)\n",
      "  --no-alt-contigs      Get rid of output records for alternate contigs.\n",
      "                        (default: None)\n",
      "  --ploidy PLOIDY       Ploidy assumed for the BAM file. Currently only\n",
      "                        haploid (ploidy 1) and diploid (ploidy 2) are\n",
      "                        supported. (default: 2)\n",
      "  --sample-sex SAMPLE_SEX\n",
      "                        Sex of the sample input. This option will override the\n",
      "                        sex determined from any X/Y read ratio range. Must be\n",
      "                        either male or female. (default: None)\n",
      "  --range-male RANGE_MALE\n",
      "                        Inclusive male range for the X/Y read ratio. The sex\n",
      "                        is declared male if the actual ratio falls in the\n",
      "                        specified range. Syntax is \"<min>-<max>\" (e.g. \"--\n",
      "                        range-male 1-10\"). (default: None)\n",
      "  --range-female RANGE_FEMALE\n",
      "                        Inclusive female range for the X/Y read ratio. The sex\n",
      "                        is declared female if the actual ratio falls in the\n",
      "                        specified range. Syntax is \"<min>-<max>\" (e.g. \"--\n",
      "                        range-female 150-250\"). (default: None)\n",
      "  --use-GRCh37-regions  Use the pseudoautosomal regions for GRCh37 reference\n",
      "                        types. This flag should be used for GRCh37 and UCSC\n",
      "                        hg19 references. By default, GRCh38 regions are used.\n",
      "                        (default: None)\n",
      "\n",
      "Performance Options:\n",
      "  Performance Tuning Options.\n",
      "\n",
      "  --bwa-nstreams BWA_NSTREAMS\n",
      "                        Number of streams per GPU to use; note: more streams\n",
      "                        increases device memory usage (default: 4)\n",
      "  --bwa-cpu-thread-pool BWA_CPU_THREAD_POOL\n",
      "                        Number of threads to devote to CPU thread pool *per\n",
      "                        GPU* (default: 16)\n",
      "  --num-cpu-threads-per-stage NUM_CPU_THREADS_PER_STAGE\n",
      "                        (Same as above) Number of threads to devote to CPU\n",
      "                        thread pool *per GPU* (default: None)\n",
      "  --gpuwrite            Use one GPU to accelerate writing final BAM/CRAM.\n",
      "                        (default: None)\n",
      "  --gpuwrite-deflate-algo GPUWRITE_DEFLATE_ALGO\n",
      "                        Choose the nvCOMP DEFLATE algorithm to use with\n",
      "                        --gpuwrite. Note these options do not correspond to\n",
      "                        CPU DEFLATE options. Valid options are 1, 2, and 4.\n",
      "                        Option 1 is fastest, while options 2 and 4 have\n",
      "                        progressively lower throughputs but higher compression\n",
      "                        ratios. The default value is 1 when the user does not\n",
      "                        provide an input (i.e., None) (default: None)\n",
      "  --gpusort             Use GPUs to accelerate sorting and marking. (default:\n",
      "                        None)\n",
      "  --use-gds             Use GPUDirect Storage (GDS) to enable a direct data\n",
      "                        path for direct memory access (DMA) transfers between\n",
      "                        GPU memory and storage. Must be used concurrently with\n",
      "                        `--gpuwrite`. Please refer to Parabricks Documentation\n",
      "                        > Best Performance for information on how to set up\n",
      "                        and use GPUDirect Storage. (default: None)\n",
      "  --memory-limit MEMORY_LIMIT\n",
      "                        System memory limit in GBs during sorting and\n",
      "                        postsorting. By default, the limit is half of the\n",
      "                        total system memory. (default: 377)\n",
      "  --low-memory          Use low memory mode; will lower the number of streams\n",
      "                        per GPU (default: None)\n",
      "  --htvc-low-memory     Use low memory mode in htvc. (default: None)\n",
      "  --num-htvc-threads NUM_HTVC_THREADS\n",
      "                        Number of CPU threads. (default: 5)\n",
      "  --run-partition       Divide the whole genome into multiple partitions and\n",
      "                        run multiple processes at the same time, each on one\n",
      "                        partition. (default: None)\n",
      "  --gpu-num-per-partition GPU_NUM_PER_PARTITION\n",
      "                        Number of GPUs to use per partition. (default: None)\n",
      "  --read-from-tmp-dir   Running variant caller reading from bin files\n",
      "                        generated by Aligner and sort. Run postsort in\n",
      "                        parallel. This option will increase device memory\n",
      "                        usage. (default: None)\n",
      "\n",
      "Run Options:\n",
      "  Options related to resource/verbosity/etc.\n",
      "\n",
      "  --verbose             Enable verbose output. (default: None)\n",
      "  --x3                  Show full command line arguments. (default: None)\n",
      "  --logfile LOGFILE     Path to the log file. If not specified, messages will\n",
      "                        only be written to the standard error output.\n",
      "                        (default: None)\n",
      "  --tmp-dir TMP_DIR     Full path to the directory where temporary files will\n",
      "                        be stored. (default: .)\n",
      "  --with-petagene-dir WITH_PETAGENE_DIR\n",
      "                        Full path to the PetaGene installation directory. By\n",
      "                        default, this should have been installed at\n",
      "                        /opt/petagene. Use of this option also requires that\n",
      "                        the PetaLink library has been preloaded by setting the\n",
      "                        LD_PRELOAD environment variable. Optionally set the\n",
      "                        PETASUITE_REFPATH and PGCLOUD_CREDPATH environment\n",
      "                        variables that are used for data and credentials.\n",
      "                        Optionally set the PetaLinkMode environment variable\n",
      "                        that is used to further configure PetaLink, notably\n",
      "                        setting it to \"+write\" to enable outputting compressed\n",
      "                        BAM and .fastq files. (default: None)\n",
      "  --keep-tmp            Do not delete the directory storing temporary files\n",
      "                        after completion. (default: None)\n",
      "  --no-seccomp-override\n",
      "                        Do not override seccomp options for docker. (default:\n",
      "                        None)\n",
      "  --version             View compatible software versions. (default: None)\n",
      "  --num-gpus NUM_GPUS   Number of GPUs to use for a run. (default: 1)\n"
     ]
    }
   ],
   "source": [
    "!pbrun germline -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ec0f5de8-8223-4a2d-9104-50cbadbdaf90",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please visit https://docs.nvidia.com/clara/#parabricks for detailed documentation\n",
      "\n",
      "\n",
      "[Parabricks Options Mesg]: Automatically generating ID prefix\n",
      "[Parabricks Options Mesg]: Read group created for /home/yingja1227/germline_data/sample_1.fq.gz and\n",
      "/home/yingja1227/germline_data/sample_2.fq.gz\n",
      "[Parabricks Options Mesg]: @RG\\tID:HK3TJBCX2.1\\tLB:lib1\\tPL:bar\\tSM:sample\\tPU:HK3TJBCX2.1\n",
      "\n",
      "\n",
      "[Parabricks Options Mesg]: Checking argument compatibility\n",
      "[Parabricks Options Mesg]: Set --bwa-options=\"-K #\" to produce compatible pair-ended results with previous versions of\n",
      "fq2bam or BWA MEM.\n",
      "[Parabricks Options Mesg]: Read group created for /home/yingja1227/germline_data/sample_1.fq.gz and\n",
      "/home/yingja1227/germline_data/sample_2.fq.gz\n",
      "[Parabricks Options Mesg]: @RG\\tID:HK3TJBCX2.1\\tLB:lib1\\tPL:bar\\tSM:sample\\tPU:HK3TJBCX2.1\n",
      "[Parabricks Options Mesg]: Using --low-memory sets the number of streams in bwa mem to 1.\n",
      "[PB Info 2025-Aug-29 13:47:27] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Aug-29 13:47:27] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2025-Aug-29 13:47:27] ||                              Version 4.4.0-1                             ||\n",
      "[PB Info 2025-Aug-29 13:47:27] ||                      GPU-PBBWA mem, Sorting Phase-I                      ||\n",
      "[PB Info 2025-Aug-29 13:47:27] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Aug-29 13:47:27] Mode = pair-ended-gpu\n",
      "[PB Info 2025-Aug-29 13:47:27] Running with 1 GPU(s), using 1 stream(s) per device with 16 worker threads per GPU\n",
      "[PB Info 2025-Aug-29 13:47:37] # 67  1  0  1  0   0 pool:  1 0 bases/GPU/minute: 0.0 \n",
      "[PB Info 2025-Aug-29 13:47:47] # 100  1  5  0  0   0 pool:  0 158320292 bases/GPU/minute: 949921752.0 \n",
      "[PB Info 2025-Aug-29 13:47:57] # 99  1  5  1  0   0 pool:  2 286480655 bases/GPU/minute: 768962178.0 \n",
      "[PB Info 2025-Aug-29 13:48:07] # 100  4  2  4  0   0 pool:  0 414641812 bases/GPU/minute: 768966942.0 \n",
      "[PB Info 2025-Aug-29 13:48:17] # 100 10  0  5  0   0 pool:  1 565417974 bases/GPU/minute: 904656972.0 \n",
      "[PB Info 2025-Aug-29 13:48:27] # 100 24  0  5  0   0 pool:  1 693577188 bases/GPU/minute: 768955284.0 \n",
      "[PB Info 2025-Aug-29 13:48:37] # 100 37  1  5  0   0 pool:  1 829276697 bases/GPU/minute: 814197054.0 \n",
      "[PB Info 2025-Aug-29 13:48:47] # 100 52  1  5  0   0 pool:  0 957440246 bases/GPU/minute: 768981294.0 \n",
      "[PB Info 2025-Aug-29 13:48:57] # 100 59  2  4  0   0 pool:  2 1085601388 bases/GPU/minute: 768966852.0 \n",
      "[PB Info 2025-Aug-29 13:49:07] # 100 53  8  1  0   0 pool:  5 1266533049 bases/GPU/minute: 1085589966.0 \n",
      "[PB Info 2025-Aug-29 13:49:17] # 100 58  3  5  0   0 pool:  1 1394690250 bases/GPU/minute: 768943206.0 \n",
      "[PB Info 2025-Aug-29 13:49:27] # 100 57  4  4  0   0 pool:  2 1560543541 bases/GPU/minute: 995119746.0 \n",
      "[PB Info 2025-Aug-29 13:49:37] # 100 71  0  9  0   0 pool:  0 1688707176 bases/GPU/minute: 768981810.0 \n",
      "[PB Info 2025-Aug-29 13:49:47] # 100 90  0  8  0   0 pool:  0 1794252939 bases/GPU/minute: 633274578.0 \n",
      "[PB Info 2025-Aug-29 13:49:57] # 100 83  8  1  0   0 pool:  8 1937494137 bases/GPU/minute: 859447188.0 \n",
      "[PB Info 2025-Aug-29 13:50:07] # 100 28 63  4  0   0 pool:  5 2065653272 bases/GPU/minute: 768954810.0 \n",
      "[PB Info 2025-Aug-29 13:50:17] # 100 33 57 10  0   0 pool:  5 2216429924 bases/GPU/minute: 904659912.0 \n",
      "[PB Info 2025-Aug-29 13:50:27] # 100 29 62  5  0   0 pool:  9 2389825507 bases/GPU/minute: 1040373498.0 \n",
      "[PB Info 2025-Aug-29 13:50:37] # 100 30 61  6  0   0 pool:  8 2540604068 bases/GPU/minute: 904671366.0 \n",
      "[PB Info 2025-Aug-29 13:50:47] # 99 43 48 19  0   0 pool:  1 2668767731 bases/GPU/minute: 768981978.0 \n",
      "[PB Info 2025-Aug-29 13:50:57] # 100 58 33 33  0   0 pool:  0 2804463023 bases/GPU/minute: 814171752.0 \n",
      "[PB Info 2025-Aug-29 13:51:07] # 100 67 24 43  0   0 pool:  1 2940161745 bases/GPU/minute: 814192332.0 \n",
      "[PB Info 2025-Aug-29 13:51:17] # 100 79 12 54  0   0 pool:  1 3083395989 bases/GPU/minute: 859405464.0 \n",
      "[PB Info 2025-Aug-29 13:51:27] # 100 95  0 67  0   0 pool:  0 3211560682 bases/GPU/minute: 768988158.0 \n",
      "[PB Info 2025-Aug-29 13:51:37] # 100 86 10 56  0   0 pool: 11 3369880078 bases/GPU/minute: 949916376.0 \n",
      "[PB Info 2025-Aug-29 13:51:47] # 100 96  3 64  0   0 pool:  3 3528196738 bases/GPU/minute: 949899960.0 \n",
      "[PB Info 2025-Aug-29 13:51:57] # 100 100  1 66  0   0 pool:  0 3686509279 bases/GPU/minute: 949875246.0 \n",
      "[PB Info 2025-Aug-29 13:52:07] # 100 100  1 66  0   0 pool:  0 3822201849 bases/GPU/minute: 814155420.0 \n",
      "[PB Info 2025-Aug-29 13:52:17] # 100 100  1 66  0   0 pool:  0 3965443648 bases/GPU/minute: 859450794.0 \n",
      "[PB Info 2025-Aug-29 13:52:26] Time spent reading: 30.320062 seconds\n",
      "[PB Info 2025-Aug-29 13:52:27] # 100 100  1 65  0   0 pool:  1 4108685062 bases/GPU/minute: 859448484.0 \n",
      "[PB Info 2025-Aug-29 13:52:37] # 75 100  1 66  0   0 pool: 25 4289613826 bases/GPU/minute: 1085572584.0 \n",
      "[PB Info 2025-Aug-29 13:52:47] # 57 100  1 66  0   0 pool: 43 4425309104 bases/GPU/minute: 814171668.0 \n",
      "[PB Info 2025-Aug-29 13:52:57] # 35 100  1 66  0   0 pool: 65 4591160876 bases/GPU/minute: 995110632.0 \n",
      "[PB Info 2025-Aug-29 13:53:07] # 15 93  8 59  0   0 pool: 92 4794717229 bases/GPU/minute: 1221338118.0 \n",
      "[PB Info 2025-Aug-29 13:53:17] #  0 83  0 67  0   0 pool: 100 4990729766 bases/GPU/minute: 1176075222.0 \n",
      "[PB Info 2025-Aug-29 13:53:27] #  0 62  0 66  0   0 pool: 100 5149043402 bases/GPU/minute: 949881816.0 \n",
      "[PB Info 2025-Aug-29 13:53:37] #  0 34  0 66  0   0 pool: 100 5360126092 bases/GPU/minute: 1266496140.0 \n",
      "[PB Info 2025-Aug-29 13:53:47] #  0  3  1 66  0   0 pool: 100 5593836300 bases/GPU/minute: 1402261248.0 \n",
      "[PB Info 2025-Aug-29 13:53:48] GPU 0 exited\n",
      "[PB Info 2025-Aug-29 13:53:57] #  0  0  0 40  0   0 pool: 100 5812462784 bases/GPU/minute: 1311758904.0 \n",
      "[PB Info 2025-Aug-29 13:54:07] #  0  0  0 10  0   0 pool: 100 6046165146 bases/GPU/minute: 1402214172.0 \n",
      "[PB Info 2025-Aug-29 13:54:11] A CPU has exited\n",
      "[PB Info 2025-Aug-29 13:54:17] #  0  0  0  0  0   0 pool: 100 6133353829 bases/GPU/minute: 523132098.0 \n",
      "[PB Info 2025-Aug-29 13:54:27] #  0  0  0  0  0   0 pool: 100 6133353829 bases/GPU/minute: 0.0 \n",
      "[PB Info 2025-Aug-29 13:54:37] Rate stats (based on sampling every 10 seconds):\n",
      "\tmin rate: 0.0 bases/GPU/minute\n",
      "\tmax rate: 1402261248.0 bases/GPU/minute\n",
      "\tavg rate: 876193404.1 bases/GPU/minute\n",
      "[PB Info 2025-Aug-29 13:54:37] Time spent monitoring (multiple of 10): 430.073\n",
      "[PB Info 2025-Aug-29 13:54:37] bwalib run finished in 423.438 seconds\n",
      "[PB Info 2025-Aug-29 13:54:37] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Aug-29 13:54:37] ||        Program:                    GPU-PBBWA mem, Sorting Phase-I        ||\n",
      "[PB Info 2025-Aug-29 13:54:37] ||        Version:                                           4.4.0-1        ||\n",
      "[PB Info 2025-Aug-29 13:54:37] ||        Start Time:                       Fri Aug 29 13:47:27 2025        ||\n",
      "[PB Info 2025-Aug-29 13:54:37] ||        End Time:                         Fri Aug 29 13:54:37 2025        ||\n",
      "[PB Info 2025-Aug-29 13:54:37] ||        Total Time:                           7 minutes 10 seconds        ||\n",
      "[PB Info 2025-Aug-29 13:54:37] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Aug-29 13:54:38] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Aug-29 13:54:38] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2025-Aug-29 13:54:38] ||                              Version 4.4.0-1                             ||\n",
      "[PB Info 2025-Aug-29 13:54:38] ||                             Sorting Phase-II                             ||\n",
      "[PB Info 2025-Aug-29 13:54:38] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Aug-29 13:54:38] progressMeter - Percentage\n",
      "[PB Info 2025-Aug-29 13:54:38] 0.0\n",
      "[PB Info 2025-Aug-29 13:54:43] 38.9\n",
      "[PB Info 2025-Aug-29 13:54:48] 77.1\n",
      "[PB Info 2025-Aug-29 13:54:53] Sorting and Marking: 15.002 seconds\n",
      "[PB Info 2025-Aug-29 13:54:53] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Aug-29 13:54:53] ||        Program:                                  Sorting Phase-II        ||\n",
      "[PB Info 2025-Aug-29 13:54:53] ||        Version:                                           4.4.0-1        ||\n",
      "[PB Info 2025-Aug-29 13:54:53] ||        Start Time:                       Fri Aug 29 13:54:38 2025        ||\n",
      "[PB Info 2025-Aug-29 13:54:53] ||        End Time:                         Fri Aug 29 13:54:53 2025        ||\n",
      "[PB Info 2025-Aug-29 13:54:53] ||        Total Time:                                     15 seconds        ||\n",
      "[PB Info 2025-Aug-29 13:54:53] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Aug-29 13:54:53] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Aug-29 13:54:53] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2025-Aug-29 13:54:53] ||                              Version 4.4.0-1                             ||\n",
      "[PB Info 2025-Aug-29 13:54:53] ||                         Marking Duplicates, BQSR                         ||\n",
      "[PB Info 2025-Aug-29 13:54:53] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Aug-29 13:54:53] BQSR using CUDA device(s): { 0 }\n",
      "[PB Info 2025-Aug-29 13:54:53] Using PBBinBamFile for BAM writing\n",
      "[PB Info 2025-Aug-29 13:54:53] progressMeter -\tPercentage\n",
      "[PB Info 2025-Aug-29 13:55:03] 0.0\n",
      "[PB Info 2025-Aug-29 13:55:13] 0.0\n",
      "[PB Info 2025-Aug-29 13:55:20] Checking if the index file exists for the input compressed file:/home/yingja1227/Ref/Homo_sapiens_assembly38.known_indels.vcf.gz for better performance\n",
      "[W::bcf_hdr_check_sanity] GL should be declared as Number=G\n",
      "[PB Info 2025-Aug-29 13:55:23] 0.1\n",
      "[PB Info 2025-Aug-29 13:55:33] 8.4\n",
      "[PB Info 2025-Aug-29 13:55:43] 17.4\n",
      "[PB Info 2025-Aug-29 13:55:53] 28.2\n",
      "[PB Info 2025-Aug-29 13:56:03] 38.2\n",
      "[PB Info 2025-Aug-29 13:56:13] 47.4\n",
      "[PB Info 2025-Aug-29 13:56:23] 57.2\n",
      "[PB Info 2025-Aug-29 13:56:33] 65.8\n",
      "[PB Info 2025-Aug-29 13:56:43] 76.1\n",
      "[PB Info 2025-Aug-29 13:56:53] 82.3\n",
      "[PB Info 2025-Aug-29 13:56:59] 13144849 reads are filtered\n",
      "[PB Info 2025-Aug-29 13:57:03] 96.0\n",
      "[PB Info 2025-Aug-29 13:57:13] 100.0\n",
      "[PB Info 2025-Aug-29 13:57:13] BQSR and writing final BAM:  140.050 seconds\n",
      "[PB Info 2025-Aug-29 13:57:13] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Aug-29 13:57:13] ||        Program:                          Marking Duplicates, BQSR        ||\n",
      "[PB Info 2025-Aug-29 13:57:13] ||        Version:                                           4.4.0-1        ||\n",
      "[PB Info 2025-Aug-29 13:57:13] ||        Start Time:                       Fri Aug 29 13:54:53 2025        ||\n",
      "[PB Info 2025-Aug-29 13:57:13] ||        End Time:                         Fri Aug 29 13:57:13 2025        ||\n",
      "[PB Info 2025-Aug-29 13:57:13] ||        Total Time:                           2 minutes 20 seconds        ||\n",
      "[PB Info 2025-Aug-29 13:57:13] ------------------------------------------------------------------------------\n",
      "/usr/local/parabricks/binaries/bin/htvc /home/yingja1227/Ref/Homo_sapiens_assembly38.fasta /home/yingja1227/germline_output/germline_out.bam 1 -o /home/yingja1227/germline_output/variants.vcf -nt 5 -a /home/yingja1227/germline_output/recal.table --low-memory\n",
      "[PB Info 2025-Aug-29 13:57:15] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Aug-29 13:57:15] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2025-Aug-29 13:57:15] ||                              Version 4.4.0-1                             ||\n",
      "[PB Info 2025-Aug-29 13:57:15] ||                         GPU-GATK4 HaplotypeCaller                        ||\n",
      "[PB Info 2025-Aug-29 13:57:15] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Aug-29 13:57:26] 0 /home/yingja1227/germline_output/germline_out.bam/home/yingja1227/germline_output/variants.vcf\n",
      "[PB Info 2025-Aug-29 13:57:26] ProgressMeter -\tCurrent-Locus\tElapsed-Minutes\tRegions-Processed\tRegions/Minute\n",
      "[PB Info 2025-Aug-29 13:57:36] chr1:7660801\t0.2\t62132\t372792\n",
      "[PB Info 2025-Aug-29 13:57:46] chr1:23078401\t0.3\t129348\t388044\n",
      "[PB Info 2025-Aug-29 13:57:56] chr1:47212801\t0.5\t245254\t490508\n",
      "[PB Info 2025-Aug-29 13:58:06] chr1:64843201\t0.7\t322239\t483358\n",
      "[PB Info 2025-Aug-29 13:58:16] chr1:114393601\t0.8\t523721\t628464\n",
      "[PB Info 2025-Aug-29 13:58:26] chr1:158865601\t1.0\t639532\t639526\n",
      "[PB Info 2025-Aug-29 13:58:36] chr1:193228801\t1.2\t776053\t665187\n",
      "[PB Info 2025-Aug-29 13:58:46] chr1:235545601\t1.3\t970909\t728181\n",
      "[PB Info 2025-Aug-29 13:58:56] chr2:17707201\t1.5\t1104200\t736133\n",
      "[PB Info 2025-Aug-29 13:59:06] chr2:39288001\t1.7\t1196391\t717834\n",
      "[PB Info 2025-Aug-29 13:59:16] chr2:49012801\t1.8\t1252359\t683104\n",
      "[PB Info 2025-Aug-29 13:59:26] chr2:81921601\t2.0\t1394131\t697065\n",
      "[PB Info 2025-Aug-29 13:59:36] chr2:102388801\t2.2\t1464327\t675843\n",
      "[PB Info 2025-Aug-29 13:59:46] chr2:128054401\t2.3\t1576294\t675554\n",
      "[PB Info 2025-Aug-29 13:59:56] chr2:149107201\t2.5\t1663983\t665593\n",
      "[PB Info 2025-Aug-29 14:00:06] chr2:174590401\t2.7\t1769161\t663435\n",
      "[PB Info 2025-Aug-29 14:00:16] chr2:207595201\t2.8\t1911131\t674516\n",
      "[PB Info 2025-Aug-29 14:00:26] chr3:172801\t3.0\t2063625\t687875\n",
      "[PB Info 2025-Aug-29 14:00:36] chr3:30648001\t3.2\t2192867\t692484\n",
      "[PB Info 2025-Aug-29 14:00:46] chr3:49684801\t3.3\t2280575\t684172\n",
      "[PB Info 2025-Aug-29 14:00:56] chr3:71198401\t3.5\t2383111\t680888\n",
      "[PB Info 2025-Aug-29 14:01:06] chr3:119256001\t3.7\t2552074\t696020\n",
      "[PB Info 2025-Aug-29 14:01:16] chr3:134155201\t3.8\t2638595\t688329\n",
      "[PB Info 2025-Aug-29 14:01:26] chr3:149572801\t4.0\t2713495\t678373\n",
      "[PB Info 2025-Aug-29 14:01:36] chr3:189868801\t4.2\t2874675\t689922\n",
      "[PB Info 2025-Aug-29 14:01:46] chr4:37958401\t4.3\t3057117\t705487\n",
      "[PB Info 2025-Aug-29 14:01:56] chr4:75859201\t4.5\t3222493\t716108\n",
      "[PB Info 2025-Aug-29 14:02:06] chr4:120057601\t4.7\t3412831\t731320\n",
      "[PB Info 2025-Aug-29 14:02:16] chr4:152409601\t4.8\t3559494\t736447\n",
      "[PB Info 2025-Aug-29 14:02:26] chr5:14481601\t5.0\t3779495\t755899\n",
      "[PB Info 2025-Aug-29 14:02:36] chr5:52785601\t5.2\t3938716\t762331\n",
      "[PB Info 2025-Aug-29 14:02:46] chr5:80654401\t5.3\t4049941\t759363\n",
      "[PB Info 2025-Aug-29 14:02:56] chr5:131976001\t5.5\t4261314\t774784\n",
      "[PB Info 2025-Aug-29 14:03:06] chr5:170030401\t5.7\t4410133\t778258\n",
      "[PB Info 2025-Aug-29 14:03:16] chr6:36072001\t5.8\t4609009\t790115\n",
      "[PB Info 2025-Aug-29 14:03:26] chr6:72859201\t6.0\t4778856\t796476\n",
      "[PB Info 2025-Aug-29 14:03:36] chr6:88896001\t6.2\t4845517\t785759\n",
      "[PB Info 2025-Aug-29 14:03:46] chr6:117364801\t6.3\t4966526\t784188\n",
      "[PB Info 2025-Aug-29 14:03:56] chr6:152428801\t6.5\t5110105\t786170\n",
      "[PB Info 2025-Aug-29 14:04:06] chr7:2227201\t6.7\t5189630\t778444\n",
      "[PB Info 2025-Aug-29 14:04:16] chr7:36691201\t6.8\t5326283\t779456\n",
      "[PB Info 2025-Aug-29 14:04:26] chr7:86284801\t7.0\t5535083\t790726\n",
      "[PB Info 2025-Aug-29 14:04:36] chr7:102936001\t7.2\t5631262\t785757\n",
      "[PB Info 2025-Aug-29 14:04:46] chr7:133171201\t7.3\t5755728\t784872\n",
      "[PB Info 2025-Aug-29 14:04:56] chr8:8558401\t7.5\t5872723\t783029\n",
      "[PB Info 2025-Aug-29 14:05:06] chr8:41942401\t7.7\t6045350\t788523\n",
      "[PB Info 2025-Aug-29 14:05:16] chr8:63460801\t7.8\t6131785\t782781\n",
      "[PB Info 2025-Aug-29 14:05:26] chr8:95524801\t8.0\t6266974\t783371\n",
      "[PB Info 2025-Aug-29 14:05:36] chr9:432001\t8.2\t6478137\t793241\n",
      "[PB Info 2025-Aug-29 14:05:46] chr9:32611201\t8.3\t6615949\t793913\n",
      "[PB Info 2025-Aug-29 14:05:56] chr9:99547201\t8.5\t6827561\t803242\n",
      "[PB Info 2025-Aug-29 14:06:06] chr9:130833601\t8.7\t6941695\t800964\n",
      "[PB Info 2025-Aug-29 14:06:16] chr10:18936001\t8.8\t7078531\t801343\n",
      "[PB Info 2025-Aug-29 14:06:26] chr10:82977601\t9.0\t7323920\t813768\n",
      "[PB Info 2025-Aug-29 14:06:36] chr10:109680001\t9.2\t7459930\t813810\n",
      "[PB Info 2025-Aug-29 14:06:46] chr11:15816001\t9.3\t7604538\t814771\n",
      "[PB Info 2025-Aug-29 14:06:56] chr11:61444801\t9.5\t7816759\t822816\n",
      "[PB Info 2025-Aug-29 14:07:06] chr11:102182401\t9.7\t8003888\t827988\n",
      "[PB Info 2025-Aug-29 14:07:16] chr12:3177601\t9.8\t8152531\t829070\n",
      "[PB Info 2025-Aug-29 14:07:26] chr12:45816001\t10.0\t8326469\t832646\n",
      "[PB Info 2025-Aug-29 14:07:36] chr12:70742401\t10.2\t8458735\t832006\n",
      "[PB Info 2025-Aug-29 14:07:46] chr12:118233601\t10.3\t8637751\t835911\n",
      "[PB Info 2025-Aug-29 14:07:56] chr13:28368001\t10.5\t8765595\t834818\n",
      "[PB Info 2025-Aug-29 14:08:06] chr13:60331201\t10.7\t8919673\t836219\n",
      "[PB Info 2025-Aug-29 14:08:16] chr14:37545601\t10.8\t9221395\t851205\n",
      "[PB Info 2025-Aug-29 14:08:26] chr14:92822401\t11.0\t9442363\t858396\n",
      "[PB Info 2025-Aug-29 14:08:36] chr15:67747201\t11.2\t9697163\t868402\n",
      "[PB Info 2025-Aug-29 14:08:46] chr16:54628801\t11.3\t10026077\t884653\n",
      "[PB Info 2025-Aug-29 14:08:56] chr17:17260801\t11.5\t10265392\t892642\n",
      "[PB Info 2025-Aug-29 14:09:06] chr17:62241601\t11.7\t10428589\t893879\n",
      "[PB Info 2025-Aug-29 14:09:16] chr18:24033601\t11.8\t10622651\t897688\n",
      "[PB Info 2025-Aug-29 14:09:26] chr19:11107201\t12.0\t10892015\t907667\n",
      "[PB Info 2025-Aug-29 14:09:36] chr19:48960001\t12.2\t11051104\t908309\n",
      "[PB Info 2025-Aug-29 14:09:46] chr20:47246401\t12.3\t11291961\t915564\n",
      "[PB Info 2025-Aug-29 14:09:56] chr21:39273601\t12.5\t11489179\t919134\n",
      "[PB Info 2025-Aug-29 14:10:06] chr22:47284801\t12.7\t11664627\t920891\n",
      "[PB Info 2025-Aug-29 14:10:16] chrX:105364801\t12.8\t12096321\t942570\n",
      "[PB Info 2025-Aug-29 14:10:26] chrUn_KI270743v1:163201\t13.0\t12389267\t953020\n",
      "[PB Info 2025-Aug-29 14:10:36] Total time taken: 790.394640\n",
      "[PB Info 2025-Aug-29 14:10:36] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Aug-29 14:10:36] ||        Program:                         GPU-GATK4 HaplotypeCaller        ||\n",
      "[PB Info 2025-Aug-29 14:10:36] ||        Version:                                           4.4.0-1        ||\n",
      "[PB Info 2025-Aug-29 14:10:36] ||        Start Time:                       Fri Aug 29 13:57:15 2025        ||\n",
      "[PB Info 2025-Aug-29 14:10:36] ||        End Time:                         Fri Aug 29 14:10:36 2025        ||\n",
      "[PB Info 2025-Aug-29 14:10:36] ||        Total Time:                          13 minutes 21 seconds        ||\n",
      "[PB Info 2025-Aug-29 14:10:36] ------------------------------------------------------------------------------\n",
      "Variant caller done, total time: 13.4 min\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[PB Info 2025-Aug-27 16:33:50] chr6:62342401\t1.8\t4777598\t2605962\n",
      "[PB Info 2025-Aug-27 16:34:00] chr6:161044801\t2.0\t5186472\t2593236\n",
      "[PB Info 2025-Aug-27 16:34:10] chr7:107764801\t2.2\t5705770\t2633432\n",
      "[PB Info 2025-Aug-27 16:34:20] chr8:62294401\t2.3\t6177174\t2647360\n",
      "[PB Info 2025-Aug-27 16:34:30] chr9:21998401\t2.5\t6609104\t2643641\n",
      "[PB Info 2025-Aug-27 16:34:40] chr10:29961601\t2.7\t7094248\t2660343\n",
      "[PB Info 2025-Aug-27 16:34:50] chr11:3681601\t2.8\t7592991\t2679879\n",
      "[PB Info 2025-Aug-27 16:35:00] chr11:122064001\t3.0\t8123867\t2707955\n",
      "[PB Info 2025-Aug-27 16:35:10] chr12:92136001\t3.2\t8551733\t2700547\n",
      "[PB Info 2025-Aug-27 16:35:20] chr13:102849601\t3.3\t9047105\t2714131\n",
      "[PB Info 2025-Aug-27 16:35:30] chr15:41577601\t3.5\t9582370\t2737820\n",
      "[PB Info 2025-Aug-27 16:35:40] chr16:56817601\t3.7\t10061794\t2744125\n",
      "[PB Info 2025-Aug-27 16:35:50] chr17:68524801\t3.8\t10464699\t2729921\n",
      "[PB Info 2025-Aug-27 16:36:00] chr19:32352001\t4.0\t10955058\t2738764\n",
      "[PB Info 2025-Aug-27 16:36:10] chr20:58627201\t4.2\t11338974\t2721353\n",
      "[PB Info 2025-Aug-27 16:36:20] chrX:30273601\t4.3\t11790102\t2720792\n",
      "[PB Info 2025-Aug-27 16:36:30] Total time taken: 270.244529\n",
      "[PB Info 2025-Aug-27 16:36:30] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Aug-27 16:36:30] ||        Program:                         GPU-GATK4 HaplotypeCaller        ||\n",
      "[PB Info 2025-Aug-27 16:36:30] ||        Version:                                           4.0.1-1        ||\n",
      "[PB Info 2025-Aug-27 16:36:30] ||        Start Time:                       Wed Aug 27 16:31:48 2025        ||\n",
      "[PB Info 2025-Aug-27 16:36:30] ||        End Time:                         Wed Aug 27 16:36:30 2025        ||\n",
      "[PB Info 2025-Aug-27 16:36:30] ||        Total Time:                           4 minutes 42 seconds        ||\n",
      "[PB Info 2025-Aug-27 16:36:30] ------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "!pbrun germline \\\n",
    "    --ref Ref/Homo_sapiens_assembly38.fasta \\\n",
    "    --in-fq germline_data/sample_1.fq.gz germline_data/sample_2.fq.gz \\\n",
    "    --knownSites Ref/Homo_sapiens_assembly38.known_indels.vcf.gz \\\n",
    "    --out-bam germline_output/germline_out.bam \\\n",
    "    --out-variants germline_output/variants.vcf \\\n",
    "    --out-recal-file germline_output/recal.table \\\n",
    "    --low-memory   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2d92fa9d-c6d0-4d54-98d5-fade52ee9449",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "germline_out.bam      germline_out_chrs.txt  variants.vcf\n",
      "germline_out.bam.bai  recal.table\n"
     ]
    }
   ],
   "source": [
    "!ls germline_output"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e340c76f-6616-4256-802f-73440d390f2a",
   "metadata": {},
   "source": [
    "### Run Germline Workflow: BWA-MEM + DeepVariant"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "995ebc61-4b77-4029-b98d-676b37eba9dd",
   "metadata": {},
   "source": [
    "Germline workflow with DeepVariant is also available using the below command."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "2df97884-e445-4971-895c-11a7bb1e7e8b",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please visit https://docs.nvidia.com/clara/#parabricks for detailed documentation\n",
      "\n",
      "\n",
      "[Parabricks Options Mesg]: Automatically generating ID prefix\n",
      "[Parabricks Options Mesg]: Read group created for /home/yingja1227/germline_data/sample_data/Data/sample_1.fq.gz and\n",
      "/home/yingja1227/germline_data/sample_data/Data/sample_2.fq.gz\n",
      "[Parabricks Options Mesg]: @RG\\tID:HK3TJBCX2.1\\tLB:lib1\\tPL:bar\\tSM:sample\\tPU:HK3TJBCX2.1\n",
      "\n",
      "\n",
      "[Parabricks Options Mesg]: Checking argument compatibility\n",
      "[Parabricks Options Mesg]: Read group created for /home/yingja1227/germline_data/sample_data/Data/sample_1.fq.gz and\n",
      "/home/yingja1227/germline_data/sample_data/Data/sample_2.fq.gz\n",
      "[Parabricks Options Mesg]: @RG\\tID:HK3TJBCX2.1\\tLB:lib1\\tPL:bar\\tSM:sample\\tPU:HK3TJBCX2.1\n",
      "[PB Info 2025-Aug-27 16:39:11] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Aug-27 16:39:11] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2025-Aug-27 16:39:11] ||                              Version 4.0.1-1                             ||\n",
      "[PB Info 2025-Aug-27 16:39:11] ||                       GPU-BWA mem, Sorting Phase-I                       ||\n",
      "[PB Info 2025-Aug-27 16:39:11] ------------------------------------------------------------------------------\n",
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
      "[PB Info 2025-Aug-27 16:39:14] GPU-BWA mem\n",
      "[PB Info 2025-Aug-27 16:39:14] ProgressMeter\tReads\t\tBase Pairs Aligned\n",
      "[PB Info 2025-Aug-27 16:39:43] 5043564\t\t550000000\n",
      "[PB Info 2025-Aug-27 16:40:03] 10087128\t1120000000\n",
      "[PB Info 2025-Aug-27 16:40:25] 15130692\t1740000000\n",
      "[PB Info 2025-Aug-27 16:40:47] 20174256\t2270000000\n",
      "[PB Info 2025-Aug-27 16:41:09] 25217820\t2920000000\n",
      "[PB Info 2025-Aug-27 16:41:30] 30261384\t3430000000\n",
      "[PB Info 2025-Aug-27 16:41:51] 35304948\t4060000000\n",
      "[PB Info 2025-Aug-27 16:42:10] 40348512\t4610000000\n",
      "[PB Info 2025-Aug-27 16:42:29] 45392076\t5210000000\n",
      "[PB Info 2025-Aug-27 16:42:49] 50435640\t5800000000\n",
      "[PB Info 2025-Aug-27 16:43:38] \n",
      "GPU-BWA Mem time: 264.044989 seconds\n",
      "[PB Info 2025-Aug-27 16:43:38] GPU-BWA Mem is finished.\n",
      "\n",
      "\n",
      "[main] CMD: /usr/local/parabricks/binaries//bin/bwa mem -Z ./pbOpts.txt /home/yingja1227/germline_data/sample_data/Ref/Homo_sapiens_assembly38.fasta /home/yingja1227/germline_data/sample_data/Data/sample_1.fq.gz /home/yingja1227/germline_data/sample_data/Data/sample_2.fq.gz @RG\\tID:HK3TJBCX2.1\\tLB:lib1\\tPL:bar\\tSM:sample\\tPU:HK3TJBCX2.1\n",
      "[main] Real time: 267.351 sec; CPU: 1852.082 sec\n",
      "[PB Info 2025-Aug-27 16:43:38] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Aug-27 16:43:38] ||        Program:                      GPU-BWA mem, Sorting Phase-I        ||\n",
      "[PB Info 2025-Aug-27 16:43:38] ||        Version:                                           4.0.1-1        ||\n",
      "[PB Info 2025-Aug-27 16:43:38] ||        Start Time:                       Wed Aug 27 16:39:11 2025        ||\n",
      "[PB Info 2025-Aug-27 16:43:38] ||        End Time:                         Wed Aug 27 16:43:38 2025        ||\n",
      "[PB Info 2025-Aug-27 16:43:38] ||        Total Time:                           4 minutes 27 seconds        ||\n",
      "[PB Info 2025-Aug-27 16:43:38] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Aug-27 16:43:39] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Aug-27 16:43:39] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2025-Aug-27 16:43:39] ||                              Version 4.0.1-1                             ||\n",
      "[PB Info 2025-Aug-27 16:43:39] ||                             Sorting Phase-II                             ||\n",
      "[PB Info 2025-Aug-27 16:43:39] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Aug-27 16:43:39] progressMeter - Percentage\n",
      "[PB Info 2025-Aug-27 16:43:39] 0.0\t 0.00 GB\n",
      "[PB Info 2025-Aug-27 16:43:49] Sorting and Marking: 10.000 seconds\n",
      "[PB Info 2025-Aug-27 16:43:49] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Aug-27 16:43:49] ||        Program:                                  Sorting Phase-II        ||\n",
      "[PB Info 2025-Aug-27 16:43:49] ||        Version:                                           4.0.1-1        ||\n",
      "[PB Info 2025-Aug-27 16:43:49] ||        Start Time:                       Wed Aug 27 16:43:39 2025        ||\n",
      "[PB Info 2025-Aug-27 16:43:49] ||        End Time:                         Wed Aug 27 16:43:49 2025        ||\n",
      "[PB Info 2025-Aug-27 16:43:49] ||        Total Time:                                     10 seconds        ||\n",
      "[PB Info 2025-Aug-27 16:43:49] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Aug-27 16:43:49] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Aug-27 16:43:49] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2025-Aug-27 16:43:49] ||                              Version 4.0.1-1                             ||\n",
      "[PB Info 2025-Aug-27 16:43:49] ||                         Marking Duplicates, BQSR                         ||\n",
      "[PB Info 2025-Aug-27 16:43:49] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Aug-27 16:43:49] progressMeter -\tPercentage\n",
      "[PB Info 2025-Aug-27 16:43:59] 22.9\t 5.45 GB\n",
      "[PB Info 2025-Aug-27 16:44:09] 49.4\t 9.83 GB\n",
      "[PB Info 2025-Aug-27 16:44:19] 76.7\t 4.02 GB\n",
      "[PB Info 2025-Aug-27 16:44:29] 100.0\t 0.00 GB\n",
      "[PB Info 2025-Aug-27 16:44:29] BQSR and writing final BAM:  40.034 seconds\n",
      "[PB Info 2025-Aug-27 16:44:29] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Aug-27 16:44:29] ||        Program:                          Marking Duplicates, BQSR        ||\n",
      "[PB Info 2025-Aug-27 16:44:29] ||        Version:                                           4.0.1-1        ||\n",
      "[PB Info 2025-Aug-27 16:44:29] ||        Start Time:                       Wed Aug 27 16:43:49 2025        ||\n",
      "[PB Info 2025-Aug-27 16:44:29] ||        End Time:                         Wed Aug 27 16:44:29 2025        ||\n",
      "[PB Info 2025-Aug-27 16:44:29] ||        Total Time:                                     40 seconds        ||\n",
      "[PB Info 2025-Aug-27 16:44:29] ------------------------------------------------------------------------------\n",
      "Detected 2 CUDA Capable device(s), considering 2 device(s)\n",
      "  CUDA Driver Version / Runtime Version          12.2 / 11.7\n",
      "Using model for CUDA Capability Major/Minor version number:    70\n",
      "/usr/local/parabricks/binaries//bin/deepvariant /home/yingja1227/germline_data/sample_data/Ref/Homo_sapiens_assembly38.fasta /home/yingja1227/germline_data/outputdir/dvgermline_out.bam 2 -o /home/yingja1227/germline_data/outputdir/variants_dv.vcf --model /usr/local/parabricks/binaries//model/70/shortread/deepvariant.eng -n 4 --channel_insert_size --pileup_image_width 221 --max_reads_per_partition 1500 --partition_size 1000 --vsc_min_count_snps 2 --vsc_min_count_indels 2 --vsc_min_fraction_snps 0.12 --min_mapping_quality 5 --min_base_quality 10 --alt_aligned_pileup none --variant_caller VERY_SENSITIVE_CALLER --dbg_min_base_quality 15 --ws_min_windows_distance 80 --aux_fields_to_keep HP\n",
      "[PB Info 2025-Aug-27 16:44:31] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Aug-27 16:44:31] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2025-Aug-27 16:44:31] ||                              Version 4.0.1-1                             ||\n",
      "[PB Info 2025-Aug-27 16:44:31] ||                                deepvariant                               ||\n",
      "[PB Info 2025-Aug-27 16:44:31] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Aug-27 16:44:31] Starting DeepVariant\n",
      "[PB Info 2025-Aug-27 16:44:31] Running with 2 gpu, each with 4 workers\n",
      "[PB Info 2025-Aug-27 16:44:42] ProgressMeter -\tCurrent-Locus\tElapsed-Minutes\n",
      "[PB Info 2025-Aug-27 16:44:48] ProgressMeter -\tchr1:10000\t0.1\n",
      "[PB Info 2025-Aug-27 16:44:54] ProgressMeter -\tchr1:6713000\t0.2\n",
      "[PB Info 2025-Aug-27 16:45:00] ProgressMeter -\tchr1:53504000\t0.3\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[PB Info 2025-Aug-27 16:45:06] ProgressMeter -\tchr1:158292000\t0.4\n",
      "[PB Info 2025-Aug-27 16:45:12] ProgressMeter -\tchr1:229843000\t0.5\n",
      "[PB Info 2025-Aug-27 16:45:18] ProgressMeter -\tchr2:40154000\t0.6\n",
      "[PB Info 2025-Aug-27 16:45:24] ProgressMeter -\tchr2:146693000\t0.7\n",
      "[PB Info 2025-Aug-27 16:45:30] ProgressMeter -\tchr3:4036000\t0.8\n",
      "[PB Info 2025-Aug-27 16:45:36] ProgressMeter -\tchr3:119975000\t0.9\n",
      "[PB Info 2025-Aug-27 16:45:42] ProgressMeter -\tchr4:15711000\t1.0\n",
      "[PB Info 2025-Aug-27 16:45:48] ProgressMeter -\tchr4:140268000\t1.1\n",
      "[PB Info 2025-Aug-27 16:45:54] ProgressMeter -\tchr5:67945000\t1.2\n",
      "[PB Info 2025-Aug-27 16:46:00] ProgressMeter -\tchr5:132578000\t1.3\n",
      "[PB Info 2025-Aug-27 16:46:06] ProgressMeter -\tchr6:36353000\t1.4\n",
      "[PB Info 2025-Aug-27 16:46:12] ProgressMeter -\tchr7:22983000\t1.5\n",
      "[PB Info 2025-Aug-27 16:46:18] ProgressMeter -\tchr7:108547000\t1.6\n",
      "[PB Info 2025-Aug-27 16:46:24] ProgressMeter -\tchr8:106980000\t1.7\n",
      "[PB Info 2025-Aug-27 16:46:30] ProgressMeter -\tchr9:83624000\t1.8\n",
      "[PB Info 2025-Aug-27 16:46:36] ProgressMeter -\tchr10:59277000\t1.9\n",
      "[PB Info 2025-Aug-27 16:46:42] ProgressMeter -\tchr11:17153000\t2.0\n",
      "[PB Info 2025-Aug-27 16:46:48] ProgressMeter -\tchr11:118426000\t2.1\n",
      "[PB Info 2025-Aug-27 16:46:54] ProgressMeter -\tchr12:50719000\t2.2\n",
      "[PB Info 2025-Aug-27 16:47:00] ProgressMeter -\tchr13:64098000\t2.3\n",
      "[PB Info 2025-Aug-27 16:47:06] ProgressMeter -\tchr15:22702000\t2.4\n",
      "[PB Info 2025-Aug-27 16:47:12] ProgressMeter -\tchr16:3077000\t2.5\n",
      "[PB Info 2025-Aug-27 16:47:18] ProgressMeter -\tchr17:61541000\t2.6\n",
      "[PB Info 2025-Aug-27 16:47:24] ProgressMeter -\tchr18:75760000\t2.7\n",
      "[PB Info 2025-Aug-27 16:47:30] ProgressMeter -\tchr20:33514000\t2.8\n",
      "[PB Info 2025-Aug-27 16:47:36] ProgressMeter -\tchr22:25506000\t2.9\n",
      "[PB Info 2025-Aug-27 16:47:43] Deepvariant is finished, total time is 181.009 seconds\n",
      "[PB Info 2025-Aug-27 16:47:43] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Aug-27 16:47:43] ||        Program:                                       deepvariant        ||\n",
      "[PB Info 2025-Aug-27 16:47:43] ||        Version:                                           4.0.1-1        ||\n",
      "[PB Info 2025-Aug-27 16:47:43] ||        Start Time:                       Wed Aug 27 16:44:31 2025        ||\n",
      "[PB Info 2025-Aug-27 16:47:43] ||        End Time:                         Wed Aug 27 16:47:43 2025        ||\n",
      "[PB Info 2025-Aug-27 16:47:43] ||        Total Time:                           3 minutes 12 seconds        ||\n",
      "[PB Info 2025-Aug-27 16:47:43] ------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "!pbrun deepvariant_germline \\\n",
    "    --ref Ref/Homo_sapiens_assembly38.fasta \\\n",
    "    --in-fq germline_data/sample_1.fq.gz germline_data/sample_2.fq.gz \\\n",
    "    --out-bam germline_output/dvgermline_out.bam \\\n",
    "    --out-variants germline_output/variants_dv.vcf \n",
    "    #--low-memory    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
