{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "18db57d9-6d77-4cce-b8f0-bacdd9b58ade",
   "metadata": {},
   "source": [
    "# Parabricks Hands-On Workshop"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1ecc79b-2629-4151-925c-55c953f34852",
   "metadata": {},
   "source": [
    "## Tumor Sequencing: Somatic Variant calling Workflow"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce2d9f58-9a5a-4c8f-abf6-a41bed770150",
   "metadata": {},
   "source": [
    "For tumor sequencing, there will be several analysis to perform:\n",
    "- Read alignment of tumor sequencing sample\n",
    "- Read alignment of normal sequencing sample (if doing tumor-normal paired analysis)\n",
    "- Somatic variant calling\n",
    "- RNA read alignment (for calling gene fusions from RNA-seq)\n",
    "- Gene fusion detection\n",
    "\n",
    "\\We will start from downloading the reference genome and sample files, and then proceed through the analysis step-by-step."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81d854c7-c48b-4e62-a843-eadcca5ab261",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### GPU Monitoring"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d5523e52-751a-4059-9751-f36883738c2f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fri Oct 11 03:09:37 2024       \n",
      "+---------------------------------------------------------------------------------------+\n",
      "| NVIDIA-SMI 535.54.03              Driver Version: 535.54.03    CUDA Version: 12.3     |\n",
      "|-----------------------------------------+----------------------+----------------------+\n",
      "| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |\n",
      "| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |\n",
      "|                                         |                      |               MIG M. |\n",
      "|=========================================+======================+======================|\n",
      "|   0  Tesla V100-SXM2-16GB-N         On  | 00000000:0A:00.0 Off |                    0 |\n",
      "| N/A   35C    P0              44W / 160W |      0MiB / 16384MiB |      0%      Default |\n",
      "|                                         |                      |                  N/A |\n",
      "+-----------------------------------------+----------------------+----------------------+\n",
      "                                                                                         \n",
      "+---------------------------------------------------------------------------------------+\n",
      "| Processes:                                                                            |\n",
      "|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |\n",
      "|        ID   ID                                                             Usage      |\n",
      "|=======================================================================================|\n",
      "|  No running processes found                                                           |\n",
      "+---------------------------------------------------------------------------------------+\n"
     ]
    }
   ],
   "source": [
    "!nvidia-smi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0374156-867b-4477-a3bf-1bfcd17fef31",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run the command below in the terminal\n",
    "### watch -n 0.5 nvidia-smi\n",
    "#"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "727be743-7a89-4242-83d3-3920884a7327",
   "metadata": {},
   "source": [
    "#### Download the Reference Genome"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af6a84dc-62c8-404a-b25d-d32c29c8e1b2",
   "metadata": {},
   "source": [
    "We will download the sample data from the germline workflow tutorial, which includes the reference genome and its index files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b939766-26a3-4d9f-93f3-3f6062c42627",
   "metadata": {},
   "outputs": [],
   "source": [
    "# The tar file is 9.3GB and, when extracted, an additional 14GB\n",
    "!mkdir sample_data\n",
    "%cd sample_data\n",
    "!wget -O parabricks_sample.tar.gz \"https://s3.amazonaws.com/parabricks.sample/parabricks_sample.tar.gz\"\n",
    "!tar xvf parabricks_sample.tar.gz\n",
    "!mv parabricks_sample/* .\n",
    "%cd .."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4449d1d7-09a9-459d-b71d-7ccc75037583",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Homo_sapiens_assembly38.dict\n",
      "Homo_sapiens_assembly38.fasta\n",
      "Homo_sapiens_assembly38.fasta.amb\n",
      "Homo_sapiens_assembly38.fasta.ann\n",
      "Homo_sapiens_assembly38.fasta.bwt\n",
      "Homo_sapiens_assembly38.fasta.fai\n",
      "Homo_sapiens_assembly38.fasta.pac\n",
      "Homo_sapiens_assembly38.fasta.sa\n",
      "Homo_sapiens_assembly38.known_indels.vcf.gz\n",
      "Homo_sapiens_assembly38.known_indels.vcf.gz.tbi\n"
     ]
    }
   ],
   "source": [
    "!ls sample_data/Ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3e865f3a-813c-4435-825b-0f022207f713",
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir outputdir"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4feacce-6f08-44b3-8b53-67b511bfcb9c",
   "metadata": {},
   "source": [
    "#### Download Sample Data Using SRAtoolkit"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f95dc733-bd77-4aa2-b40d-13d5ca529f9e",
   "metadata": {},
   "source": [
    "The SeqC2 consortium did a series of studies to o establish best practices, reference standards, and benchmark the results of somatic mutation detections under different bioinformatic and laboratory conditions (https://sites.google.com/view/seqc2 paper: https://www.nature.com/articles/s41587-021-00993-6). Download a tumor-normal paired sequencing data for analsis."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e374308-fa48-4108-9047-dc082b05f9ba",
   "metadata": {},
   "source": [
    "Download SRAtoolkit and download the sequencing files using it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b2429b4c-ffb2-41e5-8e30-564899d60fcf",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.10/dist-packages/IPython/core/magics/osm.py:417: UserWarning: This is now an optional IPython functionality, setting dhist requires you to install the `pickleshare` library.\n",
      "  self.shell.db['dhist'] = compress_dhist(dhist)[-100:]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/tutorial/sample_data\n",
      "--2024-10-16 02:32:48--  https://ftp-trace.ncbi.nlm.nih.gov/sra/sdk/current/sratoolkit.current-centos_linux64.tar.gz\n",
      "Resolving ftp-trace.ncbi.nlm.nih.gov (ftp-trace.ncbi.nlm.nih.gov)... 130.14.250.12, 130.14.250.7, 2607:f220:41e:250::10, ...\n",
      "Connecting to ftp-trace.ncbi.nlm.nih.gov (ftp-trace.ncbi.nlm.nih.gov)|130.14.250.12|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 93535881 (89M) [application/x-gzip]\n",
      "Saving to: ‘sratoolkit.tar.gz’\n",
      "\n",
      "sratoolkit.tar.gz   100%[===================>]  89.20M  33.6MB/s    in 2.7s    \n",
      "\n",
      "2024-10-16 02:32:51 (33.6 MB/s) - ‘sratoolkit.tar.gz’ saved [93535881/93535881]\n",
      "\n",
      "sratoolkit.3.1.1-centos_linux64/\n",
      "sratoolkit.3.1.1-centos_linux64/README.md\n",
      "sratoolkit.3.1.1-centos_linux64/README-vdb-config\n",
      "sratoolkit.3.1.1-centos_linux64/schema/\n",
      "sratoolkit.3.1.1-centos_linux64/schema/vdb/\n",
      "sratoolkit.3.1.1-centos_linux64/schema/vdb/vdb.vschema\n",
      "sratoolkit.3.1.1-centos_linux64/schema/vdb/built-in.vschema\n",
      "sratoolkit.3.1.1-centos_linux64/schema/insdc/\n",
      "sratoolkit.3.1.1-centos_linux64/schema/insdc/insdc.vschema\n",
      "sratoolkit.3.1.1-centos_linux64/schema/insdc/sra.vschema\n",
      "sratoolkit.3.1.1-centos_linux64/schema/insdc/seq.vschema\n",
      "sratoolkit.3.1.1-centos_linux64/schema/sra/\n",
      "sratoolkit.3.1.1-centos_linux64/schema/sra/abi.vschema\n",
      "sratoolkit.3.1.1-centos_linux64/schema/sra/pevents.vschema\n",
      "sratoolkit.3.1.1-centos_linux64/schema/sra/generic-fastq.vschema\n",
      "sratoolkit.3.1.1-centos_linux64/schema/sra/454.vschema\n",
      "sratoolkit.3.1.1-centos_linux64/schema/sra/pacbio.vschema\n",
      "sratoolkit.3.1.1-centos_linux64/schema/sra/illumina.vschema\n",
      "sratoolkit.3.1.1-centos_linux64/schema/sra/helicos.vschema\n",
      "sratoolkit.3.1.1-centos_linux64/schema/sra/nanopore.vschema\n",
      "sratoolkit.3.1.1-centos_linux64/schema/sra/ion-torrent.vschema\n",
      "sratoolkit.3.1.1-centos_linux64/schema/ncbi/\n",
      "sratoolkit.3.1.1-centos_linux64/schema/ncbi/clip.vschema\n",
      "sratoolkit.3.1.1-centos_linux64/schema/ncbi/seq-graph.vschema\n",
      "sratoolkit.3.1.1-centos_linux64/schema/ncbi/stats.vschema\n",
      "sratoolkit.3.1.1-centos_linux64/schema/ncbi/ncbi.vschema\n",
      "sratoolkit.3.1.1-centos_linux64/schema/ncbi/wgs-contig.vschema\n",
      "sratoolkit.3.1.1-centos_linux64/schema/ncbi/sra.vschema\n",
      "sratoolkit.3.1.1-centos_linux64/schema/ncbi/spotname.vschema\n",
      "sratoolkit.3.1.1-centos_linux64/schema/ncbi/pnbrdb.vschema\n",
      "sratoolkit.3.1.1-centos_linux64/schema/ncbi/varloc.vschema\n",
      "sratoolkit.3.1.1-centos_linux64/schema/ncbi/seq.vschema\n",
      "sratoolkit.3.1.1-centos_linux64/schema/ncbi/trace.vschema\n",
      "sratoolkit.3.1.1-centos_linux64/schema/align/\n",
      "sratoolkit.3.1.1-centos_linux64/schema/align/refseq.vschema\n",
      "sratoolkit.3.1.1-centos_linux64/schema/align/pileup-stats.vschema\n",
      "sratoolkit.3.1.1-centos_linux64/schema/align/mate-cache.vschema\n",
      "sratoolkit.3.1.1-centos_linux64/schema/align/seq.vschema\n",
      "sratoolkit.3.1.1-centos_linux64/schema/align/align.vschema\n",
      "sratoolkit.3.1.1-centos_linux64/schema/align/qstat.vschema\n",
      "sratoolkit.3.1.1-centos_linux64/schema/csra2/\n",
      "sratoolkit.3.1.1-centos_linux64/schema/csra2/stats.vschema\n",
      "sratoolkit.3.1.1-centos_linux64/schema/csra2/reference.vschema\n",
      "sratoolkit.3.1.1-centos_linux64/schema/csra2/read.vschema\n",
      "sratoolkit.3.1.1-centos_linux64/schema/csra2/csra2.vschema\n",
      "sratoolkit.3.1.1-centos_linux64/bin/\n",
      "sratoolkit.3.1.1-centos_linux64/bin/bam-load.3.1.1\n",
      "sratoolkit.3.1.1-centos_linux64/bin/sra-stat.3.1.1\n",
      "sratoolkit.3.1.1-centos_linux64/bin/bam-load.3\n",
      "sratoolkit.3.1.1-centos_linux64/bin/sra-search\n",
      "sratoolkit.3.1.1-centos_linux64/bin/srapath-orig.3.1.1\n",
      "sratoolkit.3.1.1-centos_linux64/bin/sra-search.3.1.1\n",
      "sratoolkit.3.1.1-centos_linux64/bin/prefetch.3\n",
      "sratoolkit.3.1.1-centos_linux64/bin/sam-dump.3\n",
      "sratoolkit.3.1.1-centos_linux64/bin/sff-dump.3\n",
      "sratoolkit.3.1.1-centos_linux64/bin/sff-dump\n",
      "sratoolkit.3.1.1-centos_linux64/bin/fastq-dump\n",
      "sratoolkit.3.1.1-centos_linux64/bin/fastq-load.3\n",
      "sratoolkit.3.1.1-centos_linux64/bin/srf-load.3.1.1\n",
      "sratoolkit.3.1.1-centos_linux64/bin/sra-sort.3\n",
      "sratoolkit.3.1.1-centos_linux64/bin/kdbmeta.3.1.1\n",
      "sratoolkit.3.1.1-centos_linux64/bin/sratools\n",
      "sratoolkit.3.1.1-centos_linux64/bin/cache-mgr.3\n",
      "sratoolkit.3.1.1-centos_linux64/bin/helicos-load.3.1.1\n",
      "sratoolkit.3.1.1-centos_linux64/bin/cache-mgr\n",
      "sratoolkit.3.1.1-centos_linux64/bin/srapath\n",
      "sratoolkit.3.1.1-centos_linux64/bin/vdb-config.3.1.1\n",
      "sratoolkit.3.1.1-centos_linux64/bin/kar.3\n",
      "sratoolkit.3.1.1-centos_linux64/bin/srf-load.3\n",
      "sratoolkit.3.1.1-centos_linux64/bin/illumina-dump\n",
      "sratoolkit.3.1.1-centos_linux64/bin/illumina-load.3.1.1\n",
      "sratoolkit.3.1.1-centos_linux64/bin/srapath.3\n",
      "sratoolkit.3.1.1-centos_linux64/bin/sam-dump\n",
      "sratoolkit.3.1.1-centos_linux64/bin/vdb-encrypt\n",
      "sratoolkit.3.1.1-centos_linux64/bin/sra-pileup.3.1.1\n",
      "sratoolkit.3.1.1-centos_linux64/bin/fastq-load.3.1.1\n",
      "sratoolkit.3.1.1-centos_linux64/bin/vdb-decrypt.3.1.1\n",
      "sratoolkit.3.1.1-centos_linux64/bin/rcexplain\n",
      "sratoolkit.3.1.1-centos_linux64/bin/prefetch-orig.3.1.1\n",
      "sratoolkit.3.1.1-centos_linux64/bin/vdb-decrypt\n",
      "sratoolkit.3.1.1-centos_linux64/bin/cache-mgr.3.1.1\n",
      "sratoolkit.3.1.1-centos_linux64/bin/illumina-dump.3.1.1\n",
      "sratoolkit.3.1.1-centos_linux64/bin/vdb-lock.3\n",
      "sratoolkit.3.1.1-centos_linux64/bin/vdb-lock\n",
      "sratoolkit.3.1.1-centos_linux64/bin/align-info.3.1.1\n",
      "sratoolkit.3.1.1-centos_linux64/bin/vdb-encrypt.3.1.1\n",
      "sratoolkit.3.1.1-centos_linux64/bin/ncbi/\n",
      "sratoolkit.3.1.1-centos_linux64/bin/ncbi/default.kfg\n",
      "sratoolkit.3.1.1-centos_linux64/bin/ncbi/certs.kfg\n",
      "sratoolkit.3.1.1-centos_linux64/bin/ncbi/vdb-copy.kfg\n",
      "sratoolkit.3.1.1-centos_linux64/bin/vdb-validate\n",
      "sratoolkit.3.1.1-centos_linux64/bin/rcexplain.3\n",
      "sratoolkit.3.1.1-centos_linux64/bin/vdb-copy.3\n",
      "sratoolkit.3.1.1-centos_linux64/bin/vdb-config\n",
      "sratoolkit.3.1.1-centos_linux64/bin/fastq-load\n",
      "sratoolkit.3.1.1-centos_linux64/bin/sra-pileup-orig.3.1.1\n",
      "sratoolkit.3.1.1-centos_linux64/bin/sff-load.3\n",
      "sratoolkit.3.1.1-centos_linux64/bin/kar.3.1.1\n",
      "sratoolkit.3.1.1-centos_linux64/bin/illumina-load.3\n",
      "sratoolkit.3.1.1-centos_linux64/bin/vdb-dump.3\n",
      "sratoolkit.3.1.1-centos_linux64/bin/test-sra.3\n",
      "sratoolkit.3.1.1-centos_linux64/bin/helicos-load\n",
      "sratoolkit.3.1.1-centos_linux64/bin/pacbio-load\n",
      "sratoolkit.3.1.1-centos_linux64/bin/helicos-load.3\n",
      "sratoolkit.3.1.1-centos_linux64/bin/fastq-dump-orig.3.1.1\n",
      "sratoolkit.3.1.1-centos_linux64/bin/abi-dump\n",
      "sratoolkit.3.1.1-centos_linux64/bin/test-sra\n",
      "sratoolkit.3.1.1-centos_linux64/bin/vdb-validate.3\n",
      "sratoolkit.3.1.1-centos_linux64/bin/abi-load.3.1.1\n",
      "sratoolkit.3.1.1-centos_linux64/bin/dump-ref-fasta.3.1.1\n",
      "sratoolkit.3.1.1-centos_linux64/bin/illumina-dump.3\n",
      "sratoolkit.3.1.1-centos_linux64/bin/pacbio-load.3.1.1\n",
      "sratoolkit.3.1.1-centos_linux64/bin/bam-load\n",
      "sratoolkit.3.1.1-centos_linux64/bin/sratools.3\n",
      "sratoolkit.3.1.1-centos_linux64/bin/test-sra.3.1.1\n",
      "sratoolkit.3.1.1-centos_linux64/bin/kdbmeta.3\n",
      "sratoolkit.3.1.1-centos_linux64/bin/sra-sort-cg\n",
      "sratoolkit.3.1.1-centos_linux64/bin/vdb-validate.3.1.1\n",
      "sratoolkit.3.1.1-centos_linux64/bin/fasterq-dump.3\n",
      "sratoolkit.3.1.1-centos_linux64/bin/latf-load\n",
      "sratoolkit.3.1.1-centos_linux64/bin/sra-sort.3.1.1\n",
      "sratoolkit.3.1.1-centos_linux64/bin/fastq-dump.3.1.1\n",
      "sratoolkit.3.1.1-centos_linux64/bin/abi-load.3\n",
      "sratoolkit.3.1.1-centos_linux64/bin/sam-dump-orig.3.1.1\n",
      "sratoolkit.3.1.1-centos_linux64/bin/srapath.3.1.1\n",
      "sratoolkit.3.1.1-centos_linux64/bin/prefetch.3.1.1\n",
      "sratoolkit.3.1.1-centos_linux64/bin/vdb-unlock\n",
      "sratoolkit.3.1.1-centos_linux64/bin/illumina-load\n",
      "sratoolkit.3.1.1-centos_linux64/bin/srf-load\n",
      "sratoolkit.3.1.1-centos_linux64/bin/vdb-dump.3.1.1\n",
      "sratoolkit.3.1.1-centos_linux64/bin/vdb-encrypt.3\n",
      "sratoolkit.3.1.1-centos_linux64/bin/sratools.3.1.1\n",
      "sratoolkit.3.1.1-centos_linux64/bin/sra-stat.3\n",
      "sratoolkit.3.1.1-centos_linux64/bin/sra-pileup.3\n",
      "sratoolkit.3.1.1-centos_linux64/bin/vdb-lock.3.1.1\n",
      "sratoolkit.3.1.1-centos_linux64/bin/align-info\n",
      "sratoolkit.3.1.1-centos_linux64/bin/sam-dump.3.1.1\n",
      "sratoolkit.3.1.1-centos_linux64/bin/sff-load.3.1.1\n",
      "sratoolkit.3.1.1-centos_linux64/bin/sra-search.3\n",
      "sratoolkit.3.1.1-centos_linux64/bin/abi-load\n",
      "sratoolkit.3.1.1-centos_linux64/bin/sra-sort-cg.3.1.1\n",
      "sratoolkit.3.1.1-centos_linux64/bin/cg-load.3.1.1\n",
      "sratoolkit.3.1.1-centos_linux64/bin/vdb-copy.3.1.1\n",
      "sratoolkit.3.1.1-centos_linux64/bin/vdb-decrypt.3\n",
      "sratoolkit.3.1.1-centos_linux64/bin/fastq-dump.3\n",
      "sratoolkit.3.1.1-centos_linux64/bin/dump-ref-fasta.3\n",
      "sratoolkit.3.1.1-centos_linux64/bin/sff-dump.3.1.1\n",
      "sratoolkit.3.1.1-centos_linux64/bin/kar\n",
      "sratoolkit.3.1.1-centos_linux64/bin/dump-ref-fasta\n",
      "sratoolkit.3.1.1-centos_linux64/bin/fasterq-dump.3.1.1\n",
      "sratoolkit.3.1.1-centos_linux64/bin/cg-load\n",
      "sratoolkit.3.1.1-centos_linux64/bin/vdb-copy\n",
      "sratoolkit.3.1.1-centos_linux64/bin/vdb-config.3\n",
      "sratoolkit.3.1.1-centos_linux64/bin/sra-sort-cg.3\n",
      "sratoolkit.3.1.1-centos_linux64/bin/sra-stat\n",
      "sratoolkit.3.1.1-centos_linux64/bin/abi-dump.3\n",
      "sratoolkit.3.1.1-centos_linux64/bin/abi-dump.3.1.1\n",
      "sratoolkit.3.1.1-centos_linux64/bin/rcexplain.3.1.1\n",
      "sratoolkit.3.1.1-centos_linux64/bin/align-info.3\n",
      "sratoolkit.3.1.1-centos_linux64/bin/sra-sort\n",
      "sratoolkit.3.1.1-centos_linux64/bin/vdb-unlock.3.1.1\n",
      "sratoolkit.3.1.1-centos_linux64/bin/kdbmeta\n",
      "sratoolkit.3.1.1-centos_linux64/bin/vdb-dump\n",
      "sratoolkit.3.1.1-centos_linux64/bin/sra-pileup\n",
      "sratoolkit.3.1.1-centos_linux64/bin/latf-load.3\n",
      "sratoolkit.3.1.1-centos_linux64/bin/cg-load.3\n",
      "sratoolkit.3.1.1-centos_linux64/bin/latf-load.3.1.1\n",
      "sratoolkit.3.1.1-centos_linux64/bin/prefetch\n",
      "sratoolkit.3.1.1-centos_linux64/bin/fasterq-dump-orig.3.1.1\n",
      "sratoolkit.3.1.1-centos_linux64/bin/fasterq-dump\n",
      "sratoolkit.3.1.1-centos_linux64/bin/pacbio-load.3\n",
      "sratoolkit.3.1.1-centos_linux64/bin/vdb-dump-orig.3.1.1\n",
      "sratoolkit.3.1.1-centos_linux64/bin/vdb-unlock.3\n",
      "sratoolkit.3.1.1-centos_linux64/bin/sff-load\n",
      "sratoolkit.3.1.1-centos_linux64/example/\n",
      "sratoolkit.3.1.1-centos_linux64/example/perl/\n",
      "sratoolkit.3.1.1-centos_linux64/example/perl/base-stats.pl\n",
      "sratoolkit.3.1.1-centos_linux64/example/perl/mismatch-stats.pl\n",
      "sratoolkit.3.1.1-centos_linux64/example/perl/gene-lookup.pl\n",
      "sratoolkit.3.1.1-centos_linux64/example/perl/simplefastq.pl\n",
      "sratoolkit.3.1.1-centos_linux64/example/perl/splitfastq.pl\n",
      "sratoolkit.3.1.1-centos_linux64/example/perl/dump-reference.pl\n",
      "sratoolkit.3.1.1-centos_linux64/example/perl/quality-stats.pl\n",
      "sratoolkit.3.1.1-centos_linux64/CHANGES\n",
      "sratoolkit.3.1.1-centos_linux64/README-blastn\n",
      "/usr/bin/sh: 1: source: not found\n",
      "/tutorial\n"
     ]
    }
   ],
   "source": [
    "## Download SRAtoolkit to process SRA files\n",
    "!wget --output-document sratoolkit.tar.gz https://ftp-trace.ncbi.nlm.nih.gov/sra/sdk/current/sratoolkit.current-centos_linux64.tar.gz\n",
    "!tar -vxzf sratoolkit.tar.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2863cdc-cc5c-4189-83df-85b3d833fab9",
   "metadata": {},
   "source": [
    "Add sratoolkit executable into the PATH for convenience. Check to make sure we are able to use the tool `fastq-dump`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "909ba91a-0769-4e0c-9570-dd606ec7ea16",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/tutorial/sample_data/sratoolkit.3.1.1-centos_linux64/bin/fastq-dump\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "sra_bin_path = os.popen(\"cd sample_data/sratoolkit.*/bin && pwd\").read().strip()\n",
    "os.environ[\"PATH\"] += os.pathsep + sra_bin_path\n",
    "!which fastq-dump"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75609797-24be-4238-9df7-0fbf68387e69",
   "metadata": {},
   "source": [
    "Check that `fastq-dump` works by running the following command. In a few seconds, you should see 10 lines of output."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "48938633-948d-4509-8335-c121c5323db1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read 2 spots for SRR390728\n",
      "Written 2 spots for SRR390728\n",
      "@SRR390728.1 1 length=72\n",
      "CATTCTTCACGTAGTTCTCGAGCCTTGGTTTTCAGCGATGGAGAATGACTTTGACAAGCTGAGAGAAGNTNC\n",
      "+SRR390728.1 1 length=72\n",
      ";;;;;;;;;;;;;;;;;;;;;;;;;;;9;;665142;;;;;;;;;;;;;;;;;;;;;;;;;;;;;96&&&&(\n",
      "@SRR390728.2 2 length=72\n",
      "AAGTAGGTCTCGTCTGTGTTTTCTACGAGCTTGTGTTCCAGCTGACCCACTCCCTGGGTGGGGGGACTGGGT\n",
      "+SRR390728.2 2 length=72\n",
      ";;;;;;;;;;;;;;;;;4;;;;3;393.1+4&&5&&;;;;;;;;;;;;;;;;;;;;;<9;<;;;;;464262\n"
     ]
    }
   ],
   "source": [
    "#Check that sratoolkit is properly installed and fastq-dump works\n",
    "!fastq-dump --stdout -X 2 SRR390728"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35b51f53-63c5-4b26-8840-59b5ace25603",
   "metadata": {},
   "source": [
    "Download sequencing files. SRR7890899 is a whole-exome sequencing for the tumor breast cell line HCC1395. SRR7890845 is a whole-exome sequencing for the matched normal HCC1395BL B lymphocyte cell line. These files are quite large and took hours to download."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "62036e4a-ebec-4a7e-86f3-e445f2a7103e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read 57058265 spots for SRR7890845\n",
      "Written 57058265 spots for SRR7890845\n"
     ]
    }
   ],
   "source": [
    "%cd sample_data/Data\n",
    "!fastq-dump --split-files --gzip SRR7890845\n",
    "!fastq-dump --split-files --gzip SRR7890899\n",
    "%cd ../.."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "321fd6c4-b73e-4319-9777-8426ea5d8395",
   "metadata": {},
   "source": [
    "Download whole-genome sequencing files of a tumor and its matched normal sample. The files are large and took a long time to download."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae860e26-d387-4585-b41e-dacbed8814c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Download publicly available SRA files using wget. Both files are\n",
    "# about 65 GB in size.\n",
    "%cd sample_data/Data\n",
    "# Normal sample\n",
    "!wget https://sra-pub-run-odp.s3.amazonaws.com/sra/SRR7890827/SRR7890827 --output-document=SRR7890827.sra\n",
    "\n",
    "# Tumor sample\n",
    "!wget https://sra-pub-run-odp.s3.amazonaws.com/sra/SRR7890824/SRR7890824 --output-document=SRR7890824.sra\n",
    "\n",
    "## Convert SRA to FASTQ files\n",
    "!fastq-dump --split-files ./SRR7890827.sra --gzip\n",
    "!fastq-dump --split-files ./SRR7890824.sra --gzip\n",
    "%cd ../.."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1259ba08-43f9-49e9-9f5d-68c5699be8d2",
   "metadata": {},
   "source": [
    "#### Run Somatic Workflow: BWA-MEM + Mutect2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7fd89790-2a95-4b4a-b00d-8e2657633475",
   "metadata": {},
   "source": [
    "The Somatic Workflow is one command that analyzes tumor sequencing from raw FASTQ file to somatic variants in VCF format. It performs alignment using BWA-MEM on both tumor and normal sequencing files and uses Mutect2 to generate a single VCF file containing somatic variants. If you wish to perform tumor-only analysis, just omit the `in-normal-fq` and `out-normal-bam` options."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "7ca5c7a7-fe01-46c9-b204-0a14c5a13d9b",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please visit https://docs.nvidia.com/clara/#parabricks for detailed documentation\n",
      "\n",
      "\n",
      "[Parabricks Options Mesg]: Read group created for /tutorial/sample_data/Data/SRR7890899_1.fastq.gz and\n",
      "/tutorial/sample_data/Data/SRR7890899_2.fastq.gz\n",
      "[Parabricks Options Mesg]: @RG\\tID:SRR7890899_rg1\\tLB:lib_tumor\\tPL:ILLUMINA\\tSM:tumor_sample\\tPU:SRR7890899_rg1\n",
      "[Parabricks Options Mesg]: Read group created for /tutorial/sample_data/Data/SRR7890845_1.fastq.gz and\n",
      "/tutorial/sample_data/Data/SRR7890845_2.fastq.gz\n",
      "[Parabricks Options Mesg]: @RG\\tID:SRR7890845_rg1\\tLB:lib_normal\\tPL:ILLUMINA\\tSM:normal_sample\\tPU:SRR7890845_rg1\n",
      "\n",
      "\n",
      "[Parabricks Options Mesg]: Checking argument compatibility\n",
      "[Parabricks Options Mesg]: Set --bwa-options=\"-K #\" to produce compatible pair-ended results with previous versions of\n",
      "fq2bam or BWA MEM.\n",
      "[Parabricks Options Mesg]: Read group created for /tutorial/sample_data/Data/SRR7890899_1.fastq.gz and\n",
      "/tutorial/sample_data/Data/SRR7890899_2.fastq.gz\n",
      "[Parabricks Options Mesg]: @RG\\tID:SRR7890899_rg1\\tLB:lib_tumor\\tPL:ILLUMINA\\tSM:tumor_sample\\tPU:SRR7890899_rg1\n",
      "[Parabricks Options Mesg]: Using --low-memory sets the number of streams in bwa mem to 1.\n",
      "[PB Info 2024-Oct-20 00:33:42] ------------------------------------------------------------------------------\n",
      "[PB Info 2024-Oct-20 00:33:42] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2024-Oct-20 00:33:42] ||                              Version 4.3.2-1                             ||\n",
      "[PB Info 2024-Oct-20 00:33:42] ||                      GPU-PBBWA mem, Sorting Phase-I                      ||\n",
      "[PB Info 2024-Oct-20 00:33:42] ------------------------------------------------------------------------------\n",
      "[PB Info 2024-Oct-20 00:33:42] Mode = pair-ended-gpu\n",
      "[PB Info 2024-Oct-20 00:33:42] Running with 1 GPU(s), using 1 stream(s) per device with 16 worker threads per GPU\n",
      "[PB Info 2024-Oct-20 00:33:52] #  0  0  0  0  0   0 pool:  0 0 bases/GPU/minute: 0.0 \n",
      "[PB Info 2024-Oct-20 00:34:02] # 94  0  1  0  0   0 pool:  0 0 bases/GPU/minute: 0.0 \n",
      "[PB Info 2024-Oct-20 00:34:12] # 100  0  0  0  0   0 pool:  1 108727500 bases/GPU/minute: 652365000.0 \n",
      "[PB Info 2024-Oct-20 00:34:22] # 100  0  1  0  0   0 pool:  0 237284250 bases/GPU/minute: 771340500.0 \n",
      "[PB Info 2024-Oct-20 00:34:32] # 100  0  1  0  0   0 pool:  0 355945350 bases/GPU/minute: 711966600.0 \n",
      "[PB Info 2024-Oct-20 00:34:42] # 100  0  1  0  0   0 pool:  0 464710650 bases/GPU/minute: 652591800.0 \n",
      "[PB Info 2024-Oct-20 00:34:52] # 100  0  1  0  0   0 pool:  0 593249400 bases/GPU/minute: 771232500.0 \n",
      "[PB Info 2024-Oct-20 00:35:02] # 100  0  1  0  0   0 pool:  0 711893250 bases/GPU/minute: 711863100.0 \n",
      "[PB Info 2024-Oct-20 00:35:12] # 100  0  1  0  0   0 pool:  0 830553450 bases/GPU/minute: 711961200.0 \n",
      "[PB Info 2024-Oct-20 00:35:22] # 100  0  1  0  0   0 pool:  0 949214550 bases/GPU/minute: 711966600.0 \n",
      "[PB Info 2024-Oct-20 00:35:32] # 100  0  0  0  0   0 pool:  1 1067888550 bases/GPU/minute: 712044000.0 \n",
      "[PB Info 2024-Oct-20 00:35:42] # 100  0  1  0  0   0 pool:  0 1166772750 bases/GPU/minute: 593305200.0 \n",
      "[PB Info 2024-Oct-20 00:35:52] # 100  0  1  0  0   0 pool:  1 1265652450 bases/GPU/minute: 593278200.0 \n",
      "[PB Info 2024-Oct-20 00:36:02] # 100  0  1  0  0   0 pool:  0 1344760950 bases/GPU/minute: 474651000.0 \n",
      "[PB Info 2024-Oct-20 00:36:12] # 100  0  1  0  0   0 pool:  0 1443648000 bases/GPU/minute: 593322300.0 \n",
      "[PB Info 2024-Oct-20 00:36:22] # 100  0  1  0  0   0 pool:  0 1562303850 bases/GPU/minute: 711935100.0 \n",
      "[PB Info 2024-Oct-20 00:36:32] # 100  0  1  0  0   0 pool:  1 1690836750 bases/GPU/minute: 771197400.0 \n",
      "[PB Info 2024-Oct-20 00:36:42] # 100  0  1  0  0   0 pool:  0 1809531000 bases/GPU/minute: 712165500.0 \n",
      "[PB Info 2024-Oct-20 00:36:52] # 100  0  1  0  0   0 pool:  1 1928205600 bases/GPU/minute: 712047600.0 \n",
      "[PB Info 2024-Oct-20 00:37:02] # 100  0  1  0  0   0 pool:  0 2056738950 bases/GPU/minute: 771200100.0 \n",
      "[PB Info 2024-Oct-20 00:37:12] # 100  0  1  0  0   0 pool:  0 2165515500 bases/GPU/minute: 652659300.0 \n",
      "[PB Info 2024-Oct-20 00:37:22] # 100  0  1  0  0   0 pool:  0 2284165200 bases/GPU/minute: 711898200.0 \n",
      "[PB Info 2024-Oct-20 00:37:32] # 100  0  1  0  0   0 pool:  0 2392953300 bases/GPU/minute: 652728600.0 \n",
      "[PB Info 2024-Oct-20 00:37:42] # 100  0  1  0  0   0 pool:  1 2501737650 bases/GPU/minute: 652706100.0 \n",
      "[PB Info 2024-Oct-20 00:37:52] # 100  0  1  0  0   0 pool:  0 2590719150 bases/GPU/minute: 533889000.0 \n",
      "[PB Info 2024-Oct-20 00:38:02] # 100  0  1  0  0   0 pool:  1 2729170050 bases/GPU/minute: 830705400.0 \n",
      "[PB Info 2024-Oct-20 00:38:12] # 100  0  1  0  0   0 pool:  0 2857721700 bases/GPU/minute: 771309900.0 \n",
      "[PB Info 2024-Oct-20 00:38:22] # 100  0  1  0  0   0 pool:  0 2966481600 bases/GPU/minute: 652559400.0 \n",
      "[PB Info 2024-Oct-20 00:38:32] # 100  0  1  0  0   0 pool:  0 3085123800 bases/GPU/minute: 711853200.0 \n",
      "[PB Info 2024-Oct-20 00:38:42] # 100  0  1  0  0   0 pool:  0 3203794350 bases/GPU/minute: 712023300.0 \n",
      "[PB Info 2024-Oct-20 00:38:52] # 100  0  1  0  0   0 pool:  0 3332358000 bases/GPU/minute: 771381900.0 \n",
      "[PB Info 2024-Oct-20 00:39:02] # 100  0  1  0  0   0 pool:  0 3470792250 bases/GPU/minute: 830605500.0 \n",
      "[PB Info 2024-Oct-20 00:39:12] # 100  0  1  0  0   0 pool:  0 3589459650 bases/GPU/minute: 712004400.0 \n",
      "[PB Info 2024-Oct-20 00:39:22] # 100  0  1  0  0   0 pool:  1 3718012950 bases/GPU/minute: 771319800.0 \n",
      "[PB Info 2024-Oct-20 00:39:32] # 100  0  1  0  0   0 pool:  0 3836701950 bases/GPU/minute: 712134000.0 \n",
      "[PB Info 2024-Oct-20 00:39:42] # 100  0  1  0  0   0 pool:  1 3955380150 bases/GPU/minute: 712069200.0 \n",
      "[PB Info 2024-Oct-20 00:39:52] # 100  0  1  0  0   0 pool:  0 4083924150 bases/GPU/minute: 771264000.0 \n",
      "[PB Info 2024-Oct-20 00:40:02] # 100  0  1  0  0   0 pool:  0 4182826350 bases/GPU/minute: 593413200.0 \n",
      "[PB Info 2024-Oct-20 00:40:12] # 100  1  0  0  0   0 pool:  0 4301491950 bases/GPU/minute: 711993600.0 \n",
      "[PB Info 2024-Oct-20 00:40:22] # 100  0  1  0  0   0 pool:  1 4430085000 bases/GPU/minute: 771558300.0 \n",
      "[PB Info 2024-Oct-20 00:40:32] # 100  0  1  0  0   0 pool:  0 4528977600 bases/GPU/minute: 593355600.0 \n",
      "[PB Info 2024-Oct-20 00:40:42] # 100  0  1  0  0   0 pool:  0 4637784750 bases/GPU/minute: 652842900.0 \n",
      "[PB Info 2024-Oct-20 00:40:52] # 100  0  1  0  0   0 pool:  0 4776257550 bases/GPU/minute: 830836800.0 \n",
      "[PB Info 2024-Oct-20 00:41:02] # 100  0  1  0  0   0 pool:  0 4904874000 bases/GPU/minute: 771698700.0 \n",
      "[PB \u001b[31mError\u001b[0m 2024-Oct-20 00:41:05][src/PBParallelFastqFile.cpp:250] Input files/tutorial/sample_data/Data/SRR7890899_1.fastq.gz and /tutorial/sample_data/Data/SRR7890899_2.fastq.gz have different sizes, exiting.\n",
      "For technical support visit https://docs.nvidia.com/clara/index.html#parabricks\n",
      "Exiting...\n",
      "\n",
      "Could not run fq2bam as part of tumor sample processing for somatic pipeline\n",
      "Exiting pbrun ...\n"
     ]
    }
   ],
   "source": [
    "!pbrun somatic \\\n",
    "    --ref sample_data/Ref/Homo_sapiens_assembly38.fasta \\\n",
    "    --in-normal-fq sample_data/Data/SRR7890845_1.fastq.gz sample_data/Data/SRR7890845_2.fastq.gz \"@RG\\tID:SRR7890845_rg1\\tLB:lib_normal\\tPL:ILLUMINA\\tSM:normal_sample\\tPU:SRR7890845_rg1\" \\\n",
    "    --in-tumor-fq sample_data/Data/SRR7890899_1.fastq.gz sample_data/Data/SRR7890899_2.fastq.gz \"@RG\\tID:SRR7890899_rg1\\tLB:lib_tumor\\tPL:ILLUMINA\\tSM:tumor_sample\\tPU:SRR7890899_rg1\" \\\n",
    "    --bwa-options=\"-Y\" \\\n",
    "    --out-normal-bam outputdir/normal.bam \\\n",
    "    --out-tumor-bam outputdir/tumor.bam \\\n",
    "    --out-vcf outputdir/somatic_variants.vcf \\\n",
    "    --low-memory \\\n",
    "    --mutect-low-memory \\\n",
    "    --num-gpus 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "715c6c42-04d8-4bcb-928f-2fdaed38a05f",
   "metadata": {},
   "source": [
    "#### Run Alignment and Variant Calling Separately"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40c46eab-af0c-47cc-86a7-e06e65b8acc1",
   "metadata": {},
   "source": [
    "The alignment and variant calling steps within the somatic workflow can be run separately. Remember to run fq2bam on the tumor fastq and normal fastqa separately. Variant calling can be done by Mutext2 or DeepSomatic."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2bcca71a-cde0-4759-bfad-063a44748f99",
   "metadata": {},
   "source": [
    "- Mutect2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c85dc5c5-98b6-4cf2-b310-be6609390757",
   "metadata": {},
   "outputs": [],
   "source": [
    "!pbrun mutectcaller \\\n",
    "    --ref sample_data/Ref/Homo_sapiens_assembly38.fasta \\\n",
    "    --in-normal-bam outputdir/normal.bam \\\n",
    "    --in-tumor-bam outputdir/tumor.bam \\\n",
    "    --out-vcf outputdir/somatic_variants.vcf \\\n",
    "    --mutect-low-memory \\\n",
    "    --num-gpus 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45ad2c10-46e5-4c34-9589-cd747228caaf",
   "metadata": {},
   "source": [
    "- DeepSomatics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "373d3b43-04cd-48c9-a956-54a2f48e183b",
   "metadata": {},
   "outputs": [],
   "source": [
    "!pbrun deepsomatic \\\n",
    "    --ref sample_data/Ref/Homo_sapiens_assembly38.fasta \\\n",
    "    --in-normal-bam outputdir/normal.bam \\\n",
    "    --in-tumor-bam outputdir/tumor.bam \\\n",
    "    --out-vcf outputdir/somatic_variants.vcf \\"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe307434-c556-4054-ac0a-79ea808906c2",
   "metadata": {},
   "source": [
    "## Gene Fusion Detection"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a450b33-9516-4164-88e9-c09226b628a6",
   "metadata": {},
   "source": [
    "#### Download Sample Dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2344227e-e2dc-47b8-8106-c0d9f5fb7764",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2024-10-18 07:02:06--  https://zenodo.org/records/13363154/files/BT474.Left.fq.gz\n",
      "Resolving zenodo.org (zenodo.org)... 188.184.103.159, 188.185.79.172, 188.184.98.238, ...\n",
      "Connecting to zenodo.org (zenodo.org)|188.184.103.159|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 1067370758 (1018M) [application/octet-stream]\n",
      "Saving to: ‘BT474.Left.fq.gz’\n",
      "\n",
      "BT474.Left.fq.gz    100%[===================>]   1018M  17.0MB/s    in 57s     \n",
      "\n",
      "2024-10-18 07:03:04 (18.0 MB/s) - ‘BT474.Left.fq.gz’ saved [1067370758/1067370758]\n",
      "\n",
      "--2024-10-18 07:03:04--  https://zenodo.org/records/13363154/files/BT474.Right.fq.gz\n",
      "Resolving zenodo.org (zenodo.org)... 188.185.79.172, 188.184.98.238, 188.184.103.159, ...\n",
      "Connecting to zenodo.org (zenodo.org)|188.185.79.172|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 1078442985 (1.0G) [application/octet-stream]\n",
      "Saving to: ‘BT474.Right.fq.gz’\n",
      "\n",
      "BT474.Right.fq.gz   100%[===================>]   1.00G  18.4MB/s    in 57s     \n",
      "\n",
      "2024-10-18 07:04:02 (17.9 MB/s) - ‘BT474.Right.fq.gz’ saved [1078442985/1078442985]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "## Download STAR-Fusion benchmark dataset breast cancer cell line  BT474\n",
    "%cd sample_data/Data\n",
    "!wget https://zenodo.org/records/13363154/files/BT474.Left.fq.gz\n",
    "!wget https://zenodo.org/records/13363154/files/BT474.Right.fq.gz\n",
    "%cd ../.."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "a19b943a-9091-4b9a-bcc6-b7339e8a981d",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2024-10-19 08:03:24--  https://zenodo.org/records/13363154/files/SRR1217085_1.fastq.gz.20M.fq.gz\n",
      "Resolving zenodo.org (zenodo.org)... 188.184.103.159, 188.185.79.172, 188.184.98.238, ...\n",
      "Connecting to zenodo.org (zenodo.org)|188.184.103.159|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 2158029851 (2.0G) [application/octet-stream]\n",
      "Saving to: ‘SRR1217085_1.fastq.gz.20M.fq.gz’\n",
      "\n",
      "SRR1217085_1.fastq. 100%[===================>]   2.01G  18.2MB/s    in 1m 56s  \n",
      "\n",
      "2024-10-19 08:05:21 (17.8 MB/s) - ‘SRR1217085_1.fastq.gz.20M.fq.gz’ saved [2158029851/2158029851]\n",
      "\n",
      "--2024-10-19 08:05:21--  https://zenodo.org/records/13363154/files/SRR1217085_2.fastq.gz.20M.fq.gz\n",
      "Resolving zenodo.org (zenodo.org)... 188.185.79.172, 188.184.103.159, 188.184.98.238, ...\n",
      "Connecting to zenodo.org (zenodo.org)|188.185.79.172|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 2048873466 (1.9G) [application/octet-stream]\n",
      "Saving to: ‘SRR1217085_2.fastq.gz.20M.fq.gz’\n",
      "\n",
      "SRR1217085_2.fastq. 100%[===================>]   1.91G  18.7MB/s    in 1m 48s  \n",
      "\n",
      "2024-10-19 08:07:10 (18.1 MB/s) - ‘SRR1217085_2.fastq.gz.20M.fq.gz’ saved [2048873466/2048873466]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "## Download STAR-Fusion benchmark dataset prostate cancer cell line VCaP\n",
    "%cd sample_data/Data\n",
    "\n",
    "!wget https://zenodo.org/records/13363154/files/SRR1217085_1.fastq.gz.20M.fq.gz\n",
    "!wget https://zenodo.org/records/13363154/files/SRR1217085_2.fastq.gz.20M.fq.gz\n",
    "%cd ../.."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da028f97-7869-4a19-bad5-b4e5b590b5b5",
   "metadata": {},
   "source": [
    "#### Prepare the Reference Genome Index File"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "675ee842-1a06-482a-97c4-7e1a2c00cd60",
   "metadata": {},
   "source": [
    "Before using Parabricks to accelerate STAR-Fusion for gene fusion detection, the reference genome index file needs to be generated using STAR v2.7.2a run on CPU."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8fa0f883-3fa2-45f5-b0a9-16d25f643b4c",
   "metadata": {},
   "source": [
    "##### Download the CTAT Genome Library"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "7336c81a-424b-40f6-b59b-431551ab155f",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/usr/bin/sh: 1: cd: can't cd to sample/Ref\n",
      "--2024-10-18 09:34:02--  https://data.broadinstitute.org/Trinity/CTAT_RESOURCE_LIB/__genome_libs_StarFv1.10/GRCh38_gencode_v37_CTAT_lib_Mar012021.source.tar.gz\n",
      "Resolving data.broadinstitute.org (data.broadinstitute.org)... 69.173.68.137\n",
      "Connecting to data.broadinstitute.org (data.broadinstitute.org)|69.173.68.137|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 4184819390 (3.9G) [application/x-gzip]\n",
      "Saving to: ‘GRCh38_gencode_v37_CTAT_lib_Mar012021.source.tar.gz’\n",
      "\n",
      "GRCh38_gencode_v37_ 100%[===================>]   3.90G  43.4MB/s    in 92s     \n",
      "\n",
      "2024-10-18 09:35:35 (43.3 MB/s) - ‘GRCh38_gencode_v37_CTAT_lib_Mar012021.source.tar.gz’ saved [4184819390/4184819390]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Download the ctat genome library\n",
    "%cd sample/Ref\n",
    "!wget https://data.broadinstitute.org/Trinity/CTAT_RESOURCE_LIB/__genome_libs_StarFv1.10/GRCh38_gencode_v37_CTAT_lib_Mar012021.source.tar.gz\n",
    "%cd ../.."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "495c5f49-5f89-4579-9dfb-c7232867c03b",
   "metadata": {},
   "source": [
    "Unpack the CTAT genome library and replace the `AnnotFilterRule.pm` file as instructed. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "23e16673-10d9-4737-b6f6-b370c3822f92",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/tutorial/sample_data/Ref\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.10/dist-packages/IPython/core/magics/osm.py:417: UserWarning: This is now an optional IPython functionality, setting dhist requires you to install the `pickleshare` library.\n",
      "  self.shell.db['dhist'] = compress_dhist(dhist)[-100:]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/fusion_lib.Mar2021.dat.gz\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/GRCh38.primary_assembly.genome.fa.pseudo_masked.fa\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__loc_chkpts/\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__loc_chkpts/customized_gtf.ok\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__loc_chkpts/ref_annot.cdsplus.fa.ok\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__loc_chkpts/pseudo_mask_genome.ok\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__loc_chkpts/ref_annot.cdna.fa.allvsall.blastn.outfmt6.toGenes.ok\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__loc_chkpts/ref_annot.cdsplus.dfam_masked.fa.ok\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__loc_chkpts/ref_annot.cdsplus.dfam_masked.fa.blidx.ok\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__loc_chkpts/ref_annot.cdsplus.dfam_masked.fa.allvsall.outfmt6.genesym.best.gz.ok\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__loc_chkpts/ref_annot.cdsplus.dfam_masked.fa.allvsall.outfmt6.genesym.best.overlaps_filt.ok\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__loc_chkpts/hmmscan.ok\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__loc_chkpts/ref_annot.cdsplus.dfam_masked.fa.allvsall.outfmt6.genesym.gz.ok\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__loc_chkpts/ref_annot.cdsplus.dfam_masked.fa.allvsall.outfmt6.ok\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__loc_chkpts/ref_annot.cdna.fa.ok\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__loc_chkpts/ref_annot.cdna.fa.blastn.allvsall.outfmt6.ok\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__loc_chkpts/ref_annot.cdna.fa.allvsall.blastn.outfmt6.toGenes.sorted.ok\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__loc_chkpts/ref_annot.cdna.fa.allvsall.blastn.outfmt6.toGenes.sorted.gzip.ok\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__loc_chkpts/gzip_pfam.ok\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__loc_chkpts/make_pep_file.ok\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__loc_chkpts/ref_annot.cdna.fa.blastidx.ok\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__loc_chkpts/revised_gencode_annots.ok\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/para.mask.list\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/_pfam_db_prep_chckpts/\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/_pfam_db_prep_chckpts/pipeliner.55865.cmds\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/_pfam_db_prep_chckpts/pfam_wget.ok\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/_pfam_db_prep_chckpts/pfam_hmmpress.ok\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/_pfam_db_prep_chckpts/pfam_gunzip.ok\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/_pfam_db_prep_chckpts/pipeliner.23423.cmds\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/GRCh38.primary_assembly.genome.fa\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/ref_annot.cdna.fa\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/ref_annot.cdna.fa.allvsall.outfmt6.toGenes.sorted.gz\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/gencode.v37.annotation.gtf.revised.custom.gtf\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__gencode_refinement_chkpts/\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__gencode_refinement_chkpts/IGH.superlocus.gtf\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__gencode_refinement_chkpts/pipeliner.25985.cmds\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__gencode_refinement_chkpts/remove_readthrus.ok\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__gencode_refinement_chkpts/IGL.superlocus.gtf\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__gencode_refinement_chkpts/IGH_locus.gtf\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__gencode_refinement_chkpts/feature_selection.ok\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__gencode_refinement_chkpts/cp_refined_gtf_to_out.ok\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__gencode_refinement_chkpts/igh_superlocus.revcomp.ok\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__gencode_refinement_chkpts/igl_locus.ok\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__gencode_refinement_chkpts/IGH.superlocus.revcomp.gtf\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__gencode_refinement_chkpts/IGL_locus.gtf\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__gencode_refinement_chkpts/igl_superlocus.ok\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__gencode_refinement_chkpts/igh_locus.ok\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__gencode_refinement_chkpts/IGL.superlocus.revcomp.gtf\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__gencode_refinement_chkpts/igh_superlocus.ok\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__gencode_refinement_chkpts/refined_gtf.ok\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__gencode_refinement_chkpts/igl_superlocus.revcomp.ok\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/gencode.v37.annotation.gtf\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/ref_annot.cdsplus.fa\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/PFAM.domtblout.dat.gz\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__dfam_ref_annot.cdsplus.fa/\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__dfam_ref_annot.cdsplus.fa/dfam.out\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__dfam_ref_annot.cdsplus.fa/pipeliner.49979.cmds\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__dfam_ref_annot.cdsplus.fa/pipeliner.69026.cmds\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__dfam_ref_annot.cdsplus.fa/dfamscan.ok\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/_dfam_db_prep_chckpts/\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/_dfam_db_prep_chckpts/dfam.h3f.ok\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/_dfam_db_prep_chckpts/pipeliner.55865.cmds\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/_dfam_db_prep_chckpts/dfam..ok\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/_dfam_db_prep_chckpts/dfam.h3i.ok\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/_dfam_db_prep_chckpts/dfam.h3p.ok\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/_dfam_db_prep_chckpts/dfam.h3m.ok\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/ref_annot.cdsplus.dfam_masked.fa.allvsall.outfmt6.genesym.best.overlaps_filt.gz\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/ref_annot.pep.fa\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/build_ctat_lib.sh\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/ref_annot.cdsplus.dfam_masked.fa\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/gencode.v37.annotation.gtf.revised.gtf\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__paralogs/\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__paralogs/__chckpts/\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__paralogs/__chckpts/blast_confounders.ok\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__paralogs/__chckpts/pipeliner.27496.cmds\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__paralogs/pipeliner.27490.cmds\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__paralogs/confounders.blastn.outfmt6\n",
      "GRCh38_gencode_v37_CTAT_lib_Mar012021.source/__paralogs/define_para_mask.ok\n",
      "/tutorial/sample_data/Ref/GRCh38_gencode_v37_CTAT_lib_Mar012021.source\n",
      "--2024-10-19 15:45:11--  https://data.broadinstitute.org/Trinity/CTAT_RESOURCE_LIB/__genome_libs_StarFv1.10/AnnotFilterRule.pm\n",
      "Resolving data.broadinstitute.org (data.broadinstitute.org)... 69.173.68.137\n",
      "Connecting to data.broadinstitute.org (data.broadinstitute.org)|69.173.68.137|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 3541 (3.5K) [text/plain]\n",
      "Saving to: ‘AnnotFilterRule.pm’\n",
      "\n",
      "AnnotFilterRule.pm  100%[===================>]   3.46K  --.-KB/s    in 0s      \n",
      "\n",
      "2024-10-19 15:45:12 (936 MB/s) - ‘AnnotFilterRule.pm’ saved [3541/3541]\n",
      "\n",
      "/tutorial\n"
     ]
    }
   ],
   "source": [
    "%cd sample_data/Ref\n",
    "!tar xvf GRCh38_gencode_v37_CTAT_lib_Mar012021.source.tar.gz \n",
    "%cd GRCh38_gencode_v37_CTAT_lib_Mar012021.source\n",
    "!wget https://data.broadinstitute.org/Trinity/CTAT_RESOURCE_LIB/__genome_libs_StarFv1.10/AnnotFilterRule.pm\n",
    "%cd ../../.."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a47875c-09bf-4892-8931-73797b004db3",
   "metadata": {},
   "source": [
    "##### Download STAR "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec465b79-9e40-40c1-b55c-8a47ca91eed3",
   "metadata": {},
   "source": [
    "Parabricks has been tested to be compatible with STAR 2.7.2a, so we will download this version of STAR to generate the reference genome index file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "c8a6ec71-34c8-4af0-a4b6-147e13e43a13",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2024-10-19 09:18:28--  https://github.com/alexdobin/STAR/archive/2.7.2a.tar.gz\n",
      "Resolving github.com (github.com)... 140.82.113.4\n",
      "Connecting to github.com (github.com)|140.82.113.4|:443... connected.\n",
      "HTTP request sent, awaiting response... 302 Found\n",
      "Location: https://codeload.github.com/alexdobin/STAR/tar.gz/refs/tags/2.7.2a [following]\n",
      "--2024-10-19 09:18:28--  https://codeload.github.com/alexdobin/STAR/tar.gz/refs/tags/2.7.2a\n",
      "Resolving codeload.github.com (codeload.github.com)... 140.82.113.10\n",
      "Connecting to codeload.github.com (codeload.github.com)|140.82.113.10|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: unspecified [application/x-gzip]\n",
      "Saving to: ‘2.7.2a.tar.gz’\n",
      "\n",
      "2.7.2a.tar.gz           [      <=>           ]   7.76M  6.99MB/s    in 1.1s    \n",
      "\n",
      "2024-10-19 09:18:30 (6.99 MB/s) - ‘2.7.2a.tar.gz’ saved [8142240]\n",
      "\n",
      "make: *** No rule to make target 'STAR'.  Stop.\n"
     ]
    }
   ],
   "source": [
    "# Download STAR to make the genome index file for rna_fq2bam\n",
    "!wget https://github.com/alexdobin/STAR/archive/2.7.2a.tar.gz\n",
    "!tar -xzf 2.7.2a.tar.gz\n",
    "%cd STAR-2.7.2a/source\n",
    "!make STAR"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "05aecd9a-e527-48d1-a928-aa8b37bc6d2e",
   "metadata": {},
   "source": [
    "Once STAR is installed, set the PATH environmental variable for convenience."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "15255655-a578-410d-afe6-7d3884f5aeef",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/usr/local/parabricks:/usr/local/nvidia/bin:/usr/local/cuda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/tutorial/sample_data/sratoolkit.3.1.1-centos_linux64/bin:/tutorial/STAR-2.7.2a/source\n"
     ]
    }
   ],
   "source": [
    "STAR_path = os.popen(\"cd STAR-2.7.2a/source && pwd\").read().strip()\n",
    "os.environ[\"PATH\"] += os.pathsep + STAR_path\n",
    "!echo $PATH"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "661e4467-9473-4027-b820-8f27f4869027",
   "metadata": {},
   "source": [
    "##### Generate Reference Genome Index"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73026a98-b6c3-482a-b080-041837f9dce4",
   "metadata": {},
   "source": [
    "Run STAR 2.7.2a to generate the reference genome index file. The read length of SRR1217085 is 99, so `--sjdbOverhang `is set to 100 (read length plus one). This step is time consuming and took 40 min on my machine (80 CPU cores, 520 GB RAM)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "ac703368-a0d0-465a-84be-218f20934e34",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage: STAR  [options]... --genomeDir REFERENCE   --readFilesIn R1.fq R2.fq\n",
      "Spliced Transcripts Alignment to a Reference (c) Alexander Dobin, 2009-2019\n",
      "\n",
      "For more details see:\n",
      "<https://github.com/alexdobin/STAR>\n",
      "<https://github.com/alexdobin/STAR/blob/master/doc/STARmanual.pdf>\n",
      "\n",
      "### versions\n",
      "versionGenome           2.7.1a\n",
      "    string: earliest genome index version compatible with this STAR release. Please do not change this value!\n",
      "\n",
      "### Parameter Files\n",
      "parametersFiles          -\n",
      "    string: name of a user-defined parameters file, \"-\": none. Can only be defined on the command line.\n",
      "\n",
      "### System\n",
      "sysShell            -\n",
      "    string: path to the shell binary, preferably bash, e.g. /bin/bash.\n",
      "                    - ... the default shell is executed, typically /bin/sh. This was reported to fail on some Ubuntu systems - then you need to specify path to bash.\n",
      "\n",
      "### Run Parameters\n",
      "runMode                         alignReads\n",
      "    string: type of the run.\n",
      "\n",
      "                                alignReads             ... map reads\n",
      "                                genomeGenerate         ... generate genome files\n",
      "                                inputAlignmentsFromBAM ... input alignments from BAM. Presently only works with --outWigType and --bamRemoveDuplicates.\n",
      "                                liftOver               ... lift-over of GTF files (--sjdbGTFfile) between genome assemblies using chain file(s) from --genomeChainFiles.\n",
      "\n",
      "runThreadN                      1\n",
      "    int: number of threads to run STAR\n",
      "\n",
      "runDirPerm                      User_RWX\n",
      "    string: permissions for the directories created at the run-time.\n",
      "                                User_RWX ... user-read/write/execute\n",
      "                                All_RWX  ... all-read/write/execute (same as chmod 777)\n",
      "\n",
      "runRNGseed                      777\n",
      "    int: random number generator seed.\n",
      "\n",
      "\n",
      "### Genome Parameters\n",
      "genomeDir                   ./GenomeDir/\n",
      "    string: path to the directory where genome files are stored (for --runMode alignReads) or will be generated (for --runMode generateGenome)\n",
      "\n",
      "genomeLoad                NoSharedMemory\n",
      "    string: mode of shared memory usage for the genome files. Only used with --runMode alignReads.\n",
      "                          LoadAndKeep     ... load genome into shared and keep it in memory after run\n",
      "                          LoadAndRemove   ... load genome into shared but remove it after run\n",
      "                          LoadAndExit     ... load genome into shared memory and exit, keeping the genome in memory for future runs\n",
      "                          Remove          ... do not map anything, just remove loaded genome from memory\n",
      "                          NoSharedMemory  ... do not use shared memory, each job will have its own private copy of the genome\n",
      "\n",
      "genomeFastaFiles            -\n",
      "    string(s): path(s) to the fasta files with the genome sequences, separated by spaces. These files should be plain text FASTA files, they *cannot* be zipped.\n",
      "                            Required for the genome generation (--runMode genomeGenerate). Can also be used in the mapping (--runMode alignReads) to add extra (new) sequences to the genome (e.g. spike-ins).\n",
      "\n",
      "genomeChainFiles            -\n",
      "    string: chain files for genomic liftover. Only used with --runMode liftOver .\n",
      "\n",
      "genomeFileSizes             0\n",
      "    uint(s)>0: genome files exact sizes in bytes. Typically, this should not be defined by the user.\n",
      "\n",
      "genomeConsensusFile         -\n",
      "    string: VCF file with consensus SNPs (i.e. alternative allele is the major (AF>0.5) allele)\n",
      "\n",
      "### Genome Indexing Parameters - only used with --runMode genomeGenerate\n",
      "genomeChrBinNbits           18\n",
      "    int: =log2(chrBin), where chrBin is the size of the bins for genome storage: each chromosome will occupy an integer number of bins. For a genome with large number of contigs, it is recommended to scale this parameter as min(18, log2[max(GenomeLength/NumberOfReferences,ReadLength)]).\n",
      "\n",
      "genomeSAindexNbases         14\n",
      "    int: length (bases) of the SA pre-indexing string. Typically between 10 and 15. Longer strings will use much more memory, but allow faster searches. For small genomes, the parameter --genomeSAindexNbases must be scaled down to min(14, log2(GenomeLength)/2 - 1).\n",
      "\n",
      "genomeSAsparseD             1\n",
      "    int>0: suffux array sparsity, i.e. distance between indices: use bigger numbers to decrease needed RAM at the cost of mapping speed reduction\n",
      "\n",
      "genomeSuffixLengthMax       -1\n",
      "    int: maximum length of the suffixes, has to be longer than read length. -1 = infinite.\n",
      "\n",
      "\n",
      "### Splice Junctions Database\n",
      "sjdbFileChrStartEnd                     -\n",
      "    string(s): path to the files with genomic coordinates (chr <tab> start <tab> end <tab> strand) for the splice junction introns. Multiple files can be supplied wand will be concatenated.\n",
      "\n",
      "sjdbGTFfile                             -\n",
      "    string: path to the GTF file with annotations\n",
      "\n",
      "sjdbGTFchrPrefix                        -\n",
      "    string: prefix for chromosome names in a GTF file (e.g. 'chr' for using ENSMEBL annotations with UCSC genomes)\n",
      "\n",
      "sjdbGTFfeatureExon                      exon\n",
      "    string: feature type in GTF file to be used as exons for building transcripts\n",
      "\n",
      "sjdbGTFtagExonParentTranscript          transcript_id\n",
      "    string: GTF attribute name for parent transcript ID (default \"transcript_id\" works for GTF files)\n",
      "\n",
      "sjdbGTFtagExonParentGene                gene_id\n",
      "    string: GTF attribute name for parent gene ID (default \"gene_id\" works for GTF files)\n",
      "\n",
      "sjdbGTFtagExonParentGeneName            gene_name\n",
      "    string(s): GTF attrbute name for parent gene name\n",
      "\n",
      "sjdbGTFtagExonParentGeneType            gene_type gene_biotype\n",
      "    string(s): GTF attrbute name for parent gene type\n",
      "\n",
      "sjdbOverhang                            100\n",
      "    int>0: length of the donor/acceptor sequence on each side of the junctions, ideally = (mate_length - 1)\n",
      "\n",
      "sjdbScore                               2\n",
      "    int: extra alignment score for alignmets that cross database junctions\n",
      "\n",
      "sjdbInsertSave                          Basic\n",
      "    string: which files to save when sjdb junctions are inserted on the fly at the mapping step\n",
      "\t\t\t\t\tBasic ... only small junction / transcript files\n",
      "\t\t\t\t\tAll   ... all files including big Genome, SA and SAindex - this will create a complete genome directory\n",
      "\n",
      "### Variation parameters\n",
      "varVCFfile                              -\n",
      "    string: path to the VCF file that contains variation data.\n",
      "\n",
      "### Input Files\n",
      "inputBAMfile                -\n",
      "    string: path to BAM input file, to be used with --runMode inputAlignmentsFromBAM\n",
      "\n",
      "### Read Parameters\n",
      "readFilesType               Fastx\n",
      "    string: format of input read files\n",
      "                            Fastx       ... FASTA or FASTQ\n",
      "                            SAM SE      ... SAM or BAM single-end reads; for BAM use --readFilesCommand samtools view\n",
      "                            SAM PE      ... SAM or BAM paired-end reads; for BAM use --readFilesCommand samtools view\n",
      "\n",
      "readFilesIn                 Read1 Read2\n",
      "    string(s): paths to files that contain input read1 (and, if needed,  read2)\n",
      "\n",
      "readFilesPrefix             -\n",
      "    string: preifx for the read files names, i.e. it will be added in front of the strings in --readFilesIn\n",
      "                            -: no prefix\n",
      "\n",
      "readFilesCommand             -\n",
      "    string(s): command line to execute for each of the input file. This command should generate FASTA or FASTQ text and send it to stdout\n",
      "               For example: zcat - to uncompress .gz files, bzcat - to uncompress .bz2 files, etc.\n",
      "\n",
      "readMapNumber               -1\n",
      "    int: number of reads to map from the beginning of the file\n",
      "                            -1: map all reads\n",
      "\n",
      "readMatesLengthsIn          NotEqual\n",
      "    string: Equal/NotEqual - lengths of names,sequences,qualities for both mates are the same  / not the same. NotEqual is safe in all situations.\n",
      "\n",
      "readNameSeparator           /\n",
      "    string(s): character(s) separating the part of the read names that will be trimmed in output (read name after space is always trimmed)\n",
      "\n",
      "clip3pNbases                 0\n",
      "    int(s): number(s) of bases to clip from 3p of each mate. If one value is given, it will be assumed the same for both mates.\n",
      "\n",
      "clip5pNbases                 0\n",
      "    int(s): number(s) of bases to clip from 5p of each mate. If one value is given, it will be assumed the same for both mates.\n",
      "\n",
      "clip3pAdapterSeq            -\n",
      "    string(s): adapter sequences to clip from 3p of each mate.  If one value is given, it will be assumed the same for both mates.\n",
      "\n",
      "clip3pAdapterMMp            0.1\n",
      "    double(s): max proportion of mismatches for 3p adpater clipping for each mate.  If one value is given, it will be assumed the same for both mates.\n",
      "\n",
      "clip3pAfterAdapterNbases    0\n",
      "    int(s): number of bases to clip from 3p of each mate after the adapter clipping. If one value is given, it will be assumed the same for both mates.\n",
      "\n",
      "\n",
      "### Limits\n",
      "limitGenomeGenerateRAM               31000000000\n",
      "    int>0: maximum available RAM (bytes) for genome generation\n",
      "\n",
      "limitIObufferSize                    150000000\n",
      "    int>0: max available buffers size (bytes) for input/output, per thread\n",
      "\n",
      "limitOutSAMoneReadBytes              100000\n",
      "    int>0: max size of the SAM record (bytes) for one read. Recommended value: >(2*(LengthMate1+LengthMate2+100)*outFilterMultimapNmax\n",
      "\n",
      "limitOutSJoneRead                    1000\n",
      "    int>0: max number of junctions for one read (including all multi-mappers)\n",
      "\n",
      "limitOutSJcollapsed                  1000000\n",
      "    int>0: max number of collapsed junctions\n",
      "\n",
      "limitBAMsortRAM                         0\n",
      "    int>=0: maximum available RAM (bytes) for sorting BAM. If =0, it will be set to the genome index size. 0 value can only be used with --genomeLoad NoSharedMemory option.\n",
      "\n",
      "limitSjdbInsertNsj                     1000000\n",
      "    int>=0: maximum number of junction to be inserted to the genome on the fly at the mapping stage, including those from annotations and those detected in the 1st step of the 2-pass run\n",
      "\n",
      "limitNreadsSoft                        -1\n",
      "    int: soft limit on the number of reads\n",
      "\n",
      "### Output: general\n",
      "outFileNamePrefix               ./\n",
      "    string: output files name prefix (including full or relative path). Can only be defined on the command line.\n",
      "\n",
      "outTmpDir                       -\n",
      "    string: path to a directory that will be used as temporary by STAR. All contents of this directory will be removed!\n",
      "            - the temp directory will default to outFileNamePrefix_STARtmp\n",
      "\n",
      "outTmpKeep                      None\n",
      "    string: whether to keep the tempporary files after STAR runs is finished\n",
      "                                None ... remove all temporary files\n",
      "                                All .. keep all files\n",
      "\n",
      "outStd                          Log\n",
      "    string: which output will be directed to stdout (standard out)\n",
      "                                Log                    ... log messages\n",
      "                                SAM                    ... alignments in SAM format (which normally are output to Aligned.out.sam file), normal standard output will go into Log.std.out\n",
      "                                BAM_Unsorted           ... alignments in BAM format, unsorted. Requires --outSAMtype BAM Unsorted\n",
      "                                BAM_SortedByCoordinate ... alignments in BAM format, unsorted. Requires --outSAMtype BAM SortedByCoordinate\n",
      "                                BAM_Quant              ... alignments to transcriptome in BAM format, unsorted. Requires --quantMode TranscriptomeSAM\n",
      "\n",
      "outReadsUnmapped                None\n",
      "   string: output of unmapped and partially mapped (i.e. mapped only one mate of a paired end read) reads in separate file(s).\n",
      "                                None    ... no output\n",
      "                                Fastx   ... output in separate fasta/fastq files, Unmapped.out.mate1/2\n",
      "\n",
      "outQSconversionAdd              0\n",
      "   int: add this number to the quality score (e.g. to convert from Illumina to Sanger, use -31)\n",
      "\n",
      "outMultimapperOrder             Old_2.4\n",
      "    string: order of multimapping alignments in the output files\n",
      "                                Old_2.4             ... quasi-random order used before 2.5.0\n",
      "                                Random              ... random order of alignments for each multi-mapper. Read mates (pairs) are always adjacent, all alignment for each read stay together. This option will become default in the future releases.\n",
      "\n",
      "### Output: SAM and BAM\n",
      "outSAMtype                      SAM\n",
      "    strings: type of SAM/BAM output\n",
      "                                1st word:\n",
      "                                BAM  ... output BAM without sorting\n",
      "                                SAM  ... output SAM without sorting\n",
      "                                None ... no SAM/BAM output\n",
      "                                2nd, 3rd:\n",
      "                                Unsorted           ... standard unsorted\n",
      "                                SortedByCoordinate ... sorted by coordinate. This option will allocate extra memory for sorting which can be specified by --limitBAMsortRAM.\n",
      "\n",
      "outSAMmode                      Full\n",
      "    string: mode of SAM output\n",
      "                                None ... no SAM output\n",
      "                                Full ... full SAM output\n",
      "                                NoQS ... full SAM but without quality scores\n",
      "\n",
      "outSAMstrandField                               None\n",
      "    string: Cufflinks-like strand field flag\n",
      "                                None        ... not used\n",
      "                                intronMotif ... strand derived from the intron motif. Reads with inconsistent and/or non-canonical introns are filtered out.\n",
      "\n",
      "outSAMattributes                Standard\n",
      "    string: a string of desired SAM attributes, in the order desired for the output SAM\n",
      "                                NH HI AS nM NM MD jM jI XS MC ch ... any combination in any order\n",
      "                                None        ... no attributes\n",
      "                                Standard    ... NH HI AS nM\n",
      "                                All         ... NH HI AS nM NM MD jM jI MC ch\n",
      "                                vA          ... variant allele\n",
      "                                vG          ... genomic coordiante of the variant overlapped by the read\n",
      "                                vW          ... 0/1 - alignment does not pass / passes WASP filtering. Requires --waspOutputMode SAMtag\n",
      "                                CR CY UR UY ... sequences and quality scores of cell barcodes and UMIs for the solo* demultiplexing\n",
      "                                Unsupported/undocumented:\n",
      "                                rB          ... alignment block read/genomic coordinates\n",
      "                                vR          ... read coordinate of the variant\n",
      "\n",
      "outSAMattrIHstart               1\n",
      "    int>=0:                     start value for the IH attribute. 0 may be required by some downstream software, such as Cufflinks or StringTie.\n",
      "\n",
      "outSAMunmapped                  None\n",
      "    string(s): output of unmapped reads in the SAM format\n",
      "                                1st word:\n",
      "                                None   ... no output\n",
      "                                Within ... output unmapped reads within the main SAM file (i.e. Aligned.out.sam)\n",
      "                                2nd word:\n",
      "                                KeepPairs ... record unmapped mate for each alignment, and, in case of unsorted output, keep it adjacent to its mapped mate. Only affects multi-mapping reads.\n",
      "\n",
      "outSAMorder                     Paired\n",
      "    string: type of sorting for the SAM output\n",
      "                                Paired: one mate after the other for all paired alignments\n",
      "                                PairedKeepInputOrder: one mate after the other for all paired alignments, the order is kept the same as in the input FASTQ files\n",
      "\n",
      "outSAMprimaryFlag\t\tOneBestScore\n",
      "    string: which alignments are considered primary - all others will be marked with 0x100 bit in the FLAG\n",
      "                                OneBestScore ... only one alignment with the best score is primary\n",
      "                                AllBestScore ... all alignments with the best score are primary\n",
      "\n",
      "outSAMreadID\t\t\tStandard\n",
      "    string: read ID record type\n",
      "                                Standard ... first word (until space) from the FASTx read ID line, removing /1,/2 from the end\n",
      "                                Number   ... read number (index) in the FASTx file\n",
      "\n",
      "outSAMmapqUnique        255\n",
      "    int: 0 to 255: the MAPQ value for unique mappers\n",
      "\n",
      "outSAMflagOR           0\n",
      "    int: 0 to 65535: sam FLAG will be bitwise OR'd with this value, i.e. FLAG=FLAG | outSAMflagOR. This is applied after all flags have been set by STAR, and after outSAMflagAND. Can be used to set specific bits that are not set otherwise.\n",
      "\n",
      "outSAMflagAND           65535\n",
      "    int: 0 to 65535: sam FLAG will be bitwise AND'd with this value, i.e. FLAG=FLAG & outSAMflagOR. This is applied after all flags have been set by STAR, but before outSAMflagOR. Can be used to unset specific bits that are not set otherwise.\n",
      "\n",
      "outSAMattrRGline        -\n",
      "    string(s): SAM/BAM read group line. The first word contains the read group identifier and must start with \"ID:\", e.g. --outSAMattrRGline ID:xxx CN:yy \"DS:z z z\".\n",
      "            xxx will be added as RG tag to each output alignment. Any spaces in the tag values have to be double quoted.\n",
      "            Comma separated RG lines correspons to different (comma separated) input files in --readFilesIn. Commas have to be surrounded by spaces, e.g.\n",
      "            --outSAMattrRGline ID:xxx , ID:zzz \"DS:z z\" , ID:yyy DS:yyyy\n",
      "\n",
      "outSAMheaderHD          -\n",
      "    strings: @HD (header) line of the SAM header\n",
      "\n",
      "outSAMheaderPG          -\n",
      "    strings: extra @PG (software) line of the SAM header (in addition to STAR)\n",
      "\n",
      "outSAMheaderCommentFile -\n",
      "    string: path to the file with @CO (comment) lines of the SAM header\n",
      "\n",
      "outSAMfilter            None\n",
      "    string(s): filter the output into main SAM/BAM files\n",
      "                        KeepOnlyAddedReferences ... only keep the reads for which all alignments are to the extra reference sequences added with --genomeFastaFiles at the mapping stage.\n",
      "                        KeepAllAddedReferences ...  keep all alignments to the extra reference sequences added with --genomeFastaFiles at the mapping stage.\n",
      "\n",
      "\n",
      "outSAMmultNmax          -1\n",
      "    int: max number of multiple alignments for a read that will be output to the SAM/BAM files.\n",
      "                        -1 ... all alignments (up to --outFilterMultimapNmax) will be output\n",
      "\n",
      "outSAMtlen              1\n",
      "    int: calculation method for the TLEN field in the SAM/BAM files\n",
      "                        1 ... leftmost base of the (+)strand mate to rightmost base of the (-)mate. (+)sign for the (+)strand mate\n",
      "                        2 ... leftmost base of any mate to rightmost base of any mate. (+)sign for the mate with the leftmost base. This is different from 1 for overlapping mates with protruding ends\n",
      "\n",
      "outBAMcompression       1\n",
      "    int: -1 to 10  BAM compression level, -1=default compression (6?), 0=no compression, 10=maximum compression\n",
      "\n",
      "outBAMsortingThreadN    0\n",
      "    int: >=0: number of threads for BAM sorting. 0 will default to min(6,--runThreadN).\n",
      "\n",
      "outBAMsortingBinsN      50\n",
      "    int: >0:  number of genome bins fo coordinate-sorting\n",
      "\n",
      "### BAM processing\n",
      "bamRemoveDuplicatesType  -\n",
      "    string: mark duplicates in the BAM file, for now only works with (i) sorted BAM fed with inputBAMfile, and (ii) for paired-end alignments only\n",
      "                        -                       ... no duplicate removal/marking\n",
      "                        UniqueIdentical         ... mark all multimappers, and duplicate unique mappers. The coordinates, FLAG, CIGAR must be identical\n",
      "                        UniqueIdenticalNotMulti  ... mark duplicate unique mappers but not multimappers.\n",
      "\n",
      "bamRemoveDuplicatesMate2basesN   0\n",
      "    int>0: number of bases from the 5' of mate 2 to use in collapsing (e.g. for RAMPAGE)\n",
      "\n",
      "### Output Wiggle\n",
      "outWigType          None\n",
      "    string(s): type of signal output, e.g. \"bedGraph\" OR \"bedGraph read1_5p\". Requires sorted BAM: --outSAMtype BAM SortedByCoordinate .\n",
      "                    1st word:\n",
      "                    None       ... no signal output\n",
      "                    bedGraph   ... bedGraph format\n",
      "                    wiggle     ... wiggle format\n",
      "                    2nd word:\n",
      "                    read1_5p   ... signal from only 5' of the 1st read, useful for CAGE/RAMPAGE etc\n",
      "                    read2      ... signal from only 2nd read\n",
      "\n",
      "outWigStrand        Stranded\n",
      "    string: strandedness of wiggle/bedGraph output\n",
      "                    Stranded   ...  separate strands, str1 and str2\n",
      "                    Unstranded ...  collapsed strands\n",
      "\n",
      "outWigReferencesPrefix    -\n",
      "    string: prefix matching reference names to include in the output wiggle file, e.g. \"chr\", default \"-\" - include all references\n",
      "\n",
      "outWigNorm              RPM\n",
      "    string: type of normalization for the signal\n",
      "                        RPM    ... reads per million of mapped reads\n",
      "                        None   ... no normalization, \"raw\" counts\n",
      "\n",
      "### Output Filtering\n",
      "outFilterType                   Normal\n",
      "    string: type of filtering\n",
      "                                Normal  ... standard filtering using only current alignment\n",
      "                                BySJout ... keep only those reads that contain junctions that passed filtering into SJ.out.tab\n",
      "\n",
      "outFilterMultimapScoreRange     1\n",
      "    int: the score range below the maximum score for multimapping alignments\n",
      "\n",
      "outFilterMultimapNmax           10\n",
      "    int: maximum number of loci the read is allowed to map to. Alignments (all of them) will be output only if the read maps to no more loci than this value.\n",
      "         Otherwise no alignments will be output, and the read will be counted as \"mapped to too many loci\" in the Log.final.out .\n",
      "\n",
      "outFilterMismatchNmax           10\n",
      "    int: alignment will be output only if it has no more mismatches than this value.\n",
      "\n",
      "outFilterMismatchNoverLmax      0.3\n",
      "    real: alignment will be output only if its ratio of mismatches to *mapped* length is less than or equal to this value.\n",
      "\n",
      "outFilterMismatchNoverReadLmax  1.0\n",
      "    real: alignment will be output only if its ratio of mismatches to *read* length is less than or equal to this value.\n",
      "\n",
      "\n",
      "outFilterScoreMin               0\n",
      "    int: alignment will be output only if its score is higher than or equal to this value.\n",
      "\n",
      "outFilterScoreMinOverLread      0.66\n",
      "    real: same as outFilterScoreMin, but  normalized to read length (sum of mates' lengths for paired-end reads)\n",
      "\n",
      "outFilterMatchNmin              0\n",
      "    int: alignment will be output only if the number of matched bases is higher than or equal to this value.\n",
      "\n",
      "outFilterMatchNminOverLread     0.66\n",
      "    real: sam as outFilterMatchNmin, but normalized to the read length (sum of mates' lengths for paired-end reads).\n",
      "\n",
      "outFilterIntronMotifs           None\n",
      "    string: filter alignment using their motifs\n",
      "\t\t\t\tNone                           ... no filtering\n",
      "\t\t\t\tRemoveNoncanonical             ... filter out alignments that contain non-canonical junctions\n",
      "\t\t\t\tRemoveNoncanonicalUnannotated  ... filter out alignments that contain non-canonical unannotated junctions when using annotated splice junctions database. The annotated non-canonical junctions will be kept.\n",
      "\n",
      "outFilterIntronStrands          RemoveInconsistentStrands\n",
      "    string: filter alignments\n",
      "                RemoveInconsistentStrands      ... remove alignments that have junctions with inconsistent strands\n",
      "                None                           ... no filtering\n",
      "\n",
      "### Output Filtering: Splice Junctions\n",
      "outSJfilterReads                All\n",
      "    string: which reads to consider for collapsed splice junctions output\n",
      "                All: all reads, unique- and multi-mappers\n",
      "                Unique: uniquely mapping reads only\n",
      "\n",
      "outSJfilterOverhangMin          30  12  12  12\n",
      "    4 integers:    minimum overhang length for splice junctions on both sides for: (1) non-canonical motifs, (2) GT/AG and CT/AC motif, (3) GC/AG and CT/GC motif, (4) AT/AC and GT/AT motif. -1 means no output for that motif\n",
      "                                does not apply to annotated junctions\n",
      "\n",
      "outSJfilterCountUniqueMin       3   1   1   1\n",
      "    4 integers: minimum uniquely mapping read count per junction for: (1) non-canonical motifs, (2) GT/AG and CT/AC motif, (3) GC/AG and CT/GC motif, (4) AT/AC and GT/AT motif. -1 means no output for that motif\n",
      "                                Junctions are output if one of outSJfilterCountUniqueMin OR outSJfilterCountTotalMin conditions are satisfied\n",
      "                                does not apply to annotated junctions\n",
      "\n",
      "outSJfilterCountTotalMin     3   1   1   1\n",
      "    4 integers: minimum total (multi-mapping+unique) read count per junction for: (1) non-canonical motifs, (2) GT/AG and CT/AC motif, (3) GC/AG and CT/GC motif, (4) AT/AC and GT/AT motif. -1 means no output for that motif\n",
      "                                Junctions are output if one of outSJfilterCountUniqueMin OR outSJfilterCountTotalMin conditions are satisfied\n",
      "                                does not apply to annotated junctions\n",
      "\n",
      "outSJfilterDistToOtherSJmin     10  0   5   10\n",
      "    4 integers>=0: minimum allowed distance to other junctions' donor/acceptor\n",
      "                                does not apply to annotated junctions\n",
      "\n",
      "outSJfilterIntronMaxVsReadN        50000 100000 200000\n",
      "    N integers>=0: maximum gap allowed for junctions supported by 1,2,3,,,N reads\n",
      "                                i.e. by default junctions supported by 1 read can have gaps <=50000b, by 2 reads: <=100000b, by 3 reads: <=200000. by >=4 reads any gap <=alignIntronMax\n",
      "                                does not apply to annotated junctions\n",
      "\n",
      "### Scoring\n",
      "scoreGap                     0\n",
      "    int: splice junction penalty (independent on intron motif)\n",
      "\n",
      "scoreGapNoncan               -8\n",
      "    int: non-canonical junction penalty (in addition to scoreGap)\n",
      "\n",
      "scoreGapGCAG                 -4\n",
      "    GC/AG and CT/GC junction penalty (in addition to scoreGap)\n",
      "\n",
      "scoreGapATAC                 -8\n",
      "    AT/AC  and GT/AT junction penalty  (in addition to scoreGap)\n",
      "\n",
      "scoreGenomicLengthLog2scale   -0.25\n",
      "    extra score logarithmically scaled with genomic length of the alignment: scoreGenomicLengthLog2scale*log2(genomicLength)\n",
      "\n",
      "scoreDelOpen                 -2\n",
      "    deletion open penalty\n",
      "\n",
      "scoreDelBase                 -2\n",
      "    deletion extension penalty per base (in addition to scoreDelOpen)\n",
      "\n",
      "scoreInsOpen                 -2\n",
      "    insertion open penalty\n",
      "\n",
      "scoreInsBase                 -2\n",
      "    insertion extension penalty per base (in addition to scoreInsOpen)\n",
      "\n",
      "scoreStitchSJshift           1\n",
      "    maximum score reduction while searching for SJ boundaries inthe stitching step\n",
      "\n",
      "\n",
      "### Alignments and Seeding\n",
      "\n",
      "seedSearchStartLmax             50\n",
      "    int>0: defines the search start point through the read - the read is split into pieces no longer than this value\n",
      "\n",
      "seedSearchStartLmaxOverLread    1.0\n",
      "    real: seedSearchStartLmax normalized to read length (sum of mates' lengths for paired-end reads)\n",
      "\n",
      "seedSearchLmax       0\n",
      "    int>=0: defines the maximum length of the seeds, if =0 max seed lengthis infinite\n",
      "\n",
      "seedMultimapNmax      10000\n",
      "    int>0: only pieces that map fewer than this value are utilized in the stitching procedure\n",
      "\n",
      "seedPerReadNmax       1000\n",
      "    int>0: max number of seeds per read\n",
      "\n",
      "seedPerWindowNmax     50\n",
      "    int>0: max number of seeds per window\n",
      "\n",
      "seedNoneLociPerWindow    10\n",
      "    int>0: max number of one seed loci per window\n",
      "\n",
      "seedSplitMin                12\n",
      "    int>0: min length of the seed sequences split by Ns or mate gap\n",
      "\n",
      "alignIntronMin              21\n",
      "    minimum intron size: genomic gap is considered intron if its length>=alignIntronMin, otherwise it is considered Deletion\n",
      "\n",
      "alignIntronMax              0\n",
      "    maximum intron size, if 0, max intron size will be determined by (2^winBinNbits)*winAnchorDistNbins\n",
      "\n",
      "alignMatesGapMax            0\n",
      "    maximum gap between two mates, if 0, max intron gap will be determined by (2^winBinNbits)*winAnchorDistNbins\n",
      "\n",
      "alignSJoverhangMin          5\n",
      "    int>0: minimum overhang (i.e. block size) for spliced alignments\n",
      "\n",
      "alignSJstitchMismatchNmax   0 -1 0 0\n",
      "    4*int>=0: maximum number of mismatches for stitching of the splice junctions (-1: no limit).\n",
      "                            (1) non-canonical motifs, (2) GT/AG and CT/AC motif, (3) GC/AG and CT/GC motif, (4) AT/AC and GT/AT motif.\n",
      "\n",
      "alignSJDBoverhangMin        3\n",
      "    int>0: minimum overhang (i.e. block size) for annotated (sjdb) spliced alignments\n",
      "\n",
      "alignSplicedMateMapLmin     0\n",
      "    int>0: minimum mapped length for a read mate that is spliced\n",
      "\n",
      "alignSplicedMateMapLminOverLmate 0.66\n",
      "    real>0: alignSplicedMateMapLmin normalized to mate length\n",
      "\n",
      "alignWindowsPerReadNmax     10000\n",
      "    int>0: max number of windows per read\n",
      "\n",
      "alignTranscriptsPerWindowNmax       100\n",
      "    int>0: max number of transcripts per window\n",
      "\n",
      "alignTranscriptsPerReadNmax               10000\n",
      "    int>0: max number of different alignments per read to consider\n",
      "\n",
      "alignEndsType           Local\n",
      "    string: type of read ends alignment\n",
      "                        Local             ... standard local alignment with soft-clipping allowed\n",
      "                        EndToEnd          ... force end-to-end read alignment, do not soft-clip\n",
      "                        Extend5pOfRead1   ... fully extend only the 5p of the read1, all other ends: local alignment\n",
      "                        Extend5pOfReads12 ... fully extend only the 5p of the both read1 and read2, all other ends: local alignment\n",
      "\n",
      "alignEndsProtrude       0    ConcordantPair\n",
      "    int, string:        allow protrusion of alignment ends, i.e. start (end) of the +strand mate downstream of the start (end) of the -strand mate\n",
      "                        1st word: int: maximum number of protrusion bases allowed\n",
      "                        2nd word: string:\n",
      "                                            ConcordantPair ... report alignments with non-zero protrusion as concordant pairs\n",
      "                                            DiscordantPair ... report alignments with non-zero protrusion as discordant pairs\n",
      "\n",
      "alignSoftClipAtReferenceEnds    Yes\n",
      "    string: allow the soft-clipping of the alignments past the end of the chromosomes\n",
      "                                Yes ... allow\n",
      "                                No  ... prohibit, useful for compatibility with Cufflinks\n",
      "\n",
      "alignInsertionFlush     None\n",
      "    string: how to flush ambiguous insertion positions\n",
      "                        None    ... insertions are not flushed\n",
      "                        Right   ... insertions are flushed to the right\n",
      "\n",
      "### Paired-End reads\n",
      "peOverlapNbasesMin          0\n",
      "    int>=0:             minimum number of overlap bases to trigger mates merging and realignment\n",
      "\n",
      "peOverlapMMp                0.01\n",
      "    real, >=0 & <1:     maximum proportion of mismatched bases in the overlap area\n",
      "\n",
      "### Windows, Anchors, Binning\n",
      "\n",
      "winAnchorMultimapNmax           50\n",
      "    int>0: max number of loci anchors are allowed to map to\n",
      "\n",
      "winBinNbits                     16\n",
      "    int>0: =log2(winBin), where winBin is the size of the bin for the windows/clustering, each window will occupy an integer number of bins.\n",
      "\n",
      "winAnchorDistNbins              9\n",
      "    int>0: max number of bins between two anchors that allows aggregation of anchors into one window\n",
      "\n",
      "winFlankNbins                   4\n",
      "    int>0: log2(winFlank), where win Flank is the size of the left and right flanking regions for each window\n",
      "\n",
      "winReadCoverageRelativeMin      0.5\n",
      "    real>=0: minimum relative coverage of the read sequence by the seeds in a window, for STARlong algorithm only.\n",
      "\n",
      "winReadCoverageBasesMin      0\n",
      "    int>0: minimum number of bases covered by the seeds in a window , for STARlong algorithm only.\n",
      "\n",
      "### Chimeric Alignments\n",
      "chimOutType                 Junctions\n",
      "    string(s): type of chimeric output\n",
      "                            Junctions       ... Chimeric.out.junction\n",
      "                            SeparateSAMold  ... output old SAM into separate Chimeric.out.sam file\n",
      "                            WithinBAM       ... output into main aligned BAM files (Aligned.*.bam)\n",
      "                            WithinBAM HardClip  ... (default) hard-clipping in the CIGAR for supplemental chimeric alignments (defaultif no 2nd word is present)\n",
      "                            WithinBAM SoftClip  ... soft-clipping in the CIGAR for supplemental chimeric alignments\n",
      "\n",
      "chimSegmentMin              0\n",
      "    int>=0: minimum length of chimeric segment length, if ==0, no chimeric output\n",
      "\n",
      "chimScoreMin                0\n",
      "    int>=0: minimum total (summed) score of the chimeric segments\n",
      "\n",
      "chimScoreDropMax            20\n",
      "    int>=0: max drop (difference) of chimeric score (the sum of scores of all chimeric segments) from the read length\n",
      "\n",
      "chimScoreSeparation         10\n",
      "    int>=0: minimum difference (separation) between the best chimeric score and the next one\n",
      "\n",
      "chimScoreJunctionNonGTAG    -1\n",
      "    int: penalty for a non-GT/AG chimeric junction\n",
      "\n",
      "chimJunctionOverhangMin     20\n",
      "    int>=0: minimum overhang for a chimeric junction\n",
      "\n",
      "chimSegmentReadGapMax       0\n",
      "    int>=0: maximum gap in the read sequence between chimeric segments\n",
      "\n",
      "chimFilter                  banGenomicN\n",
      "    string(s): different filters for chimeric alignments\n",
      "                            None ... no filtering\n",
      "                            banGenomicN ... Ns are not allowed in the genome sequence around the chimeric junction\n",
      "\n",
      "chimMainSegmentMultNmax        10\n",
      "    int>=1: maximum number of multi-alignments for the main chimeric segment. =1 will prohibit multimapping main segments.\n",
      "\n",
      "chimMultimapNmax                    0\n",
      "    int>=0: maximum number of chimeric multi-alignments\n",
      "                                0 ... use the old scheme for chimeric detection which only considered unique alignments\n",
      "\n",
      "chimMultimapScoreRange          1\n",
      "    int>=0: the score range for multi-mapping chimeras below the best chimeric score. Only works with --chimMultimapNmax > 1\n",
      "\n",
      "chimNonchimScoreDropMin         20\n",
      "    int>=0: to trigger chimeric detection, the drop in the best non-chimeric alignment score with respect to the read length has to be smaller than this value\n",
      "\n",
      "chimOutJunctionFormat           0\n",
      "    int: formatting type for the Chimeric.out.junction file\n",
      "                                0 ... no comment lines/headers\n",
      "                                1 ... comment lines at the end of the file: command line and Nreads: total, unique, multi\n",
      "\n",
      "### Quantification of Annotations\n",
      "quantMode                   -\n",
      "    string(s): types of quantification requested\n",
      "                            -                ... none\n",
      "                            TranscriptomeSAM ... output SAM/BAM alignments to transcriptome into a separate file\n",
      "                            GeneCounts       ... count reads per gene\n",
      "\n",
      "quantTranscriptomeBAMcompression    1       1\n",
      "    int: -2 to 10  transcriptome BAM compression level\n",
      "                            -2  ... no BAM output\n",
      "                            -1  ... default compression (6?)\n",
      "                             0  ... no compression\n",
      "                             10 ... maximum compression\n",
      "\n",
      "quantTranscriptomeBan       IndelSoftclipSingleend\n",
      "    string: prohibit various alignment type\n",
      "                            IndelSoftclipSingleend  ... prohibit indels, soft clipping and single-end alignments - compatible with RSEM\n",
      "                            Singleend               ... prohibit single-end alignments\n",
      "\n",
      "### 2-pass Mapping\n",
      "twopassMode                 None\n",
      "    string: 2-pass mapping mode.\n",
      "                            None        ... 1-pass mapping\n",
      "                            Basic       ... basic 2-pass mapping, with all 1st pass junctions inserted into the genome indices on the fly\n",
      "\n",
      "twopass1readsN              -1\n",
      "    int: number of reads to process for the 1st step. Use very large number (or default -1) to map all reads in the first step.\n",
      "\n",
      "\n",
      "### WASP parameters\n",
      "waspOutputMode              None\n",
      "    string: WASP allele-specific output type. This is re-implemenation of the original WASP mappability filtering by Bryce van de Geijn, Graham McVicker, Yoav Gilad & Jonathan K Pritchard. Please cite the original WASP paper: Nature Methods 12, 1061–1063 (2015), https://www.nature.com/articles/nmeth.3582 .\n",
      "                            SAMtag      ... add WASP tags to the alignments that pass WASP filtering\n",
      "\n",
      "### STARsolo (single cell RNA-seq) parameters\n",
      "soloType                    None\n",
      "    string(s): type of single-cell RNA-seq\n",
      "                            Droplet   ... one cell barcode and one UMI barcode in read2, e.g. Drop-seq and 10X Chromium\n",
      "\n",
      "soloCBwhitelist             -\n",
      "    string: file with whitelist of cell barcodes\n",
      "\n",
      "soloCBstart                 1\n",
      "    int>0: cell barcode start base\n",
      "\n",
      "soloCBlen                   16\n",
      "    int>0: cell barcode length\n",
      "\n",
      "soloUMIstart                17\n",
      "    int>0: UMI start base\n",
      "\n",
      "soloUMIlen                  10\n",
      "    int>0: UMI length\n",
      "\n",
      "soloBarcodeReadLength       1\n",
      "    int: length of the barcode read\n",
      "                            1   ... equal to sum of soloCBlen+soloUMIlen\n",
      "                            0  ... not defined, do not check\n",
      "\n",
      "soloStrand                  Forward\n",
      "    string: strandedness of the solo libraries:\n",
      "                            Unstranded  ... no strand information\n",
      "                            Forward     ... read strand same as the original RNA molecule\n",
      "                            Reverse     ... read strand opposite to the original RNA molecule\n",
      "\n",
      "soloFeatures                Gene\n",
      "    string(s):              genomic features for which the UMI counts per Cell Barcode are collected\n",
      "                            Gene            ... genes: reads match the gene transcript\n",
      "                            SJ              ... splice junctions: reported in SJ.out.tab\n",
      "                            GeneFull        ... full genes: count all reads overlapping genes' exons and introns\n",
      "\n",
      "soloUMIdedup                1MM_All\n",
      "    string(s):              type of UMI deduplication (collapsing) algorithm\n",
      "                            1MM_All             ... all UMIs with 1 mismatch distance to each other are collapsed (i.e. counted once)\n",
      "                            1MM_Directional     ... follows the \"directional\" method from the UMI-tools by Smith, Heger and Sudbery (Genome Research 2017).\n",
      "                            1MM_NotCollapsed      ... UMIs with 1 mismatch distance to others are not collapsed (i.e. all counted)\n",
      "\n",
      "soloOutFileNames            Solo.out/ genes.tsv barcodes.tsv matrix.mtx matrixSJ.mtx matrixGeneFull.mtx\n",
      "    string(s)               file names for STARsolo output\n",
      "                            1st word    ... file name prefix\n",
      "                            2nd word    ... gene IDs and names\n",
      "                            3rd word    ... barcode sequences\n",
      "                            4th word    ... cell/Gene counts matrix\n",
      "                            5th word    ... cell/SJ counts matrix\n",
      "                            6th word    ... cell/GeneFull counts matrix\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!STAR --help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "cf13f8bf-ad3d-4d69-9e49-b60d7125a5f5",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Oct 20 06:51:36 ..... started STAR run\n",
      "Oct 20 06:51:36 ... starting to generate Genome files\n",
      "Oct 20 06:53:15 ... starting to sort Suffix Array. This may take a long time...\n",
      "Oct 20 06:53:42 ... sorting Suffix Array chunks and saving them to disk...\n",
      "Oct 20 07:49:02 ... loading chunks from disk, packing SA...\n",
      "Oct 20 07:50:51 ... finished generating suffix array\n",
      "Oct 20 07:50:51 ... generating Suffix Array index\n",
      "Oct 20 07:55:07 ... completed Suffix Array index\n",
      "Oct 20 07:55:07 ..... processing annotations GTF\n",
      "Oct 20 07:55:33 ..... inserting junctions into the genome indices\n",
      "Oct 20 08:01:19 ... writing Genome to disk ...\n",
      "Oct 20 08:01:24 ... writing Suffix Array to disk ...\n",
      "Oct 20 08:02:11 ... writing SAindex to disk\n",
      "Oct 20 08:02:16 ..... finished successfully\n"
     ]
    }
   ],
   "source": [
    "## Prepare the reference genome index file using STAR\n",
    "!STAR --runMode genomeGenerate \\\n",
    "     --genomeDir /tutorial/sample_data/Ref/GRCh38_gencode_v37_CTAT_lib_Mar012021.source/ \\\n",
    "     --genomeFastaFiles /tutorial/sample_data/Ref/GRCh38_gencode_v37_CTAT_lib_Mar012021.source/GRCh38.primary_assembly.genome.fa \\\n",
    "     --sjdbGTFfile /tutorial/sample_data/Ref/GRCh38_gencode_v37_CTAT_lib_Mar012021.source/gencode.v37.annotation.gtf \\\n",
    "     --sjdbOverhang 100 \\\n",
    "     --runThreadN 16"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ffe5ceb5-ffb8-4889-b0eb-0047783c69e4",
   "metadata": {},
   "source": [
    "#### Run rna_fq2bam for Gene Fusion Detection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c7b4006-3622-4829-9eb5-e6ed209c471d",
   "metadata": {},
   "outputs": [],
   "source": [
    "!pbrun rna_fq2bam \\\n",
    "    --genome-lib-dir sample_data/Ref/GRCh38_gencode_v37_CTAT_lib_Mar012021.source \\\n",
    "    --ref sample_data/Ref/GRCh38_gencode_v37_CTAT_lib_Mar012021.source/GRCh38.primary_assembly.genome.fa \\\n",
    "    --in-fq sample_data/Data/SRR1217085_1.fastq.gz.20M.fq.gz sample_data/Data/SRR1217085_2.fastq.gz.20M.fq.gz\\\n",
    "    --out-bam outputdir/rna_fq2bam.bam \\\n",
    "    --output-dir /utputdir "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf4a4d30-d57d-4332-89ec-68294882da14",
   "metadata": {},
   "source": [
    "#### Run Gene Fusion Detection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "78676329-51d7-46dd-8df4-c2e95b3d7ad6",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please visit https://docs.nvidia.com/clara/#parabricks for detailed documentation\n",
      "\n",
      "[PB Info 2024-Oct-20 15:49:58] ------------------------------------------------------------------------------\n",
      "[PB Info 2024-Oct-20 15:49:58] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2024-Oct-20 15:49:58] ||                              Version 4.3.2-1                             ||\n",
      "[PB Info 2024-Oct-20 15:49:58] ||                                starfusion                                ||\n",
      "[PB Info 2024-Oct-20 15:49:58] ------------------------------------------------------------------------------\n",
      "[PB Info 2024-Oct-20 15:49:58] Running starfusion...\n",
      "[PB Info 2024-Oct-20 15:49:58] Reading \"ref_annot.gtf.gene_spans\"...\n",
      "[PB Info 2024-Oct-20 15:49:58] Reading \"blast_pairs.idx\"...\n",
      "[PB Info 2024-Oct-20 15:49:58] Reading \"trans.blast.align_coords.align_coords.dat\"...\n",
      "[PB Info 2024-Oct-20 15:49:58] Reading reference file...\n",
      "[PB Info 2024-Oct-20 15:49:58] Reading \"fusion_annot_lib.idx\"...\n",
      "[PB Info 2024-Oct-20 15:49:58] Running map_chimeric_reads_to_genes...\n",
      "[PB Info 2024-Oct-20 15:49:58] Building interval tree...\n",
      "[PB \u001b[31mError\u001b[0m 2024-Oct-20 15:49:58][./inc/StarFusion.h:317] Cannot open file \"/tutorial/sample_data/Ref/GRCh38_gencode_v37_CTAT_lib_Mar012021.source/blast_pairs.idx\"No such file or directory, exiting.\n",
      "For technical support visit https://docs.nvidia.com/clara/index.html#parabricks\n",
      "Exiting...\n",
      "\n",
      "Could not run starfusion\n",
      "Exiting pbrun ...\n"
     ]
    }
   ],
   "source": [
    "!pbrun starfusion \\\n",
    "    --output-dir outputdir/starfusion_output \\\n",
    "    --genome-lib-dir sample_data/Ref/GRCh38_gencode_v37_CTAT_lib_Mar012021.source \\\n",
    "    --chimeric-junction outputdir/Chimeric.out.junction"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
